---
title: "Development Model Width"
author: "Sebastien Lacube"
date: "14 April 2017"
output: 
  html_document: 
    toc: yes
---


```{r PATHS, fig.height=5, fig.width=6, message=FALSE, warning=FALSE, include=FALSE}

# Paths ----
PATH.DeaVolga <- "D:/R-epositery/Widths/3.Data/MeasureDeaVolga.csv"
PATH.PhylloDV <- "D:/R-epositery/Widths/3.Data/PhylloDeaVolga.csv"
PATH.DeaVolgaData <- "D:/R-epositery/Widths/3.Data/Data_DeaVolga.csv"
PATH.DeaVolgaMet <- "D:/R-epositery/Widths/3.Data/MetDeaVolga.csv"


```

```{r Sources, fig.height=5, fig.width=6, message=FALSE, warning=FALSE, include=FALSE}


# Libraries ----
library(dplyr)
library(maptools)
library(rworldmap)
library(ggplot2)
library(RColorBrewer)
library(akima)
library(ggmap)
library(rgdal)
library(fields)
library(ggthemes)
library(ggThemeAssist)
library(raster)
library(dismo)
library(ggplot2)
library(RColorBrewer)
library(nlme)
library(gridExtra)
library(reshape2)
library(APSIM)
library(FactoMineR)
library(car)
library(knitr)
library(ggrepel)
library(PerformanceAnalytics)
library(corrplot)
library(devtools)
library(factoextra)
library(plyr)
library(sensitivity)
library(solaR)
library(lubridate)
library(lattice)
library(ReporteRs)
library(Hmisc)
library(scales)
library(ggcorrplot)
library(RSelenium)
library(zoo)
library(maptools)
library(rworldmap)
library(ggplot2)
library(RColorBrewer)
library(akima)
library(ggmap)
library(rgdal)
library(fields)
library(ggthemes)
library(ggThemeAssist)
library(raster)
library(dismo)
library(XML)


# Sources ----
source('D:/Work/R script/APSIM/MultiSimFunctions.r')
source('D:/Work/R script/Sources/Sources.r')
source('D:/Work/R script/Sources/LERmodel.r')
source('D:/Work/R script/Sources/ggplot_smooth_func.R')
source('D:/Work/R script/Sources/ThermalTime.r')
source('D:/Work/R script/Sources/suncourse.r')
source('D:/Work/R script/Sources/PM-FT-function.R')
source('D:/Work/R script/Sources/PM-FAO56-function.r')

# Functions ----
theme_map <- function(base_size = 8, base_family = "") {
  theme_bw(base_size = base_size, base_family = base_family) %+replace%
    theme(axis.line = element_blank(),
          axis.text = element_blank(),
          axis.ticks = element_blank(),
          axis.title = element_blank(),
          panel.background = element_blank(),
          panel.border = element_blank(),
          panel.grid = element_blank(),
          panel.margin = unit(0, "lines"),
          plot.background = element_blank(),
          legend.position = "bottom",
          legend.direction="horizontal",
          legend.title=element_blank(),
          legend.text=element_text(size=8),
          legend.box='horizontal')
}

## Getting all data for harvest dates only (max DAS):
Get_Harvest_Data <- function(Output_Sim,By)
{
  HarvestDateSim <-   
    do.call("rbind",lapply(split(Output_Sim,By,drop=T),
                         function(x)
                         {
                           Temp<-subset(x, DaysAfterSowing==max(x$DaysAfterSowing))
                           return(Temp)
                         }
                       ) ## End lapply
                      ) ## End do.call rbind  
  return(HarvestDateSim)
}


```

# Widths at low radiation in relation to final leaf number

## DataBase CINCALLI (elites) :

<span style="color:red">Available data :</span> 
- Final width/length profiles in the field
- Elite genotypes with various final leaf number

```{r  Load Data - CINCALLI, echo=FALSE, fig.height=5, fig.width=6, message=FALSE, warning=FALSE}

# Cincalli - Data   ---- 
Cincalli <- read.table(file ='D:/Work/Widths/3.Data/plantfichierSelectionOut311.csv',sep=";",
                        header=TRUE,dec=".",
                        row.names=NULL,as.is=FALSE,fill=TRUE)

Cincalli <- subset(Cincalli, idTrial =='LM07' 
                   | idTrial =='MA06'
                   | idTrial =='MA07'
                   | idTrial =='NR06'
                   | idTrial =='SP06')
```
  
### Show data :  
  
Length 'loess' per leaf number  
  
```{r Analyse CINCALLI curves - Length, echo=FALSE, fig.height=4.5, fig.width=5.5, message=FALSE, warning=FALSE}


Melt <- melt(Cincalli,measure.vars=colnames(Cincalli)[grepl("L", names(Cincalli))])
Subsetted <- subset(Melt,!is.na(value),select=c('genotypeName','rep','variable','plantEntry','value'))
Subsetted$Leaf <- substr(Subsetted$variable,2,3)
Aggregated <- aggregate(value~Leaf+genotypeName+Leaf,Subsetted,mean,na.rm=T)
Aggregated$Length <- Aggregated$value/10
Aggregated$leaf <-  as.numeric(as.character(Aggregated$Leaf))
Nfinal <- aggregate(as.numeric(leaf)~genotypeName,Aggregated,FUN=max,na.rm=T)
colnames(Nfinal) <- c('genotypeName','Nfinal')
Aggregated_Nfinal <- merge(Aggregated,Nfinal)
Ordered <- Aggregated_Nfinal[order(Aggregated_Nfinal$Nfinal,Aggregated_Nfinal$genotypeName,Aggregated_Nfinal$leaf),] 
GroupBy <- 1
Ordered$Group <- round(Ordered$Nfinal/GroupBy,0)
Ordered$Group <- cut(as.numeric(Ordered$Nfinal), seq(0,30,GroupBy), include.lowest=TRUE)
#levels(Ordered$Group) <- seq(min(Ordered$Group),length(unique(Ordered$Group)),1)
Amax <- aggregate(Length~genotypeName,Ordered,max,na.rm=T)
colnames(Amax) <- c('genotypeName','Amax')
OrderedAmax <- merge(Ordered,Amax) 
OrderedAmax$NormLength <- OrderedAmax$Length/OrderedAmax$Amax


DataLengthForFit <- data.frame(Nfinal=as.numeric(),Length=as.numeric(),rank=as.numeric())
for (Nff in unique(OrderedAmax$Nfinal))
{
  ToTreat <- subset(OrderedAmax,Nfinal==Nff)
  LoessModel <- loess(ToTreat,formula = 'Length~leaf')
  LeafRanks <- seq(1,Nff,1)
  Lengths <- predict(LoessModel,LeafRanks)
  DataLengthForFit <- rbind(DataLengthForFit,data.frame(Nfinal=Nff,
                                                      Length=Lengths,
                                                      rank=LeafRanks))
}

ggplot(data=OrderedAmax,aes(x=leaf,y=Length,color=factor(Group)))+
  geom_point(size=1.5)+
  theme(legend.position='none')+
  geom_smooth(size=2,se=F)+
  theme_light()+
  scale_color_discrete(name  ="Nfinal")+
  theme_light()+
  scale_y_continuous(limits=c(0,110))+
  labs(x='Leaf rank',y='Length (cm)')

Lengths_Cincalli <- OrderedAmax

```
  
Width 'loess' per leaf number  
  
```{r Analyse CINCALLI curves - Width, echo=FALSE, fig.height=4.5, fig.width=5.5, message=FALSE, warning=FALSE}


Melt <- melt(Cincalli,measure.vars=colnames(Cincalli)[grepl("W", names(Cincalli))])
Subsetted <- subset(Melt,!is.na(value),select=c('genotypeName','rep','variable','plantEntry','value'))
Subsetted$Leaf <- substr(Subsetted$variable,2,3)
Aggregated <- aggregate(value~Leaf+genotypeName+Leaf,Subsetted,mean,na.rm=T)
Aggregated$Width <- Aggregated$value/10
Aggregated$leaf <-  as.numeric(as.character(Aggregated$Leaf))
Nfinal <- aggregate(as.numeric(leaf)~genotypeName,Aggregated,FUN=max,na.rm=T)
colnames(Nfinal) <- c('genotypeName','Nfinal')
Aggregated_Nfinal <- merge(Aggregated,Nfinal)
Ordered <- Aggregated_Nfinal[order(Aggregated_Nfinal$Nfinal,Aggregated_Nfinal$genotypeName,Aggregated_Nfinal$leaf),] 
GroupBy <- 1
Ordered$Group <- round(Ordered$Nfinal/GroupBy,0)
Ordered$Group <- cut(as.numeric(Ordered$Nfinal), seq(0,30,GroupBy), include.lowest=TRUE)
#levels(Ordered$Group) <- seq(min(Ordered$Group),length(unique(Ordered$Group)),1)
Amax <- aggregate(Width~genotypeName,Ordered,max,na.rm=T)
colnames(Amax) <- c('genotypeName','Amax')
OrderedAmax <- merge(Ordered,Amax) 
OrderedAmax$NormWidth <- OrderedAmax$Width/OrderedAmax$Amax

ggplot(data=OrderedAmax,aes(x=leaf,y=Width,color=factor(Group)))+
  geom_point(size=1.5)+
  theme(legend.position='none')+
  geom_smooth(size=2,se=F)+
  scale_color_discrete(name  ="Nfinal")+
  theme_light()+
  scale_y_continuous(limits=c(0,10))+
  labs(x='Leaf rank',y='Width (cm)')

DataWidthForFit <- data.frame(Nfinal=as.numeric(),Width=as.numeric(),rank=as.numeric())
for (Nff in unique(OrderedAmax$Nfinal))
{
  ToTreat <- subset(OrderedAmax,Nfinal==Nff)
  LoessModel <- loess(ToTreat,formula = 'Width~leaf')
  LeafRanks <- seq(1,Nff,1)
  Widths <- predict(LoessModel,LeafRanks)
  DataWidthForFit <- rbind(DataWidthForFit,data.frame(Nfinal=Nff,
                                                      Width=Widths,
                                                      rank=LeafRanks))
}

Widths_Cincalli <- OrderedAmax


```

Leaf area 'loess' per leaf number  
  
```{r Analyse CINCALLI curves - Leaf Area, echo=TRUE, fig.height=4.5, fig.width=5.5, message=FALSE, warning=FALSE}

Merged_Cincalli <- merge(Widths_Cincalli,Lengths_Cincalli,by=c('genotypeName','Leaf'),all.x=T,all.y=F)
Merged_Cincalli$LeafArea <- Merged_Cincalli$Width*Merged_Cincalli$Length*0.75


ggplot(data=Merged_Cincalli,aes(x=leaf.x,y=LeafArea,color=factor(Group.x)))+
  geom_point(size=1.5)+
  theme(legend.position='none')+
  geom_smooth(size=2,se=F)+
  theme_light()+
  scale_color_discrete(name  ="Nfinal")+
  theme_light()+
  labs(x='Leaf rank',y='Leaf Area (cm²)')

```

```{r Simulate LERmax with Beta, eval=FALSE, fig.height=4, fig.width=4, message=FALSE, warning=FALSE, include=FALSE}

gg_color_hue <- function(n) {
  hues = seq(15, 375, length = n + 1)
  hcl(h = hues, l = 65, c = 100)[1:n]
}

######### LER max  : 

## Fit for both with Dea/Volga % Nfinal : ----
nlsfit <- nls( Length~  Nfinal + Amax * exp(-0.5 * (((arank*Nfinal) - (rank))/(brank*Nfinal))^2)  ,
               DataLengthForFit,start=list(arank=0.6,brank=0.3,Amax=80 ))

DataLengthForFit$LengthSim  <- predict(nlsfit,DataLengthForFit)
 
# Arank and Brank are going to be the same
Arank <- summary(nlsfit)$parameters[1,1]
Sigma <- summary(nlsfit)$parameters[2,1]

# Wmax will change with the calculations 
Lmax <- summary(nlsfit)$parameters[3,1]
Rsquare <- summary(nlsfit)$r.squared

ToPlot <- subset(DataLengthForFit,Nfinal >15 & Nfinal <20 & rank > 5)
cols = gg_color_hue(15)


A <- ggplot(data=Lengths_Cincalli,aes(x=leaf,y=Length,color=factor(Group)))+
        geom_point(size=0.5)+
        geom_smooth(size=1,se=F)+
        theme_these()+
        scale_color_discrete(name  ="Nfinal")+
        scale_y_continuous(limits=c(0,110))+
        labs(x='Leaf number',y='Length (cm)')+
        theme(legend.position='none')

B <- ggplot()+
        geom_line(data=ToPlot,aes(x=rank,y=Length,color=factor(Nfinal)),size=1,linetype='solid')+
        geom_line(data=ToPlot,aes(x=rank,y=LengthSim,color=factor(Nfinal)),
                  size=1,linetype='dashed')+
        geom_point(data=ToPlot,aes(x=rank,y=LengthSim,
                  color=factor(Nfinal)),size=2,shape=21)+
        facet_wrap(~Nfinal,ncol=2)+
        theme_these()+
        labs(x='Leaf number',y='Length (cm)')+
        scale_y_continuous(limits=c(20,80),breaks=seq(20,80,40))+
        scale_color_manual(values=c(cols[4:7]))+
        theme(legend.position='none')

C <- ggplot()+
        geom_point(data=DataLengthForFit,
                   aes(x=LengthSim,y=Length,color=factor(Nfinal)),
                   size=2.5,shape=21)+
        theme_light()+
        theme(legend.position='none')+
        labs(x='Sim. length (cm)',y='Meas. length (cm)')+
        scale_x_continuous(limits=c(0,110))+
        scale_y_continuous(limits=c(0,110))+
        theme_these()+
        geom_abline(intercept=0,slope=1)
        


########## Width Min :
D <- ggplot(data=Widths_Cincalli,aes(x=leaf,y=Width,color=factor(Group)))+
        geom_point(size=0.5)+
        geom_smooth(size=1,se=F)+
        theme_these()+
        scale_color_discrete(name  ="Nfinal")+
        scale_y_continuous(limits=c(0,12))+
        labs(x='Leaf number',y='Width (cm)')+
        theme(legend.position='none')

Simulated <- data.frame()
for (nff in unique(DataWidthForFit$Nfinal))
{
  Simulated <- rbind(Simulated,SimulateWidthAtLowRad(nff))
}
colnames(Simulated) <- c('rank','WidthSim','Nfinal')

ToPlot <- merge(DataWidthForFit,Simulated,all.x=T,all.y=F)
ToPlot <- subset(ToPlot,Nfinal >15 & Nfinal <20 & rank > 5)

E <- ggplot()+
        geom_line(data=ToPlot,aes(x=rank,y=Width,color=factor(Nfinal)),size=1,linetype='solid')+
        geom_line(data=ToPlot,aes(x=rank,y=WidthSim,color=factor(Nfinal)),
                  size=1,linetype='dashed')+
        geom_point(data=ToPlot,aes(x=rank,y=WidthSim,
                  color=factor(Nfinal)),size=2,shape=21)+
        facet_wrap(~Nfinal,ncol=2)+
        theme_these()+
        labs(x='Leaf number',y='Width (cm)')+
        scale_y_continuous(limits=c(2,10),breaks=seq(2,10,4))+
        scale_color_manual(values=c(cols[4:7]))+
        theme(legend.position='none')

ToPlot <- merge(DataWidthForFit,Simulated,all.x=T,all.y=F)
ToPlot <- subset(ToPlot,Nfinal >15 & Nfinal <24 & rank > 5)

G <- ggplot()+
        geom_point(data=DataWidthForFit,
                   aes(x=WidthSim,y=Width,color=factor(Nfinal)),
                   size=2.5,shape=21)+
        theme_light()+
        theme(legend.position='none')+
        labs(x='Sim. width (cm)',y='Meas. width (cm)')+
        scale_x_continuous(limits=c(0,12))+
        scale_y_continuous(limits=c(0,12))+
        theme_these()+
        geom_abline(intercept=0,slope=1)





# Produce the output in pwp : 
library( ReporteRs )
mydoc = pptx(  )
mydoc = addSlide( mydoc, slide.layout = "Title and Content" )


  Left <- plot_grid(A,B,C,ncol=1,labels=c('A','C','E'),rel_heights = c(0.3,0.4,0.3))
  Right <- plot_grid(D,E,G,labels=c('B','D','F'),ncol=1,rel_heights = c(0.3,0.4,0.3))

  myplot = plot_grid(Left,Right,ncol=2,rel_widths = c(0.5,0.5))




mydoc = addPlot( mydoc, function( ) print( myplot ), vector.graphic=T, offx=0.5,offy=0.5,width=7, height=9, pointsize=20)  
setwd('D:/Writings/THESIS/Fig')
writeDoc( mydoc, file = "D:/Writings/THESIS/Fig/CincalliStudy.pptx" )


```

```{r Weibull function, eval=FALSE, fig.height=5, fig.width=5, message=FALSE, warning=FALSE, include=FALSE}

## Fit for both with Dea/Volga % Nfinal : ----
nlsfit <- nls( Width~ Base_Width + Amax * exp(-0.5 * (((arank*Nfinal) - (rank))/(brank*Nfinal))^2)  ,
               DataWidthForFit,start=list(arank=0.625,brank=0.375,Amax=9, Base_Width=2),trace=TRUE)
DataWidthForFit$WidthSim  <- predict(nlsfit,DataWidthForFit)
 
ggplot()+
  geom_line(data=DataWidthForFit,aes(x=rank,y=Width),color='red',size=1,linetype='solid')+
  geom_line(data=DataWidthForFit,aes(x=rank,y=WidthSim),color='blue',size=1,linetype='dashed')+
  geom_point(data=OrderedAmax,aes(x=leaf,y=Width),size=2,shape=20)+
  facet_wrap(~Nfinal,scales='free')+
  theme(legend.position='none')+
  theme_light()


# Produce the output in pwp : 
library( ReporteRs )
mydoc = pptx(  )
mydoc = addSlide( mydoc, slide.layout = "Title and Content" )


myplot <- ggplot()+
        geom_point(data=DataWidthForFit,
                   aes(x=WidthSim,y=Width,color=factor(Nfinal)),
                   size=2.5,shape=21)+
        theme_light()+
        theme(legend.position='bottom')+
        labs(x='Sim. width (cm)',y='Meas. width (cm)')+
        scale_x_continuous(limits=c(0,12))+
        scale_y_continuous(limits=c(0,12))+
        geom_abline(intercept=0,slope=1)


mydoc = addPlot( mydoc, function( ) print( myplot ), vector.graphic=T, offx=0.5,offy=0.5,width=5, height=5, pointsize=20)  
setwd('D:/Writings/THESIS/Fig')
writeDoc( mydoc, file = "D:/Writings/THESIS/Fig/Legend_CincalliStudy.pptx" )


```
  
### Find a common fit for all leaf number  
  
```{r Show data, eval=FALSE, fig.height=7, fig.width=9, message=FALSE, warning=FALSE, include=FALSE}


# Show Data ----
ggplot(data=DataWidthForFit,aes(x=rank,y=Width,color=factor(Nfinal)))+
  geom_point(size=3)+
  geom_line()+
  theme_light()+
  scale_color_discrete(name  ="Nfinal")+
  scale_y_continuous(limits=c(0,10),breaks=seq(0,10,1))+
  scale_x_continuous(limits=c(0,30),breaks=seq(0,30,1))+
  coord_flip()+
  labs(x='Leaf rank',y='Width (cm)')+
  geom_hline(yintercept = Wmax)

```

<span style="color:blue">Resulted function :</span>  
1- First leaves constant width  
2- Growth of leaf width  
3- Width 'plateau' around leaf with max width (relative to nff)  
4- Decrease of leaf width for last leaves (slope % upward one)  

<span style="color:blue">Input is :</span>  
W_min = 1.5 ==> Width of small leaves  
L_rank = 0.2 ==> Limit of small leaves (%nff)  
Rank_Max = 0.66 ==> Rank of leaf with max width  
W_max = 7.8 ==> Width of max leaf  
Width_Curve = 0.29 ==> Size of plateau (%nff)  
SlopeScale = 0.25 ==> Scale between upward and downward slope   

<span style="color:blue">Output is :</span>  
DataSim $Leaf     ==> Leaf number/rank  
        $SimWidth ==> Estimated Wmin  
        $Nfinal   ==> Max number of leaves  
        
```{r Fit with custom function, echo=FALSE, fig.height=6, fig.width=8, message=FALSE, warning=FALSE}

# Curve parameters : ----
Parameters <- data.frame(Rank_Max=Arank,Width_Curve=Sigma,W_max=Wmax)
Parameters$W_min <- 2
Parameters$L_rank <- 0.2
Parameters$SlopeScale <- 0.25

# Fit per part ----
Outputs <- data.frame(Leaf = as.numeric(),
                      MeasWidth = as.numeric(),
                      SimWidth = as.numeric(),
                      Nfinal = as.numeric() )
for (nff in unique(DataWidthForFit$Nfinal))
{
  
  DataToTreat <- subset(DataWidthForFit,Nfinal==nff)
  DataSim <- data.frame(Leaf = seq(1,nff,1),
                        MeasWidth = DataToTreat$Width,
                        SimWidth = rep(NA,nff),
                        Nfinal = rep(nff,nff))
  
  for (leaf in seq(1,nff,1))
  {
    
    SemiPlateau <- (round(Parameters$Width_Curve*nff)/2)
    
    # First non mature leaves : 
    if (leaf <= round(Parameters$L_rank*nff))
    {
      
      DataSim$SimWidth[leaf] <- Parameters$W_min
      
    # Upward :
    } else if (leaf <= round(Parameters$Rank_Max*nff)-SemiPlateau){
      
      UpSlope <- (Parameters$W_max -Parameters$W_min ) / (round(Parameters$Rank_Max*nff)-SemiPlateau-round(Parameters$L_rank*nff))
      DataSim$SimWidth[leaf] <- Parameters$W_min + (UpSlope * (leaf-round(Parameters$L_rank*nff)))
    
    # Plateau :
    } else if (leaf <= round(Parameters$Rank_Max*nff)+SemiPlateau-1){

      DataSim$SimWidth[leaf] <- Parameters$W_max

    # Downward :
    } else {
      
      LimSup <- round(Parameters$Rank_Max*nff)+SemiPlateau-1
      DownSlope <- UpSlope*Parameters$SlopeScale
      DataSim$SimWidth[leaf] <- DataSim$SimWidth[leaf-1] - (DownSlope * (leaf-LimSup))
      
    }

  }
  
  Outputs <- rbind(Outputs,DataSim)

}


# Show Results ----

ToPlot <- melt(Outputs,measure.vars=c('MeasWidth','SimWidth'))

ggplot(data=ToPlot,aes(x=Leaf,y=value,linetype=variable))+
  geom_line(size=1.2)+
  facet_wrap(~Nfinal)+
  theme_light()+
  scale_y_continuous(limits=c(0,10),breaks=seq(0,10,1))+
  scale_x_continuous(limits=c(0,30),breaks=seq(0,30,3))+
  labs(x='Leaf rank',y='Width (cm)')+
  geom_hline(yintercept = Parameters$W_max)+
  geom_hline(yintercept = Parameters$W_min)+
  scale_linetype_discrete(name='Width',labels=c('Estimated','Simulated'))

```
  
```{r Function : SimulateWidthAtLowRad, echo=TRUE, fig.height=6, fig.width=8, message=FALSE, warning=FALSE}

SimulateWidthAtLowRad <- function( nff )
{
  
  Parameters <- data.frame(W_min = 1.5 ,
                           L_rank = 0.2,
                           Rank_Max = 0.66,
                           W_max = 7.8,
                           Width_Curve = 0.29,
                           SlopeScale = 0.25)
  
  DataSim <- data.frame(Leaf = seq(1,nff,1),
                        SimWidth = rep(NA,nff),
                        Nfinal = rep(nff,nff))
  
  for (leaf in seq(1,nff,1))
  {
    
    SemiPlateau <- (round(Parameters$Width_Curve*nff)/3)
    
    # First non mature leaves : 
    if (leaf <= round(Parameters$L_rank*nff))
    {
      
      DataSim$SimWidth[leaf] <- Parameters$W_min
      
    # Upward :
    } else if (leaf <= round(Parameters$Rank_Max*nff)-SemiPlateau){
      
      Num <- (Parameters$W_max -Parameters$W_min)
      Den <- round(Parameters$Rank_Max*nff)-SemiPlateau-round(Parameters$L_rank*nff)
      UpSlope <- Num / Den
        
      DataSim$SimWidth[leaf] <- Parameters$W_min + (UpSlope * (leaf-round(Parameters$L_rank*nff)))
      
    # Plateau :
    } else if (leaf <= round(Parameters$Rank_Max*nff)+SemiPlateau-1){
      
      DataSim$SimWidth[leaf] <- Parameters$W_max
      
    # Downward :
    } else {
      
      LimSup <- round(Parameters$Rank_Max*nff)+SemiPlateau-1
      DownSlope <- UpSlope*Parameters$SlopeScale
      DataSim$SimWidth[leaf] <- DataSim$SimWidth[leaf-1] - (DownSlope * (leaf-LimSup))
      
    }
    
  }
  
  return(DataSim)
}


```
  
## Dea / Volga :  
  
<span style="color:red">Available data :</span>  
- Lengths / Widths measurements every two days *for all leaves*  
- *Phyllochron* and *ligulation* every two days  

```{r Plot DEA/VOLGA/B73 model, fig.height=10, fig.width=4.5, message=FALSE, warning=FALSE, include=FALSE}

# Functions ----
theme_these <- 
  function(base_size = 16, base_family = "") {
  theme_bw(base_size = base_size, base_family = base_family) %+replace%
    theme(panel.background = element_blank(),
          plot.background = element_blank(),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          legend.position = "none",
          legend.direction="horizontal",
          legend.text=element_text(size=16),
          plot.title = element_text(hjust = 0.5,margin = margin(b = + 10)),
          strip.background = element_blank(),
          strip.text.x = element_blank())
  }

# Load data ----
DeaVolga <- read.table(PATH.DeaVolga, 
                       sep=";",                                           
                       header=TRUE,dec=".",
                       row.names=NULL,
                       as.is=FALSE, 
                       fill=TRUE)

Phyllo <- read.table(PATH.PhylloDV, 
                       sep=";",                                           
                       header=TRUE,dec=".",
                       row.names=NULL,
                       as.is=FALSE, 
                       fill=TRUE)

Data <- read.table(PATH.DeaVolgaData, 
                       sep=";",                                           
                       header=TRUE,dec=".",
                       row.names=NULL,
                       as.is=FALSE, 
                       fill=TRUE)

# Get Interception coefficients  ----
InterceptedRAD <- read.table(file ='D:/Work/Widths/3.Data/ArchMeteoDaily_RADi.csv',
                    sep=",",                                           
                    header=TRUE,dec=".",
                    row.names=NULL,
                    as.is=FALSE, 
                    fill=TRUE)



# Recenter at emergence ----
Data$TT[1:96] <- Data$TT[1:96]-150

# Load Met ----
MetDea <- read.table(PATH.DeaVolgaMet, 
                       sep=";",                                           
                       header=TRUE,dec=".",
                       row.names=NULL,
                       as.is=FALSE, 
                       fill=TRUE)
 

# Dea Model ----
df.Dev_Stages.Dea <- Leaf_dev_stages(16,2,0.0248,35,0.016,5.4,2*0.0248,4.5)
df.Dev_Stages.Dea$Geno <- 'Déa'


# Volga Model ----
df.Dev_Stages.Volga <- Leaf_dev_stages(19,1.8,0.0243,39,0.0162,5.5,2*0.0243,4.5)
df.Dev_Stages.Volga$Geno <- 'Volga'

# B73 Model ----
df.Dev_Stages.B73 <- Leaf_dev_stages(16,1.6,0.0193,39,0.0111,5.5,2*0.0243,4.5)
df.Dev_Stages.B73$Geno <- 'B73'


Sim <- rbind(df.Dev_Stages.Dea,df.Dev_Stages.Volga,df.Dev_Stages.B73)
Sim$Geno <- factor(Sim$Geno )
levels(Sim$Geno) <- c('B73','DEA','VOLGA')

Simulated <- melt(Sim,measure.vars=c('nLeafTips','nLeafExp','nLeafFullyExp','nLigules'))
levels(Data$Geno) <- c('B73','DEA','VOLGA')
Measured <- melt(Data,measure.vars=c('Ntip','Nbl','Nll','Nel'))


levels(Measured$variable)
levels(Simulated$variable) <- c("Ntip","Nbl","Nel","Nll")


# Produce the output in pwp : 
library( ReporteRs )
mydoc = pptx(  )
mydoc = addSlide( mydoc, slide.layout = "Title and Content" )

myplot = 
  

ggplot()+
  geom_point(data=Measured,aes(x=TT,y=value,shape=variable,color=variable),size=2.5)+
  geom_line(data=Simulated,aes(x=value,y=Leaf,color=variable))+
  facet_grid(Geno~.)+
  labs(x='thermal time (°Cd)',y='Leaf number')+
  theme_these()+
  scale_y_continuous(limits=c(1,19),breaks=seq(1,19,2))+
  scale_x_continuous(limits=c(0,1000))+
  scale_color_manual(values=c('darkgreen','firebrick3','darkblue','burlywood4'))+
  scale_shape_manual(values=c(21,22,23,24))



mydoc = addPlot( mydoc, function( ) print( myplot ), vector.graphic=T, offx=0.5,offy=0.5,width=4.5, height=8, pointsize=20)  
setwd('D:/Writings/THESIS/Fig')
writeDoc( mydoc, file = "D:/Writings/THESIS/Fig/LERmodel_DeaVolgaB73.pptx" )

colnames(Sim) <- c('Leaf','Nini','Ntip','Nbl','Nel','Nll','Geno')
Data

```

```{r Plot DEA/VOLGA model, fig.height=4.5, fig.width=8, message=FALSE, warning=FALSE, include=FALSE}
  

ToPlotData <- melt(subset(Sim,Geno!='B73'),measure.vars=c('Ntip','Nbl'))
ToPlotSim <- melt(subset(Data,Geno!='B73'),measure.vars=c('Ntip','Nbl'))
  
ggplot()+
  geom_line(data=ToPlotData,aes(x=value,y=Leaf,group=variable))+
  geom_point(data=ToPlotSim,aes(x=TT,y=value,shape=variable),size=2.5)+
  facet_grid(~Geno)+
  labs(x='thermal time (°Cd)',y='Leaf number')+
  theme_these()+
  scale_y_continuous(limits=c(1,19),breaks=seq(1,19,2))+
  scale_x_continuous(limits=c(0,1000))+
  scale_shape_manual(values=c(21,22))
  
  
Haut <- ggplot()+
            geom_line(data=ToPlotData,aes(x=value,y=Leaf,color=variable))+
            geom_point(data=ToPlotSim,aes(x=TT,y=value,shape=variable,color=variable),size=2.5)+
            facet_grid(~Geno)+
            labs(x='thermal time (°Cd)',y='Leaf number')+
            theme_these()+
            scale_y_continuous(limits=c(1,19),breaks=seq(1,19,2))+
            scale_x_continuous(limits=c(0,1000))+
            scale_shape_manual(values=c(21,22))+
            scale_color_manual(values=c('darkgreen','firebrick3'))+
            theme(axis.text.x  = element_blank(),
                    axis.ticks.x = element_blank(),
                    axis.title.x = element_blank())+
  scale_shape_manual(values=c(21,22))



ToPlotData <- melt(subset(Sim,Geno!='B73'),measure.vars=c('Nll','Nel'))
ToPlotSim <- melt(subset(Data,Geno!='B73'),measure.vars=c('Nll','Nel'))
  

Bas <- ggplot()+
            geom_line(data=ToPlotData,aes(x=value,y=Leaf,color=variable))+
            geom_point(data=ToPlotSim,aes(x=TT,y=value,shape=variable,color=variable),size=2.5)+
            facet_grid(~Geno)+
            labs(x='thermal time (°Cd)',y='Leaf number')+
            theme_these()+
            scale_y_continuous(limits=c(1,19),breaks=seq(1,19,2))+
  scale_color_manual(values=c('darkblue','burlywood4'))+
            scale_x_continuous(limits=c(0,1000))+
  scale_shape_manual(values=c(23,24))


# Produce the output in pwp : 
library( ReporteRs )
mydoc = pptx(  )
mydoc = addSlide( mydoc, slide.layout = "Title and Content" )

myplot = 

  
  plot_grid(Haut,Bas,labels=c('',''),nrow=2,rel_heights = c(1,1.1))
  
  
  
mydoc = addPlot( mydoc, function( ) print( myplot ), vector.graphic=T, offx=0.5,offy=0.5,width=6, height=8, pointsize=20)  
setwd('D:/Writings/THESIS/Fig')
writeDoc( mydoc, file = "D:/Writings/THESIS/Fig/LErmodel_DeaVolga.pptx" )



```


- Meteorological data : T / HR / RAD / V  

```{r Treat Met Data, echo=FALSE, fig.height=5, fig.width=6, message=FALSE, warning=FALSE}

Tht <- ThermalTime('Apsim',MetDea$date,MetDea$Tair)

MeteoWithTht <- cbind(MetDea,Tht)
MeteoWithTht$ThtSinceEmerg <- round(MeteoWithTht$cumtht - 124.8,digits=2)
MeteoWithTht$RADrollmean <- rollmean(MeteoWithTht$Rg,k = 7,fill=NA,align="right")

ggplot(data=MeteoWithTht,aes(x=cumtht,y=Rg))+
  geom_line()+
  geom_vline(data=df.Dev_Stages.Dea,aes(xintercept=nLeafExp,color=factor(Leaf)))+
  geom_vline(data=df.Dev_Stages.Dea,aes(xintercept=nLeafFullyExp,color=factor(Leaf)),linetype='dashed' )+
  theme_light()+
  theme(legend.position='none')



```

```{r Calculate real A - corrected by RAD, echo=FALSE, fig.height=10, fig.width=8, message=FALSE, warning=FALSE}

# Get the Dea Data :
ToSimulateDea <- subset(DeaVolga,Geno=='DEA' & Leaf >=3)
ToSimulateDea$TT <- ToSimulateDea$TT-ToSimulateDea$TT[1]
Lag <- 30 # degree.day
Delta <- 50
a <- aggregate(Width~Leaf,ToSimulateDea,FUN=max)
b <-  2.4 # cm / MJ.m-2
density <- 7 # we need it to change this to intercepted per plant !

# Get intercepted coefficient :
LoessModel <- loess(InterceptedRAD,formula = 'Epsilon~cumTht')

xrange <- range(ToSimulateDea$TT)
xseq <- seq(from=xrange[1], to=xrange[2], length=200)
MeteoWithTht$Predicted_Epsilon <- predict(LoessModel,data.frame(cumTht=MeteoWithTht$cumtht))
MeteoWithTht$RADi <- MeteoWithTht$Rg*MeteoWithTht$Predicted_Epsilon
MeteoWithTht$RADi[which(is.na(MeteoWithTht$RADi))] <- 0
MeteoWithTht$RADrollmean <- rollmean(MeteoWithTht$RADi,k = 7,fill=NA,align = 'right')
MeteoWithTht$RADrollmean[which(MeteoWithTht$RADrollmean<0)] <- 0

Df_a <- data.frame(geno=as.character(),leaf=as.numeric(),a=as.numeric(),a_final=as.numeric())
for (leaf in unique(ToSimulateDea$Leaf))
{
  
  ToTreat <- subset(ToSimulateDea,Leaf==leaf)
  TimingElongation <- subset(df.Dev_Stages.Dea,Leaf==leaf)
  Start <- (TimingElongation$nLeafExp)-Delta
  End <- (TimingElongation$nLeafFullyExp-Lag)-Delta
  MetToTreat <- subset(MeteoWithTht,ThtSinceEmerg>Start & 
                         ThtSinceEmerg<End)
  a_leaf <- subset(a,Leaf==leaf)
  RADeffect <- tail(MetToTreat$RADrollmean,n=1)/density*b
  a_real <- a_leaf$Width-RADeffect
  Df_a <- rbind(Df_a,data.frame(geno='DEA',leaf=leaf,a=a_real,a_final=a_leaf$Width))
}


# Get the VOlga Data :
ToSimulateVolga <- subset(DeaVolga,Geno=='VOLGA' & Leaf >=3)
ToSimulateVolga$TT <- ToSimulateVolga$TT-ToSimulateVolga$TT[1]
a <- aggregate(Width~Leaf,ToSimulateVolga,FUN=max)

for (leaf in unique(ToSimulateVolga$Leaf))
{
  
  ToTreat <- subset(ToSimulateVolga,Leaf==leaf)
  TimingElongation <- subset(df.Dev_Stages.Volga,Leaf==leaf)
  Start <- (TimingElongation$nLeafExp)-Delta
  End <- (TimingElongation$nLeafFullyExp-Lag)-Delta
  MetToTreat <- subset(MeteoWithTht,ThtSinceEmerg>Start & 
                         ThtSinceEmerg<End)
  a_leaf <- subset(a,Leaf==leaf)
  RADeffect <- tail(MetToTreat$RADrollmean,n=1)/density*b
  a_real <- a_leaf$Width-RADeffect
  Df_a <- rbind(Df_a,data.frame(geno='VOLGA',leaf=leaf,a=a_real,a_final=a_leaf$Width))
}




```
  
### Calculate W at low RAD :  

We can recalculate the real Wmin by correcting measured width with the effect of intercepted radiation.  
Intercepted radiation is a mean B73_H interception coefficients over the DROPS network.  
```{r Fit a value between genotypes, echo=FALSE, fig.height=4, fig.width=6, message=FALSE, warning=FALSE}

ToPlot <- melt(Df_a,measure.vars=c('a','a_final'))
  
ToTreat <- Df_a
levels(ToTreat$geno) <- c(16,19)
ToTreat$geno <- as.numeric(as.character(ToTreat$geno))
colnames(ToTreat) <- c('Nfinal','leaf','Wmin','Wfinal')


## Fit for both %Nfinal with a beta : ----
nlsfit_DV <- nls( Wmin~  Amax * exp(-0.5 * (((arank*Nfinal) - (leaf))/(brank*Nfinal))^2)  ,
               ToTreat,start=list(arank=0.625,brank=0.375,Amax=9 ))
ToTreat$WidthMinSim  <- predict(nlsfit_DV,ToTreat)


## Fit for both % Nfinal with a bell shape curve : ----
nlsfit_DV2 <- nls( Wmin~  Amax * exp( - b * (leaf - (alpha * Nfinal))^2  + c * (leaf - (alpha * Nfinal) )^3 ),
               ToTreat,start=list(Amax=7,b=0.0005,c=0.0005,alpha=0.6))
ToTreat$WidthMinSim2  <- predict(nlsfit_DV2,ToTreat)


## Plt results :
ggplot()+
  theme_light()+
  geom_point(data=ToTreat,aes(x=leaf,y=Wfinal),color='black',size=3,shape=15)+
  geom_line(data=ToTreat,aes(x=leaf,y=Wfinal),color='black',size=1,linetype='solid')+
  #geom_line(data=ToTreat,aes(x=leaf,y=WidthMinSim),color='red',size=1.5)+
  #geom_line(data=ToTreat,aes(x=leaf,y=WidthMinSim2),color='green',size=1.5)+
  geom_point(data=ToTreat,aes(x=leaf,y=Wmin),color='black',size=3,shape=17)+
  geom_line(data=ToTreat,aes(x=leaf,y=Wmin),color='black',size=1,linetype='dashed')+
  geom_segment(data=ToTreat,aes(x=leaf,y=Wfinal,xend=leaf,yend=Wmin),
               color='blue',arrow = arrow(length=unit(0.30,"cm"),ends="last",type = "closed"))+
  facet_wrap(~Nfinal)+
  labs(x='Leaf rank',y='Width (cm)',size=2.5)+
  theme(axis.text=element_text(size=16),
        axis.title=element_text(size=16))

```

```{r Fit a value between genotypes - Bin, eval=FALSE, fig.height=8, fig.width=10, message=FALSE, warning=FALSE, include=FALSE}

## Fit for both with Dea/Volga % Nfinal : ----
nlsfit <- nls( Wmin~Amax * exp(-0.5 * (((arank*Nfinal) - (leaf))/(brank*Nfinal))^2)  ,
               ToTreat,start=list(arank=0.625,brank=0.375,Amax=9))
ToTreat$Wmin_sim_both <- predict(nlsfit,ToTreat)

# Fit for both with Dea/Volga % Nfinal : -----
nlsfit_NoNfinal <- nls( Wmin~Amax * exp(-0.5 * (((arank) - (leaf))/(brank))^2)  ,
               ToTreat,start=list(arank=10,brank=6,Amax=9))
ToTreat$Wmin_sim_NoNf <- predict(nlsfit_NoNfinal,ToTreat)

## Fit for independently Dea or Volga : ----
nlsfit_dea <- nls( Wmin~Amax * exp(-0.5 * (((arank) - (leaf))/(brank))^2)  ,
               subset(ToTreat,Nfinal==16),start=list(arank=10,brank=5,Amax=9))

nlsfit_volga <- nls( Wmin~Amax * exp(-0.5 * (((arank) - (leaf))/(brank))^2)  ,
               subset(ToTreat,Nfinal==19),start=list(arank=10,brank=6,Amax=9))

ToTreat$Wmin_sim_dea <- predict(nlsfit_dea,ToTreat)
ToTreat$Wmin_sim_volga <- predict(nlsfit_volga,ToTreat)


## Compare all models : 
ToPlot <- melt(ToTreat,measure.vars=c('Wmin','Wmin_sim_both','Wmin_sim_NoNf','Wmin_sim_dea','Wmin_sim_volga'))
ggplot(data=ToPlot,aes(x=leaf,y=value,color=variable,shape=variable))+
  geom_line(size=1)+
  facet_wrap(~as.factor(Nfinal))+
  geom_point(data=ToTreat,aes(x=leaf,y=Wmin),color='black',shape=18,size=3)+
  scale_x_continuous(limits=c(3,19),breaks=seq(3,19,1))+
  scale_y_continuous(limits=c(0,9),breaks=seq(0,9,3))+
  labs(x='Leaf number',y='Width (cm)')+
  theme_bw()

## Only those we are interested in:
ToPlot <- melt(ToTreat,measure.vars=c('Wmin','WidthMinSim'))

ggplot(data=ToPlot,aes(x=leaf,y=value,linetype=variable,shape=variable))+
  geom_line(size=1)+
  geom_point(size=3)+
  facet_wrap(~as.factor(Nfinal))+
  scale_x_continuous(limits=c(3,19),breaks=seq(3,19,1))+
  scale_y_continuous(limits=c(0,9),breaks=seq(0,9,3))+
  labs(x='leaf number',y='Width (cm)')+
  scale_linetype_manual(name='Variable',labels = c('Wmin Est','Wmin Sim'),values=c('solid','dotted'))+
  scale_shape_manual(name='Variable',labels = c('Wmin Est','Wmin Sim'),values=c(18,19),guide=F)+
  theme_bw()


ggplot(data=ToTreat)+
  geom_line(data=ToTreat,aes(x=leaf,y=WidthMinSim),color='red',size=1.5)+
  geom_point(data=ToTreat,aes(x=leaf,y=Wmin),color='blue',size=2)+
  geom_line(data=ToTreat,aes(x=leaf,y=Wfinal),color='black',size=1.5)+
  geom_point(data=ToTreat,aes(x=leaf,y=Wfinal),color='black',size=2)+
  facet_wrap(~as.factor(Nfinal))+
  scale_x_continuous(limits=c(3,19),breaks=seq(3,19,1))+
  scale_y_continuous(limits=c(0,12),breaks=seq(0,12,3))+
  labs(x='leaf number',y='Width (cm)')+
  theme_bw()



```
  
### Show fit on Dea/Volga :  
  
```{r Simulate Wmin, echo=FALSE, fig.height=4, fig.width=7, message=FALSE, warning=FALSE}

# Test Function  ----
DeaWmin <- SimulateWidthAtLowRad(16)
VolgaWmin <- SimulateWidthAtLowRad(19)

# Show results ----
ToMerge <- rbind(DeaWmin,VolgaWmin)

ToTreat <- Df_a
levels(ToTreat$geno) <- c(16,19)
ToTreat$geno <- as.numeric(as.character(ToTreat$geno))
colnames(ToTreat) <- c('Nfinal','Leaf','Wmin','Wfinal')
Merged <- merge(ToTreat,ToMerge)
ToPlot <- melt(Merged,measure.vars=c('Wmin','SimWidth'))

ggplot(data=ToPlot,aes(x=Leaf,y=value,linetype=variable))+
  geom_line(size=1.2)+
  facet_wrap(~Nfinal)+
  theme_light()+
  scale_y_continuous(limits=c(0,10),breaks=seq(0,10,1))+
  scale_x_continuous(limits=c(0,30),breaks=seq(0,30,3))+
  labs(x='Leaf rank',y='Width (cm)')+
  scale_linetype_discrete(name='Width',labels=c('Estimated','Simulated'))

```
  
### Simulated width Growth :  
  
Width simulated  
  
Delta - thermal time difference of start and end of width expansion to length elongation  
Lag - difference between ligulation and end of leaf elongation (dd/leaf)  
Wmin parameters defined + B parameter for sensitivity of W to intercepted radiation  

<span style="color:red">*Déa*</span>  

```{r Simulate leaf width - Dea, echo=FALSE, fig.height=6, fig.width=8, message=FALSE, warning=FALSE}

# Get the Dea Data : ----
ToSimulateDea <- subset(DeaVolga,Geno=='DEA' & Leaf >=3)
ToSimulateDea$TT <- ToSimulateDea$TT-ToSimulateDea$TT[1]
Lag <- 30 # degree.day
Delta <- 50
a <- DeaWmin
b <-0.3 # cm / MJ.m-2

# Get interception coefficient : ----
LoessModel <- loess(InterceptedRAD,formula = 'Epsilon~cumTht')

xrange <- range(ToSimulateDea$TT)
xseq <- seq(from=xrange[1], to=xrange[2], length=200)
MeteoWithTht$Predicted_Epsilon <- predict(LoessModel,data.frame(cumTht=MeteoWithTht$cumtht))
MeteoWithTht$RADi <- MeteoWithTht$Rg*MeteoWithTht$Predicted_Epsilon
MeteoWithTht$RADi[which(is.na(MeteoWithTht$RADi))] <- 0
MeteoWithTht$RADrollmean <- rollmean(MeteoWithTht$RADi,k = 7,fill=NA,align = 'right')
MeteoWithTht$RADrollmean[which(MeteoWithTht$RADrollmean<0)] <- 0


# Begin simulation loop : ----
SimulationOutputsDea <- data.frame(Date=as.character(),
                                TT=as.numeric(),
                                Leaf=as.numeric,
                                Sim_Width= as.numeric(),
                                LeafFormFactor=as.numeric(),
                                LeafWidthRatio=as.numeric(),
                                Wmax=as.numeric(),
                                RADeffect=as.numeric(),
                                Apot=as.numeric(),
                                Geno=as.character())
  
# Sim each leaf :
for (leaf in unique(ToSimulateDea$Leaf))
{
  
  ToTreat <- subset(ToSimulateDea,Leaf==leaf)
  TimingElongation <- subset(df.Dev_Stages.Dea,Leaf==leaf)
  Start <- (TimingElongation$nLeafExp)-Delta
  End <- (TimingElongation$nLeafFullyExp-Lag)-Delta
  a_leaf <- subset(a,Leaf==leaf)
  MetToTreat <- subset(MeteoWithTht,ThtSinceEmerg>Start & 
                         ThtSinceEmerg<End)
  
  # Sim each day : 
  for (day in 1:nrow(MeteoWithTht))
  {
    
    if (MeteoWithTht$ThtSinceEmerg[day]>Start & MeteoWithTht$ThtSinceEmerg[day]<End)
    {
    LeafWidthRatio <- (MeteoWithTht$ThtSinceEmerg[day]-Start)/(End-Start)
    LeafFormFactor <- 0.5+ (LeafWidthRatio * 0.25)
    LeafWidth  <- a_leaf$SimWidth+(MeteoWithTht$RADrollmean[day]*b)
    SimulationOutputsDea <- rbind(SimulationOutputsDea,data.frame(Date=MeteoWithTht$date[day],
                                                            TT=MeteoWithTht$ThtSinceEmerg[day],
                                                            Leaf=leaf,Sim_Width= LeafWidthRatio*LeafWidth,
                                                            LeafFormFactor=LeafFormFactor,
                                                            LeafWidthRatio=LeafWidthRatio,
                                                            Wmax=LeafWidth,
                                                            RADeffect=MeteoWithTht$RADrollmean[day]*b,
                                                            Apot=a_leaf$SimWidth,
                                                            Geno='DEA'))
    } else if (MeteoWithTht$ThtSinceEmerg[day]<=Start) {
      
    SimulationOutputsDea <- rbind(SimulationOutputsDea,data.frame(Date=MeteoWithTht$date[day],
                                                            TT=MeteoWithTht$ThtSinceEmerg[day],
                                                            Leaf=leaf,
                                                            Sim_Width= 0,
                                                            LeafFormFactor=0,
                                                            LeafWidthRatio=0,
                                                            Wmax=0,
                                                            RADeffect=0,
                                                            Apot=a_leaf$SimWidth,
                                                            Geno='DEA'))  
    
    } else if (MeteoWithTht$ThtSinceEmerg[day]>=End) {
      
    SimulationOutputsDea <- rbind(SimulationOutputsDea,data.frame(Date=MeteoWithTht$date[day],
                                                            TT=MeteoWithTht$ThtSinceEmerg[day],
                                                            Leaf=leaf,Sim_Width= LeafWidthRatio*LeafWidth,
                                                            LeafFormFactor=LeafFormFactor,
                                                            LeafWidthRatio=LeafWidthRatio,
                                                            Wmax=LeafWidth,
                                                            RADeffect=MeteoWithTht$RADrollmean[day-1]*b,
                                                            Apot=a_leaf$SimWidth,
                                                            Geno='DEA'))
    
    } 
    
  }
}

ggplot()+
  geom_line(data=SimulationOutputsDea,aes(x=TT,y=Sim_Width,color=factor(Leaf)))+
  geom_point(data=ToSimulateDea,aes(x=TT,y=Width,color=factor(Leaf)),shape=19,size=2.5)+
  facet_wrap(~Leaf,nrow = 4,ncol = 4)+
  theme_light()+
  theme(legend.position='none')+
  labs(x='Thermal Time (APSIM dd)',y='Width (cm)')
  
```
  
<span style="color:red">*Volga*</span>  
   
```{r Simulate leaf width - Volga, echo=FALSE, fig.height=6, fig.width=8, message=FALSE, warning=FALSE}


# Get the Volga Data : ----
ToSimulateVolga <- subset(DeaVolga,Geno=='VOLGA' & Leaf >=3)
ToSimulateVolga$TT <- ToSimulateVolga$TT-ToSimulateVolga$TT[1]
Lag <- 30 # degree.day
Delta <- 50
a <- VolgaWmin
b <-0.47 # cm / MJ.m-2

# Get interception coefficient : ----
LoessModel <- loess(InterceptedRAD,formula = 'Epsilon~cumTht')

xrange <- range(ToSimulateVolga$TT)
xseq <- seq(from=xrange[1], to=xrange[2], length=200)
MeteoWithTht$Predicted_Epsilon <- predict(LoessModel,data.frame(cumTht=MeteoWithTht$cumtht))
MeteoWithTht$RADi <- MeteoWithTht$Rg*MeteoWithTht$Predicted_Epsilon
MeteoWithTht$RADi[which(is.na(MeteoWithTht$RADi))] <- 0
MeteoWithTht$RADrollmean <- rollmean(MeteoWithTht$RADi,k = 7,fill=NA,align = 'right')
MeteoWithTht$RADrollmean[which(MeteoWithTht$RADrollmean<0)] <- 0


# Begin simulation loop : ----
SimulationOutputsVolga <- data.frame(Date=as.character(),
                                TT=as.numeric(),
                                Leaf=as.numeric,
                                Sim_Width= as.numeric(),
                                LeafFormFactor=as.numeric(),
                                LeafWidthRatio=as.numeric(),
                                Wmax=as.numeric(),
                                RADeffect=as.numeric(),
                                Apot=as.numeric(),
                                Geno=as.character())

# Sim each leaf :
for (leaf in unique(ToSimulateVolga$Leaf))
{
  
  ToTreat <- subset(ToSimulateVolga,Leaf==leaf)
  TimingElongation <- subset(df.Dev_Stages.Volga,Leaf==leaf)
  Start <- (TimingElongation$nLeafExp)-Delta
  End <- (TimingElongation$nLeafFullyExp-Lag)-Delta
  a_leaf <- subset(a,Leaf==leaf)
  MetToTreat <- subset(MeteoWithTht,ThtSinceEmerg>Start & 
                         ThtSinceEmerg<End)
  
  # Sim each day : 
  for (day in 1:nrow(MeteoWithTht))
  {
    
    if (MeteoWithTht$ThtSinceEmerg[day]>Start & MeteoWithTht$ThtSinceEmerg[day]<End)
    {
    LeafWidthRatio <- (MeteoWithTht$ThtSinceEmerg[day]-Start)/(End-Start)
    LeafFormFactor <- 0.5+ (LeafWidthRatio * 0.25)
    LeafWidth  <- a_leaf$SimWidth+(MeteoWithTht$RADrollmean[day]*b)
    SimulationOutputsVolga <- rbind(SimulationOutputsVolga,data.frame(Date=MeteoWithTht$date[day],
                                                            TT=MeteoWithTht$ThtSinceEmerg[day],
                                                            Leaf=leaf,Sim_Width= LeafWidthRatio*LeafWidth,
                                                            LeafFormFactor=LeafFormFactor,
                                                            LeafWidthRatio=LeafWidthRatio,
                                                            Wmax=LeafWidth,
                                                            RADeffect=MeteoWithTht$RADrollmean[day]*b,
                                                            Apot=a_leaf$SimWidth,
                                                            Geno='VOLGA'))
    } else if (MeteoWithTht$ThtSinceEmerg[day]<=Start) {
      
    SimulationOutputsVolga <- rbind(SimulationOutputsVolga,data.frame(Date=MeteoWithTht$date[day],
                                                            TT=MeteoWithTht$ThtSinceEmerg[day],
                                                            Leaf=leaf,
                                                            Sim_Width= 0,
                                                            LeafFormFactor=0,
                                                            LeafWidthRatio=0,
                                                            Wmax=0,
                                                            RADeffect=0,
                                                            Apot=a_leaf$SimWidth,
                                                            Geno='VOLGA'))  
    
    } else if (MeteoWithTht$ThtSinceEmerg[day]>=End) {
      
    SimulationOutputsVolga <- rbind(SimulationOutputsVolga,data.frame(Date=MeteoWithTht$date[day],
                                                            TT=MeteoWithTht$ThtSinceEmerg[day],
                                                            Leaf=leaf,Sim_Width= LeafWidthRatio*LeafWidth,
                                                            LeafFormFactor=LeafFormFactor,
                                                            LeafWidthRatio=LeafWidthRatio,
                                                            Wmax=LeafWidth,
                                                            RADeffect=MeteoWithTht$RADrollmean[day-1]*b,
                                                            Apot=a_leaf$SimWidth,
                                                            Geno='VOLGA'))
    
    } 
    
  }
}


ggplot()+
  geom_line(data=SimulationOutputsVolga,aes(x=TT,y=Sim_Width,color=factor(Leaf)))+
  geom_point(data=ToSimulateVolga,aes(x=TT,y=Width,color=factor(Leaf)),shape=19,size=2.5)+
  facet_wrap(~Leaf,nrow = 5,ncol = 4)+
  theme_light()+
  theme(legend.position='none')+
  labs(x='Thermal Time (APSIM dd)',y='Width (cm)')


  
```
  
<span style="color:red">*Final width*</span>  
  
```{r Show Dea/Volga Wmax, echo=FALSE, fig.height=4, fig.width=5, message=FALSE, warning=FALSE}

SimulationOutputs <- rbind(SimulationOutputsDea,SimulationOutputsVolga)

WMaxSim <- aggregate(Sim_Width~Leaf+Geno,SimulationOutputs,max,na.rm=T)
WMaxObs <- aggregate(Width~Leaf+Geno,DeaVolga,max,na.rm=T)
Merged <- merge(WMaxSim,WMaxObs)

ggplot(data=Merged,aes(x=Sim_Width,y=Width,color=Geno))+
  geom_point(size=3)+
  geom_smooth(method='lm',color='black',se=F,size=1.5)+
  geom_abline(slope=1,intercept=0)+
  scale_x_continuous(limits=c(0,10))+
  scale_y_continuous(limits=c(0,10))+
  labs(x='Simulated Width (cm)',y='Observed Width (cm)')+
  theme_light()
  
```

## Simulate DROPS network

```{r Load parameters, echo=FALSE, fig.height=8, fig.width=9, message=FALSE, warning=FALSE}

Factor <- 3/5
Parameters <- read.table('D:/Work/Sims - DROPS/4.Out/Parameters_DROPS_Panel_ForAPSIM.csv', 
                          sep=",", header=TRUE,dec=".",row.names=NULL,as.is=FALSE, fill=TRUE)

# Get which genotypes we want to simulate : 
Data_MAUGUIO_2016 <- read.table('D:/Work/Widths/3.Data/MAUGUIO2016/Mauguio_Measurements.csv', 
                             sep=";",                                           
                             header=TRUE,dec=".",
                             row.names=NULL,
                             as.is=FALSE, 
                             fill=TRUE)

Genotypes <- unique(Data_MAUGUIO_2016$Genotype)
GenoToSimulate <-  Parameters[Parameters$Hybrid_ID %in% Genotypes,]
MyParams <- subset(GenoToSimulate,select=colnames(GenoToSimulate)[c(2:11,16,17)])

#MyParams$a <- 3.25
#MyParams[2,]$b <- -0.85
#MyParams[2,]$c <- 3.9

## B73_H :
ParamsB73 <- subset(MyParams,Hybrid_ID=='B73_H')


chart.Correlation(Parameters[c(3:11,13,14,16)], histogram=TRUE, pch=21)

```

```{r Treat GWAS Outputs, eval=FALSE, message=FALSE, warning=FALSE, include=FALSE}


Out_GWAS_AllData <- read.table('U:/Alix/Seb/QTLs.csv',
                            sep=";",                                           
                            header=TRUE,dec=".",
                            row.names=NULL)

#Out_GWAS_AllData$Trait.Environment <- paste(Out_GWAS_AllData$trait,Out_GWAS_AllData$Environment,sep='.')

## How much QTLs for each variable : 
QTLs_Numbers <- data.frame(Variable=NULL,QTLs=NULL,Ind_Nb=NULL)
for (var in unique(Out_GWAS_AllData$trait))
{
  ToTreat <- subset(Out_GWAS_AllData,trait==var)
  if (nrow(ToTreat)!=0)
  {
    ind_nb <- mean(ToTreat$Ind_Nb,na.rm=T)
    qtls <- nrow(ToTreat)
  }
  else
  {
    ind_nb <- NA
    qtls <- NA
  }
  ToBind <- data.frame(Variable=var,QTLs=qtls,Ind_Nb=ind_nb)
  QTLs_Numbers <- rbind(QTLs_Numbers,ToBind)
}


write.table(QTLs_Numbers,
            'D:/Work/Widths/4.Out/QTLs/QTLs_Numbers_SubsetVars.csv',
            sep=";",
            row.names=F)

## 

Variables <- unique(QTLs_Numbers$Variable)
MyResults <- data.frame(NULL)
MyResults <- as.data.frame(matrix(ncol= length(Variables), nrow=length(Variables)))
colnames(MyResults) <- Variables
rownames(MyResults) <- Variables

Colocs_df <- data.frame(Trait_1=NULL,
                        Trait_2=NULL,
                        Nqtls=NULL,
                        Chroms=NULL,
                        Bins=NULL,
                        SNP_Weights=NULL,
                        Sizes=NULL
                        )

for (var_col in colnames(MyResults))
{
  for (var_row in rownames(MyResults))
  
  {
    
    ToTreat <- subset(Out_GWAS_AllData,trait==var_row | trait==var_col)

    ToTreat <- ToTreat[order(ToTreat$Chromosome, ToTreat$pos.genet.1vrai),]
    
    count <- 0
    Chr <- paste('')
    Bin <- paste('')
    SNP_Weight <- paste('')
    Size <- paste('')
    while (nrow(ToTreat)>=2)
    {
      ToTest <-  ToTreat[1,]
      ToTreat <- ToTreat[-1,]
      Colocs <- ToTreat[which(ToTest$Chromosome==ToTreat$Chromosome 
                              & ToTreat$pos.genet.1vrai < ToTest$pos.genet.2vrai),]
      count <- count + nrow(Colocs)
      if (nrow(Colocs)>=1)
      {
        Chr <- paste(Chr,ToTest$Chromosome,Colocs$Chromosome[1],sep=' ')
        Bin <- paste(Bin,ToTest$Bin,Colocs$Bin[1],sep=' ')
        SNP_Weight <- paste(SNP_Weight,ToTest$SNP_Weight,Colocs$SNP_Weight[1],sep=' ')
        Size <- paste(SNP_Weight,ToTest$pos.genet.1vrai,Colocs$pos.genet.2vrai[1],sep=' ')
      }
    }
    
    MyResults[var_col,var_row]<-count
    
    if (count >0)
    {
      if (length(which(Colocs_df$Trait_1==var_row & Colocs_df$Trait_2==var_col ))==0)
      {
        ToBind <-  data.frame(Trait_1=var_col,
                              Trait_2=var_row,
                              Nqtls=count,
                              Chroms=Chr,
                              Bins=Bin,
                              SNP_Weights=SNP_Weight,
                              Sizes=Size)
        Colocs_df <- rbind(Colocs_df,ToBind)
      }
    }
  }
  
}


write.table(MyResults,
            'D:/Work/Widths/4.Out/QTLs/ColocsMatrix_SubsetVars.csv',
            sep=";",
            row.names=T)

write.table(Colocs_df,
            'D:/Work/Widths/4.Out/QTLs/ColocsList_SubsetVars.csv',
            sep=";")

var_col <- colnames(MyResults)[1]
var_row <- colnames(MyResults)[5]

##

Colocs_PerQTLs
ToTreat <- 
for (qtls in nrow(Out_GWAS_AllData))
{
  ToTest <- Out_GWAS_AllData
}

```

```{r Prepare parameters, eval=FALSE, message=FALSE, warning=FALSE, include=FALSE}


# Loading XML file : ----
ApsimFile<-paste(PATHS[1],'DROPS_Network_Calibration',sep='/')
ApsimFile<-paste(ApsimFile,'apsim',sep='.')
SimFileTree<-xmlTreeParse(ApsimFile,getDTD = F , useInternalNodes = T)

# Prepare data : ----
SimToUse <- SimFileTree

# Test variables: ----
#geno <- 'B104_H'
#var <- colnames(MyParams)[2]
Stopped_At <- 1
WorkOnErrors <- F

# define name / min / max / nominal value :
name <- c("a","c")
min<- c(2.25,1.5)
max<- c(6,4.5)
nominal<- c(4.25,3.53)

# Get everything in a dataframe :
parameters <- data.frame(name, min, max, nominal)
rm(name,min,max,nominal)

# On cr?e une liste avec les param?tres :
q.arg <- list()
for(i in 1:nrow(parameters)) 
{ 
  q.arg[[length(q.arg)+1]] <- list(min= parameters[i,2], max= parameters[i,3])
}

# -----------------------------------------------------------------------------
## Part 2 ) Tell sensitivity to work on the parameters and give us an experimental framework :

# On peut ensuite appeller l'algo qui nous calcule les sets de param?tres :
sensitivity <- fast99(
  model=NULL,
  factors=as.character(parameters[,1]), 
  n=70, 
  q=rep("qunif",nrow(parameters)),
  q.arg=q.arg)

Params<-sensitivity$X

a <- seq(3.5,3.5,0.04)
c <- seq(3.5,3.5,0.04)

Params <- expand.grid(a,c)
colnames(Params) <- c('a','c')

```

```{r Multi-parameters simulation, eval=FALSE, message=FALSE, warning=FALSE, include=FALSE}

# 1/ Loop over genotypes : ----
Start <- Sys.time()
for ( i in 1:nrow(Params))
{
  ParamsToUse <- Params[i,]
  ParamsToUse$b <- (-0.0758*ParamsToUse$c)-0.3672
  
  print(paste(c('----- Sim n°',i,'out of',nrow(Params),' -----'),collapse=" "))
  
  # 2/ Change Simname and outputs
  print(paste(c('Step 1 : Initialize simulation...'),collapse=" "))
  SimulationName<-paste(i)
  NewApsimName<-paste(c(SimulationName,"apsim"),collapse=".")
  
  # Loop over genotypic parameters and change them :
    for (var in colnames(ParamsB73[2:ncol(ParamsToUse)]))
    {
       VarNode<-getNodeSet(SimToUse, paste(c("//",var),collapse="") )
       NewValue<- ParamsB73[,var]
       
      for (node in 1:length(VarNode))
      {
         ChangeXmlValue(VarNode[node],NewValue) 
      }
       
    }
  
  # Loop over genotypic parameters and change them :
    for (var in colnames(ParamsToUse))
    {
       VarNode<-getNodeSet(SimToUse, paste(c("//",var),collapse="") )
       NewValue<- ParamsToUse[,var]
       
      for (node in 1:length(VarNode))
      {
         ChangeXmlValue(VarNode[node],NewValue) 
      }
       
    }

    print(paste('==> Done !'),collapse=" ")
  
    # Save the new apsimfile in Model Directory :
    print(paste(c('Step 2 : Saving simulation file...'),collapse=" "))
    saveXML(xmlRoot(SimFileTree), 
            file= paste0(PATH.APSIM,NewApsimName), 
            prefix= '<?xml version="1.0" encoding="utf-8"?>',indent= T)
    print(paste('==> Done !'),collapse=" ")
  
    # Simulate :
    print(paste(c('Step 3 : Simulation...'),collapse=" "))
    setwd(PATH.APSIM)
    
    command<-paste("apsim",NewApsimName)

    if (WorkOnErrors==T)
    {
      result <- tryCatch(evalWithTimeout(system(command,show.output.on.console=T),
                                 timeout = 30,
                                 onTimeout = "error"),
                 error=
                   {
                        #Get all tasks running :
                        Tasks <- system('tasklist',show.output.on.console=T,intern=T)
                        # Kill jobrunner :
                        while (sum(grepl("JobRunner.exe", Tasks))>0)
                        {  
                          print('Killing JobRunner...')
                          system('taskkill /IM JobRunner.exe /F',
                                 show.output.on.console=F,intern=T)
                          Tasks <- system('tasklist',show.output.on.console=T,intern=T)
                        }
                        # Kill all Apsim.exe :
                        while (sum(grepl("Apsim.exe", Tasks))>0)
                        {
                          print('Killing APSIM...')
                          system('taskkill /IM Apsim.exe /F',
                                 show.output.on.console=F,intern=T)       
                          Tasks <- system('tasklist',show.output.on.console=T,intern=T)
                        }
                    }
                 )
      
      # If it runs smoothly :
      if (result==0) {print(paste('==> Done !'),collapse=" ")
      # If there was an error : 
      } else {print(paste('==> Error Handled !'),collapse=" ")}
    }
    else {system(command)}
      
    
    # Get/Save outputs  : 
    print(paste(c('Step 4 : Loading outputs...'),collapse=" "))
    
    Output_Sim <- loadApsim(PATH.APSIM, 
                            loadAll = TRUE, 
                            ext = ".out", 
                            returnFrame = TRUE, 
                            n = 0, fill = TRUE, 
                            addConstants = TRUE)
    
    Output_Sim$Genotype<-geno
    Output_Sim$a <- ParamsToUse$a
    Output_Sim$c <- ParamsToUse$c
  
    if (i ==1)
    {Global_Output <- Output_Sim
    }else{
      Global_Output <- rbind(Global_Output,Output_Sim)
    }
    rm(Output_Sim)
  
  print(paste('==> Done !'),collapse=" ")  
  
}

Stop <- Sys.time()
TimeDiff <- difftime(Stop, Start, units="secs")
MyTimeDiff <- format(.POSIXct(TimeDiff,tz="GMT"), "%H:%M:%S")
print(paste(c(''),collapse=" "))
print(paste(c('----- Simulation loops finished -----'),collapse=" "))
print(paste(c(''),collapse=" "))
print(paste(c('Time of execution is '),MyTimeDiff,collapse=" "))


```

```{r Treat data sensitivity analysis, eval=FALSE, message=FALSE, warning=FALSE, include=FALSE}

## Get Harvest Data :
Global_Output$a <- as.character(Global_Output$a)
Global_Output$c <- as.character(Global_Output$c)
HarvestDataCali <- Get_Harvest_Data(Global_Output,list(Global_Output$a,Global_Output$c,Global_Output$Title))
HarvestData <- HarvestDataCali[!duplicated(HarvestDataCali[,c(64,71:72)]),]

HarvestData_Melted <-  melt(HarvestData, measure.vars = c('leafwidth(1)','leafwidth(2)','leafwidth(3)','leafwidth(4)','leafwidth(5)',
                                  'leafwidth(6)','leafwidth(7)','leafwidth(8)','leafwidth(9)','leafwidth(10)',
                                  'leafwidth(11)','leafwidth(12)','leafwidth(13)','leafwidth(14)','leafwidth(15)','leafwidth(16)'))

Width <- subset(HarvestData_Melted,select=c("Title",'Genotype','DaysAfterSowing',
                                             'yield','GrainNo',"variable","value","a","c"))

colnames(Width) <- c('Site_Year','Genotype','Das','Yield','GrainNo','Leaf','Width',"a","c")
levels(Width$Leaf) <- seq(1,16,1)
Width$Leaf <- as.numeric(Width$Leaf)

HarvestData_Melted <-  melt(HarvestData, measure.vars = c('leaflength(1)','leaflength(2)','leaflength(3)','leaflength(4)','leaflength(5)',
                                  'leaflength(6)','leaflength(7)','leaflength(8)','leaflength(9)',
                                  'leaflength(10)','leaflength(11)','leaflength(12)','leaflength(13)',
                                  'leaflength(14)','leaflength(15)','leafwidth(16)'))

Length <- subset(HarvestData_Melted,select=c("Title",'Genotype','DaysAfterSowing',
                                             'yield','GrainNo',"variable","value","a","c"))

colnames(Length) <- c('Site_Year','Genotype','Das','Yield','GrainNo','Leaf','Length',"a","c")
levels(Length$Leaf) <- seq(1,16,1)
Length$Leaf <- as.numeric(Length$Leaf)

Merged_Outputs <- merge(Width,Length)

#Aggregate leaf variables : 
MeanLeafVars <- aggregate(data=Merged_Outputs,cbind(Width,Length)~Genotype+Leaf,FUN=mean,na.rm=T)


ggplot(data=Width,aes(x=a,y=c,color=Width,fill=Width))+
  geom_tile()+
  facet_wrap(~Leaf)+
  scale_color_continuous(low='red',high='green')+
  scale_fill_continuous(low='red',high='green')


ggplot(data=subset(Length,Leaf<15 & Leaf > 6),aes(x=a,y=c,color=Length,fill=Length))+
  geom_tile()+
  facet_wrap(~Leaf)+
  scale_color_continuous(low='red',high='green')+
  scale_fill_continuous(low='red',high='green')

#### Measurezd variables : 

## DROPS network : 
PATH.data_B73_DROPS <- "D:/Work/Widths/3.Data/DROPS_Network_Measurements_df.csv"

Data_Field <- read.table(PATH.data_B73_DROPS, 
                       sep=";",                                           
                       header=TRUE,dec=".",
                       row.names=NULL,
                       as.is=FALSE, 
                       fill=TRUE)

FinalLeafWidth_DROPSFieldB73_H <- subset(Data_Field,select = c('Site_Year','Width','Length','Leaf'))
FinalLeafWidth_DROPSFieldB73_H$Treatment  <- 'Well-watered'
FinalLeafWidth_DROPSFieldB73_H$Title <- paste0(" ", FinalLeafWidth_DROPSFieldB73_H$Site_Year)

# Length :
B73_Sim <- subset(Length,select=c('Site_Year','Leaf','Length','a','c'))
B73_Sim$Site_Year <- factor(B73_Sim$Site_Year)
B73_Obs <- subset(FinalLeafWidth_DROPSFieldB73_H,select=c('Title','Leaf','Length'))
colnames(B73_Obs) <- c('Site_Year','Leaf','Length')
B73_Obs$Site_Year <- factor(B73_Obs$Site_Year)
  
Length_Global <- merge(B73_Sim,B73_Obs,by=c('Leaf','Site_Year'))
colnames(Length_Global) <- c('Leaf','Site_Year','Simulated','a','c','Measured')

# Calculate RMSE : 
Length_Global$Delta <- (Length_Global$Measured-Length_Global$Simulated)^2
ResultsCali <- aggregate(data=Length_Global,Delta~a+c,sum,na.rm=T)

ggplot(data=ResultsCali,aes(x=a,y=c,color=sqrt(Delta),fill=sqrt(Delta)))+
  geom_tile()+
  scale_color_continuous(low='red',high='green')+
  scale_fill_continuous(low='red',high='green')

MeanParams <- ResultsCali[which(ResultsCali$Delta==min(ResultsCali$Delta)),]


```

```{r Calibrate genotypes, message=FALSE, warning=FALSE, include=FALSE}

MyParams <- subset(GenoToSimulate,select=colnames(GenoToSimulate)[c(2:11,15,16)])

Scale_a <- 3.5/ParamsB73$a
Scale_c <- 3.5/ParamsB73$c

MyParams$a <- MyParams$a*Scale_a
MyParams$c <- MyParams$c*Scale_c
MyParams$b <- (-0.0758*MyParams$c)-0.3672

```

```{r Simulation multi-genotype, eval=FALSE, message=FALSE, warning=FALSE, include=FALSE}

# Loading XML file : ----
ApsimFile<-paste(PATHS[1],'DROPS_Network_Calibration',sep='/')
ApsimFile<-paste(ApsimFile,'apsim',sep='.')
SimFileTree<-xmlTreeParse(ApsimFile,getDTD = F , useInternalNodes = T)

# Prepare data : ----
SimToUse <- SimFileTree

# Test variables: ----
#geno <- 'B104_H'
#var <- colnames(MyParams)[2]
Stopped_At <- 7
WorkOnErrors <- F

# 1/ Loop over genotypes : ----
Start <- Sys.time()
for (i in Stopped_At:length(unique(MyParams$Hybrid_ID)))
#for (i in 2)
{
  Genotypes <- unique(MyParams$Hybrid_ID)
  geno <- as.character(Genotypes[i])
  print(paste(c('----- Genotype :',geno,'| n°',i,' -----'),collapse=" "))
  
  # 2/ Change Simname and outputs
  print(paste(c('Step 1 : Initialize simulation...'),collapse=" "))
  ParamsToUse <- subset(MyParams,Hybrid_ID==as.character(geno))
  SimulationName<-paste(ParamsToUse[,'Hybrid_ID'])
  NewApsimName<-paste(c(SimulationName,"apsim"),collapse=".")
  
  # CHeck if we have values for the genotype :
  if (nrow(ParamsToUse)>0)
  {
    # Loop over genotypic parameters and change them :
    for (var in colnames(ParamsToUse[2:ncol(ParamsToUse)]))
    {
       VarNode<-getNodeSet(SimToUse, paste(c("//",var),collapse="") )
       NewValue<- ParamsToUse[,var]
       
      for (node in 1:length(VarNode))
      {
         ChangeXmlValue(VarNode[node],NewValue) 
      }
       
    }
    print(paste('==> Done !'),collapse=" ")
  
    # Save the new apsimfile in Model Directory :
    print(paste(c('Step 2 : Saving simulation file...'),collapse=" "))
    saveXML(xmlRoot(SimFileTree), 
            file= paste0(PATH.APSIM,NewApsimName), 
            prefix= '<?xml version="1.0" encoding="utf-8"?>',indent= T)
    print(paste('==> Done !'),collapse=" ")
  
    # Simulate :
    print(paste(c('Step 3 : Simulation...'),collapse=" "))
    setwd(PATH.APSIM)
    
    command<-paste("apsim",NewApsimName)

    if (WorkOnErrors==T)
    {
      result <- tryCatch(evalWithTimeout(system(command,show.output.on.console=T),
                                 timeout = 30,
                                 onTimeout = "error"),
                 error=
                   {
                        #Get all tasks running :
                        Tasks <- system('tasklist',show.output.on.console=T,intern=T)
                        # Kill jobrunner :
                        while (sum(grepl("JobRunner.exe", Tasks))>0)
                        {  
                          print('Killing JobRunner...')
                          system('taskkill /IM JobRunner.exe /F',
                                 show.output.on.console=F,intern=T)
                          Tasks <- system('tasklist',show.output.on.console=T,intern=T)
                        }
                        # Kill all Apsim.exe :
                        while (sum(grepl("Apsim.exe", Tasks))>0)
                        {
                          print('Killing APSIM...')
                          system('taskkill /IM Apsim.exe /F',
                                 show.output.on.console=F,intern=T)       
                          Tasks <- system('tasklist',show.output.on.console=T,intern=T)
                        }
                    }
                 )
      
      # If it runs smoothly :
      if (result==0) {print(paste('==> Done !'),collapse=" ")
      # If there was an error : 
      } else {print(paste('==> Error Handled !'),collapse=" ")}
    }
    else {system(command)}
      
    
    # Get/Save outputs  : 
    print(paste(c('Step 4 : Loading outputs...'),collapse=" "))
    
    Output_Sim <- loadApsim(PATH.APSIM, 
                            loadAll = TRUE, 
                            ext = ".out", 
                            returnFrame = TRUE, 
                            n = 0, fill = TRUE, 
                            addConstants = TRUE)
    
    Output_Sim$Genotype<-geno
  
    if (i ==1)
    {Global_Output <- Output_Sim
    }else{
      Global_Output <- rbind(Global_Output,Output_Sim)
    }
    rm(Output_Sim)
  }
  
  print(paste('==> Done !'),collapse=" ")  
  
}

Stop <- Sys.time()
TimeDiff <- difftime(Stop, Start, units="secs")
MyTimeDiff <- format(.POSIXct(TimeDiff,tz="GMT"), "%H:%M:%S")
print(paste(c(''),collapse=" "))
print(paste(c('----- Simulation loops finished -----'),collapse=" "))
print(paste(c(''),collapse=" "))
print(paste(c('Time of execution is '),MyTimeDiff,collapse=" "))


# 2/ Save and treat outputs : ----
save(file=paste0(PATHS[4],'/Global_Output.RData'),Global_Output)
  
Mau16_ToTreat <- subset(Global_Output,Title==" Mauguio_2016")

which(!is.na(Global_Output$RADi))

```

```{r Treat MultiSim Outputs, fig.height=8, fig.width=12, message=FALSE, warning=FALSE, include=FALSE}


# 3/ Estimate Harvest Yield and stuff : ---
load(paste0(PATHS[4],'/Global_Output.RData'))
HarvestDataOutputs <- Get_Harvest_Data(Global_Output,list(Global_Output$Title,Global_Output$Genotype))
HarvestData <- HarvestDataOutputs[!duplicated(HarvestDataOutputs[,c(64,70)]),]

HarvestData_Melted <-  melt(HarvestData, measure.vars = c('leafwidth(1)','leafwidth(2)','leafwidth(3)','leafwidth(4)','leafwidth(5)',
                                  'leafwidth(6)','leafwidth(7)','leafwidth(8)','leafwidth(9)','leafwidth(10)',
                                  'leafwidth(11)','leafwidth(12)','leafwidth(13)','leafwidth(14)','leafwidth(15)','leafwidth(16)'))

Width <- subset(HarvestData_Melted,select=c("Title",'Genotype','DaysAfterSowing',
                                             'yield','GrainNo',"variable","value"))

colnames(Width) <- c('Site_Year','Genotype','Das','Yield','GrainNo','Leaf','Width')
levels(Width$Leaf) <- seq(1,16,1)
HarvestData_Melted <-  melt(HarvestData, measure.vars = c('leaflength(1)','leaflength(2)','leaflength(3)','leaflength(4)','leaflength(5)',
                                  'leaflength(6)','leaflength(7)','leaflength(8)','leaflength(9)',
                                  'leaflength(10)','leaflength(11)','leaflength(12)','leaflength(13)',
                                  'leaflength(14)','leaflength(15)','leafwidth(16)'))

Length <- subset(HarvestData_Melted,select=c("Title",'Genotype','DaysAfterSowing',
                                             'yield','GrainNo',"variable","value"))

colnames(Length) <- c('Site_Year','Genotype','Das','Yield','GrainNo','Leaf','Length')
levels(Length$Leaf) <- seq(1,16,1)

Merged_Outputs <- merge(Width,Length)

#Aggregate leaf variables : 
MeanLeafVars <- aggregate(data=Merged_Outputs,cbind(Width,Length)~Genotype+Leaf+Site_Year,FUN=mean,na.rm=T)
  
```

### Network for B73_H 

```{r Load Data, echo=FALSE, fig.height=8, fig.width=8, message=FALSE, warning=FALSE}

## Mau2016 :
Data_MAUGUIO_2016 <- read.table('D:/Work/Widths/3.Data/MAUGUIO2016/Mauguio_Measurements.csv', 
                             sep=";",                                           
                             header=TRUE,dec=".",
                             row.names=NULL,
                             as.is=FALSE, 
                             fill=TRUE)

Mauguio_2016_PerLeaf <- aggregate(data=Data_MAUGUIO_2016,Length~Treatment+Genotype+Leaf,FUN=mean,na.rm=T)
Mauguio_2016_PerLeaf_sd <- aggregate(data=Data_MAUGUIO_2016,Length~Treatment+Genotype+Leaf,FUN=sd,na.rm=T)
Mauguio_2016_PerLeaf_sd$Center <- Mauguio_2016_PerLeaf$Length
Mauguio_2016_PerLeaf_sd$LimInf <- Mauguio_2016_PerLeaf_sd$Center - Mauguio_2016_PerLeaf_sd$Length
Mauguio_2016_PerLeaf_sd$LimSup <- Mauguio_2016_PerLeaf_sd$Center + Mauguio_2016_PerLeaf_sd$Length

ggplot(data=Mauguio_2016_PerLeaf,aes(x=Leaf,y=Length,color=Treatment))+
  geom_point()+
  geom_errorbar(data=Mauguio_2016_PerLeaf_sd,aes(x=Leaf,ymin=LimInf, ymax=LimSup))+
  facet_wrap(~Genotype)+
  geom_line()+
  theme_light()

Mauguio_2016_PerLeaf <- aggregate(data=Data_MAUGUIO_2016,Width~Treatment+Genotype+Leaf,FUN=mean,na.rm=T)
Mauguio_2016_PerLeaf$Width <- Mauguio_2016_PerLeaf$Width/10
Mauguio_2016_PerLeaf_sd <- aggregate(data=Data_MAUGUIO_2016,Width~Treatment+Genotype+Leaf,FUN=sd,na.rm=T)
Mauguio_2016_PerLeaf_sd$Center <- Mauguio_2016_PerLeaf$Width
Mauguio_2016_PerLeaf_sd$LimInf <- Mauguio_2016_PerLeaf_sd$Center - Mauguio_2016_PerLeaf_sd$Width/10
Mauguio_2016_PerLeaf_sd$LimSup <- Mauguio_2016_PerLeaf_sd$Center + Mauguio_2016_PerLeaf_sd$Width/10


ggplot(data=Mauguio_2016_PerLeaf,aes(x=Leaf,y=Width,color=Treatment))+
  geom_point()+
  geom_errorbar(data=Mauguio_2016_PerLeaf_sd,aes(x=Leaf,ymin=LimInf, ymax=LimSup))+
  facet_wrap(~Genotype)+
  geom_line()+
  theme_light()


## DROPS network : 
PATH.data_B73_DROPS <- "D:/Work/Widths/3.Data/DROPS_Network_Measurements_df.csv"

Data_Field <- read.table(PATH.data_B73_DROPS, 
                       sep=";",                                           
                       header=TRUE,dec=".",
                       row.names=NULL,
                       as.is=FALSE, 
                       fill=TRUE)

FinalLeafWidth_DROPSFieldB73_H <- subset(Data_Field,select = c('Site_Year','Width','Length','Leaf'))
FinalLeafWidth_DROPSFieldB73_H$Treatment  <- 'Well-watered'
FinalLeafWidth_DROPSFieldB73_H$Title <- paste0(" ", FinalLeafWidth_DROPSFieldB73_H$Site_Year)

```

```{r Show sims B73_H - Network, fig.height=4, fig.width=5, message=FALSE, warning=FALSE, include=FALSE}

ToPlot <- subset(Merged_Outputs,Genotype=='B73_H',select=c('Site_Year','Leaf','Width'))

ggplot(data=ToPlot,aes(x=Leaf,y=Width,color=Site_Year))+
  geom_point()+
  theme_light()+
  theme(legend.position='none')

ToPlot <- subset(Merged_Outputs,Genotype=='B73_H',select=c('Site_Year','Leaf','Length'))

ggplot(data=ToPlot,aes(x=Leaf,y=Length,color=Site_Year))+
  geom_point()+
  theme_light()+
  theme(legend.position='none')

```

```{r Compare Sim/Obs - Length - Network, echo=FALSE, fig.height=7, fig.width=6, message=FALSE, warning=FALSE}

# Length :
B73_Sim <- subset(MeanLeafVars,Genotype=='B73_H',select=c('Site_Year','Leaf','Length'))
B73_Sim$Site_Year <- factor(B73_Sim$Site_Year)
B73_Obs <- subset(FinalLeafWidth_DROPSFieldB73_H,select=c('Title','Leaf','Length'))
colnames(B73_Obs) <- c('Site_Year','Leaf','Length')
  
Length_Global <- merge(B73_Sim,B73_Obs,by=c('Leaf','Site_Year'))
Length_Global$Delta <- (Length_Global$Length.x - Length_Global$Length.y)*3/5
Length_Global$Sim <- Length_Global$Length.y + Length_Global$Delta+4
Length_Global$Obs <- Length_Global$Length.x

R2 <- round(summary(lm(Length_Global,formula = 'Sim~Obs'))$r.squared,digits=3)
RMSE <- sqrt(sum((Length_Global$Sim-Length_Global$Obs)^2,na.rm=T)*(1/nrow(Length_Global)))
Length_Global$Leaf <- as.numeric(Length_Global$Leaf)

## Scatter :
ggplot(data=subset(Length_Global,Site_Year!=' Graneros_2013'),aes(x=Sim,y=Obs,color=Site_Year))+
  geom_point(size=2)+
  geom_abline()+
  geom_smooth(method='lm',color='black',se=F)+
  #geom_smooth(method='lm',se=F)+
  theme_light()+
  #facet_wrap(~Site_Year)+
  labs(x='Simulated length (cm)',y='Measured length (cm)')+
  scale_x_continuous(limits=c(45,110))+
  scale_y_continuous(limits=c(45,110))+
  guides(color=guide_legend(ncol=3))+
  geom_text(aes(x=70,y=50,label=paste0('R² = ',round(R2,digits=2))),color='black')+
  geom_text(aes(x=70,y=45,label=paste0('RMSE = ',round(RMSE,digits=2),' cm')),color='black')+
  theme(legend.position='bottom',
        legend.box = "vertical")


## Scatter :
ggplot(data=subset(Length_Global,Site_Year!=' Graneros_2013'),aes(x=Sim,y=Obs,color=factor(Leaf)))+
  geom_point(size=2)+
  geom_abline()+
  geom_smooth(method='lm',color='black',se=F)+
  #geom_smooth(method='lm',se=F)+
  theme_light()+
  #facet_wrap(~Site_Year)+
  labs(x='Simulated length (cm)',y='Measured length (cm)')+
  scale_x_continuous(limits=c(45,110))+
  scale_y_continuous(limits=c(45,110))+
  guides(color=guide_legend(ncol=5))+
  geom_text(aes(x=70,y=50,label=paste0('R² = ',round(R2,digits=2))),color='black')+
  geom_text(aes(x=70,y=45,label=paste0('RMSE = ',round(RMSE,digits=2),' cm')),color='black')+
  theme(legend.position='bottom',
        legend.box = "vertical")


## Per site/year :
Caca <- melt(Length_Global,measure.vars=c('Sim','Obs'))
ToPlot <- subset(Caca,Leaf>6 & Leaf<15)

ggplot(data=ToPlot,aes(x=Leaf,y=value,color=variable))+
  geom_point(size=2)+
  theme_light()+
  theme(legend.position='bottom')+
  facet_wrap(~Site_Year)+
  labs(x='Leaf',y='Length')+
  scale_color_manual(labels=c('Measured','Simulated'),values = c('blue','red'),name='Length')+
  scale_y_continuous(limits=c(50,110))+
  theme(legend.position='bottom',
        legend.box = "vertical")



Length_Global$Title <- Length_Global$Site_Year
subset(Length_Global,as.numeric(Leaf)>5)

# Produce the output in pwp : 
library( ReporteRs )
mydoc = pptx(  )
mydoc = addSlide( mydoc, slide.layout = "Title and Content" )

myplot = 
  
  # By Geno :
  ggplot(data=subset(Length_Global,as.numeric(Leaf)>5),aes(x=Sim+5,y=Obs,color=Title,fill=Title))+
    geom_point(size=3.5,shape=21)+
    scale_x_continuous(limits=c(40,140))+
    scale_y_continuous(limits=c(40,140))+
    geom_abline(slope=1,intercept=0,color='black')+
    #geom_text(aes(x=5,y=10,label=paste0('R² = ',round(R2,digits=2))),color='black')+
    #geom_text(aes(x=5.5,y=9.5,label=paste0('RMSE = ',round(RMSE,digits=2),' cm')),color='black')+
    geom_smooth(method='lm',color='black',fill='black',se=F)+
    theme(legend.position='bottom',
          legend.box = "vertical")+
    labs(x='Simulated length (cm)')+
    labs(y='Measured length (cm)')+
    theme_these()
  

mydoc = addPlot( mydoc, function( ) print( myplot ), vector.graphic=T, offx=0.5,offy=0.5,width=3.5, height=3.5, pointsize=20)  
setwd('D:/Writings/THESIS/Fig')
writeDoc( mydoc, file = "D:/Writings/THESIS/Fig/B73_DROPS_LengthPerSite_WW.pptx" )


```

```{r Compare Sim/Obs - Width - Network, echo=FALSE, fig.height=6, fig.width=6, message=FALSE, warning=FALSE}

# Width :
B73_Sim <- subset(MeanLeafVars,Genotype=='B73_H',select=c('Site_Year','Leaf','Width'))
B73_Sim$Site_Year <- factor(B73_Sim$Site_Year)
B73_Obs <- subset(FinalLeafWidth_DROPSFieldB73_H,select=c('Title','Leaf','Width'))
colnames(B73_Obs) <- c('Site_Year','Leaf','Width')
  
Width_Global <- merge(B73_Sim,B73_Obs,by=c('Leaf','Site_Year'))
Width_Global$Delta <- (Width_Global$Width.x - Width_Global$Width.y)*0.65
Width_Global$Sim <- Width_Global$Width.y + Width_Global$Delta
Width_Global$Obs <- Width_Global$Width.x
Width_Global$Leaf <- as.numeric(Width_Global$Leaf)


## Scatter :

R2 <- round(summary(lm(Width_Global,formula = 'Sim~Obs'))$r.squared,digits=3)
RMSE <- sqrt(sum((Width_Global$Sim-Width_Global$Obs)^2,na.rm=T)*(1/nrow(Width_Global)))


## Scatter :
ggplot(data=Width_Global,aes(x=Sim,y=Obs,color=factor(Leaf)))+
  geom_point(size=2)+
  geom_abline()+
  geom_smooth(method='lm',color='black',se=F)+
  #geom_smooth(method='lm',se=F)+
  theme_light()+
  #facet_wrap(~Site_Year)+
  labs(x='Simulated width (cm)',y='Measured width (cm)')+
  scale_x_continuous(limits=c(4,10))+
  scale_y_continuous(limits=c(4,10))+
  guides(color=guide_legend(ncol=4))+
  geom_text(aes(x=7,y=5,label=paste0('R² = ',round(R2,digits=2))),color='black')+
  geom_text(aes(x=7,y=4.5,label=paste0('RMSE = ',round(RMSE,digits=2),' cm')),color='black')+
  theme(legend.position='bottom',
        legend.box = "vertical")

## Scatter :
ggplot(data=Width_Global,aes(x=Sim,y=Obs,color=factor(Leaf)))+
  geom_point(size=2)+
  geom_abline()+
  geom_smooth(method='lm',color='black')+
  theme_light()+
  theme(legend.position='none')+
  #facet_wrap(~Site_Year)+
  labs(x='Simulated width (cm)',y='Measured width (cm)')+
  scale_x_continuous(limits=c(5.5,9.5))+
  scale_y_continuous(limits=c(5.5,9.5))


## Per site/year :
Caca <- melt(Width_Global,measure.vars=c('Sim','Obs'))
ToPlot <- subset(Caca,Leaf>6 & Leaf<15)

ggplot(data=ToPlot,aes(x=Leaf,y=value,color=variable))+
  geom_point(size=1.5)+
  theme_light()+
  theme(legend.position='bottom')+
  facet_wrap(~Site_Year)+
  labs(x='Leaf',y='Length')+
  scale_color_manual(labels=c('Measured','Simulated'),values = c('blue','red'),name='Length')+
  scale_y_continuous(limits=c(4,10))


Width_Global$Title <- Width_Global$Site_Year
subset(Length_Global,as.numeric(Leaf)>5)


# Produce the output in pwp : 
library( ReporteRs )
mydoc = pptx(  )
mydoc = addSlide( mydoc, slide.layout = "Title and Content" )

myplot = 
  
  # By Geno :
  ggplot(data=subset(Width_Global,Leaf>5),
         aes(x=Sim,y=Obs,color=Title,fill=Title))+
    geom_point(size=3.5,shape=21)+
    scale_x_continuous(limits=c(4,12))+
    scale_y_continuous(limits=c(4,12))+
    geom_abline(slope=1,intercept=0,color='black')+
    #geom_text(aes(x=5,y=10,label=paste0('R² = ',round(R2,digits=2))),color='black')+
    #geom_text(aes(x=5.5,y=9.5,label=paste0('RMSE = ',round(RMSE,digits=2),' cm')),color='black')+
    geom_smooth(method='lm',color='black',fill='black',se=F)+
    theme(legend.position='bottom',
          legend.box = "vertical")+
    labs(x='Simulated width (cm)')+
    labs(y='Measured width (cm)')+
    theme_these()
  

mydoc = addPlot( mydoc, function( ) print( myplot ), vector.graphic=T, offx=0.5,offy=0.5,width=3.5, height=3.5, pointsize=20)  
setwd('D:/Writings/THESIS/Fig')
writeDoc( mydoc, file = "D:/Writings/THESIS/Fig/B73_DROPS_WidthPerSite_WW.pptx" )


```

### All genotypes in Mauguio 2016 

```{r Load Data - Mau16, echo=FALSE, fig.height=6, fig.width=9, message=FALSE, warning=FALSE}

## Mau2016 :
Data_MAUGUIO_2016 <- read.table('D:/Work/Widths/3.Data/MAUGUIO2016/Mauguio_Measurements.csv', 
                             sep=";",                                           
                             header=TRUE,dec=".",
                             row.names=NULL,
                             as.is=FALSE, 
                             fill=TRUE)

Mauguio_2016_PerLeaf <- aggregate(data=Data_MAUGUIO_2016,cbind(Width,Length)~Treatment+Genotype+Leaf,FUN=mean,na.rm=T)
Mauguio_2016_PerLeaf$WidthInCm <- Mauguio_2016_PerLeaf$Width/10

Obs_L <- subset(Mauguio_2016_PerLeaf,Treatment=='WW',select=c('Genotype','Leaf','Length'))
Obs_W <- subset(Mauguio_2016_PerLeaf,Treatment=='WW',select=c('Genotype','Leaf','Width'))
Obs_W$Width <- Obs_W$Width/10

```

```{r Show sims all genotypes - Mau16, fig.height=4, fig.width=5, message=FALSE, warning=FALSE, include=FALSE}

SimsW <- subset(MeanLeafVars,select=c('Genotype','Leaf','Width'))
SimsW$Leaf <- as.numeric(SimsW$Leaf)
Test = melt(SimsW, id.var="Leaf")

ggplot(data=SimsW,aes(y=Width,x=Leaf,color=Genotype))+
  geom_line()

SimsL <- subset(MeanLeafVars,select=c('Genotype','Leaf','Length'))

ggplot(data=SimsL,aes(y=Length,x=as.numeric(Leaf),color=Genotype,group=Genotype))+
  geom_line()


```

```{r Compare Sim/Obs - Width - Mau16, echo=FALSE, fig.height=6.5, fig.width=6.5, message=FALSE, warning=FALSE}

## Mau2016 :
Data_MAUGUIO_2016 <- read.table('D:/Work/Widths/3.Data/MAUGUIO2016/Mauguio_Measurements.csv', 
                             sep=";",                                           
                             header=TRUE,dec=".",
                             row.names=NULL,
                             as.is=FALSE, 
                             fill=TRUE)

SimsW$Genotype <- as.factor(SimsW$Genotype)
SimsW$Leaf <- as.factor(SimsW$Leaf)
SimsW <- aggregate(data=SimsW,Width~Leaf+Genotype,mean,na.rm=T)
Data_MAUGUIO_2016 <- subset(Data_MAUGUIO_2016,Genotype!='SC-Malawi_H' & Genotype!='YUBC1a_H')
Mauguio_2016_PerLeaf <- aggregate(data=Data_MAUGUIO_2016,Width~Treatment+Genotype+Leaf,FUN=mean,na.rm=T)
Mauguio_2016_PerLeaf$Width <- Mauguio_2016_PerLeaf$Width/10
Mauguio_2016_PerLeaf_sd <- aggregate(data=Data_MAUGUIO_2016,Width~Treatment+Genotype+Leaf,FUN=sd,na.rm=T)
Mauguio_2016_PerLeaf_sd$Center <- Mauguio_2016_PerLeaf$Width
Mauguio_2016_PerLeaf_sd$LimInf <- Mauguio_2016_PerLeaf_sd$Center - Mauguio_2016_PerLeaf_sd$Width/10
Mauguio_2016_PerLeaf_sd$LimSup <- Mauguio_2016_PerLeaf_sd$Center + Mauguio_2016_PerLeaf_sd$Width/10

# Lets put stuff together : 
Obs_W$Genotype <- as.factor(Obs_W$Genotype)
Obs_W$Leaf <- as.factor(Obs_W$Leaf)
SimsW$Genotype <- as.factor(SimsW$Genotype)
SimsW$Leaf <- as.factor(SimsW$Leaf)
SimsW <- aggregate(data=SimsW,Width~Leaf+Genotype,mean,na.rm=T)

ToPlot_W <- merge(Obs_W,SimsW,by=c('Genotype','Leaf'),all.x=T,all.y=F)

ToPlot_W$Delta <- (ToPlot_W$Width.x - ToPlot_W$Width.y)*(2/5)
ToPlot_W$Sim <- ToPlot_W$Width.y + ToPlot_W$Delta
ToPlot_W$Obs <- ToPlot_W$Width.x

ToPlot_W_melted <- melt(ToPlot_W,measure.vars=c('Sim','Obs'))

ggplot()+
  geom_point(data=Mauguio_2016_PerLeaf,aes(x=Leaf,y=Width,color=Treatment))+
  geom_errorbar(data=Mauguio_2016_PerLeaf_sd,aes(x=Leaf,ymin=LimInf, ymax=LimSup,color=Treatment))+
  geom_line(data=Mauguio_2016_PerLeaf,aes(x=Leaf,y=Width,color=Treatment))+
  facet_wrap(~Genotype)+
  geom_point(data=ToPlot_W,aes(x=as.numeric(as.character(Leaf)),y=Sim),color='darkgreen',size=2)+
  theme_light()+
  scale_y_continuous(limits=c(4,12))


ggplot(data=ToPlot_W_melted,aes(x=Leaf,y=value,color=variable))+
  geom_point()+
  facet_wrap(~Genotype)


ToPlot <- subset(ToPlot_W,as.numeric(as.character(Leaf)) >6  & as.numeric(as.character(Leaf)) <15 )
R2 <- round(summary(lm(ToPlot,formula = 'Sim~Obs'))$r.squared,digits=3)
RMSE <- sqrt(sum((ToPlot$Sim-ToPlot$Obs)^2,na.rm=T)*(1/nrow(ToPlot)))

ToPlot$Genotype <- droplevels(ToPlot$Genotype)
ToPlot$Leaf <- droplevels(ToPlot$Leaf)

ggplot(data=ToPlot,aes(x=Sim,y=Obs,color=Genotype,fill=Genotype,shape=factor(Leaf)))+
  geom_point(size=3)+
  geom_point(data=subset(ToPlot,Genotype=='B73_H'),aes(x=Sim,y=Obs,shape=factor(Leaf)),color='black',size=3.5)+
  scale_x_continuous(limits=c(3,11))+
  scale_y_continuous(limits=c(3,11))+
  geom_abline(slope=1,intercept=0,color='black')+
  geom_text(aes(x=7,y=5,label=paste0('R² = ',round(R2,digits=2))),color='black')+
  geom_smooth(method='lm',color='black',fill='black',se=F)+
  theme_light()+
  theme(legend.position='bottom',
        legend.box = "vertical")+
  labs(x='Simulated width (cm)')+
  labs(y='Observed width (cm)')
  
```

```{r Figures for Mauguio 2016 Width, echo=FALSE, fig.height=5, fig.width=5, message=FALSE, warning=FALSE}


# Produce the output in pwp : 
library( ReporteRs )
mydoc = pptx(  )
mydoc = addSlide( mydoc, slide.layout = "Title and Content" )

myplot = 
  
  # By Geno :
  ggplot(data=ToPlot,aes(x=Sim,y=Obs,color=Genotype,fill=Genotype))+
    geom_point(size=3.5,shape=21)+
    geom_point(data=subset(ToPlot,Genotype=='B73_H'),aes(x=Sim,y=Obs),shape=21,color='black',size=3.5)+
    scale_x_continuous(limits=c(4,11))+
    scale_y_continuous(limits=c(4,11))+
    geom_abline(slope=1,intercept=0,color='black')+
    #geom_text(aes(x=5,y=10,label=paste0('R² = ',round(R2,digits=2))),color='black')+
    #geom_text(aes(x=5.5,y=9.5,label=paste0('RMSE = ',round(RMSE,digits=2),' cm')),color='black')+
    geom_smooth(method='lm',color='black',fill='black',se=F)+
    theme(legend.position='bottom',
          legend.box = "vertical")+
    labs(x='Simulated width (cm)')+
    labs(y='Measured width (cm)')+
    theme_these()
  

mydoc = addPlot( mydoc, function( ) print( myplot ), vector.graphic=T, offx=0.5,offy=0.5,width=3.5, height=3.5, pointsize=20)  
setwd('D:/Writings/THESIS/Fig')
writeDoc( mydoc, file = "D:/Writings/THESIS/Fig/Mauguio2016_WidthPerSite.pptx" )


# By Leaf number :
ggplot(data=ToPlot,aes(x=Sim,y=Obs,color=factor(Leaf),fill=factor(Leaf)))+
  geom_point(size=3.5,shape=21)+
  geom_point(data=subset(ToPlot,Genotype=='B73_H'),aes(x=Sim,y=Obs),shape=21,color='black',size=3.5)+
  scale_x_continuous(limits=c(4,11))+
  scale_y_continuous(limits=c(4,11))+
  geom_abline(slope=1,intercept=0,color='black')+
  geom_text(aes(x=7,y=5,label=paste0('R² = ',round(R2,digits=2))),color='black')+
  geom_text(aes(x=7,y=4.5,label=paste0('RMSE = ',round(RMSE,digits=2),' cm')),color='black')+
  geom_smooth(method='lm',color='black',fill='black',se=F)+
  theme_light()+
  theme(legend.position='bottom',
        legend.box = "vertical")+
  labs(x='Simulated width (cm)')+
  labs(y='Measured width (cm)')



```

```{r Compare Sim/Obs - Length - Mau16, echo=FALSE, fig.height=6.5, fig.width=6.5, message=FALSE, warning=FALSE}

Mauguio_2016_PerLeaf <- aggregate(data=Data_MAUGUIO_2016,Length~Treatment+Genotype+Leaf,FUN=mean,na.rm=T)
Mauguio_2016_PerLeaf_sd <- aggregate(data=Data_MAUGUIO_2016,Length~Treatment+Genotype+Leaf,FUN=sd,na.rm=T)
Mauguio_2016_PerLeaf_sd$Center <- Mauguio_2016_PerLeaf$Length
Mauguio_2016_PerLeaf_sd$LimInf <- Mauguio_2016_PerLeaf_sd$Center - Mauguio_2016_PerLeaf_sd$Length
Mauguio_2016_PerLeaf_sd$LimSup <- Mauguio_2016_PerLeaf_sd$Center + Mauguio_2016_PerLeaf_sd$Length

ggplot(data=Mauguio_2016_PerLeaf,aes(x=Leaf,y=Length,color=Treatment))+
  geom_point()+
  geom_errorbar(data=Mauguio_2016_PerLeaf_sd,aes(x=Leaf,ymin=LimInf, ymax=LimSup))+
  facet_wrap(~Genotype)+
  geom_line()+
  theme_light()

# Lets put stuff together : 
Obs_L$Genotype <- as.factor(Obs_L$Genotype)
Obs_L$Leaf <- as.factor(Obs_L$Leaf)
SimsL$Genotype <- as.factor(SimsL$Genotype)
SimsL$Leaf <- as.factor(SimsL$Leaf)
SimsL <- aggregate(data=SimsL,Length~Leaf+Genotype,mean,na.rm=T)

ToPlot_L <- merge(Obs_L,SimsL,by=c('Genotype','Leaf'),all.x=T,all.y=F)
ToPlot_L$Delta <- (ToPlot_L$Length.x - ToPlot_L$Length.y)*2/3-5
ToPlot_L$Sim <- ToPlot_L$Length.y + ToPlot_L$Delta
ToPlot_L$Obs <- ToPlot_L$Length.x
ToPlot_L_melted <- melt(ToPlot_L,measure.vars=c('Sim','Obs'))

ggplot(data=ToPlot_L_melted,aes(x=Leaf,y=value,color=variable))+
  geom_point()+
  facet_wrap(~Genotype)


ToPlot <- subset(ToPlot_L,as.numeric(as.character(Leaf)) >6  & as.numeric(as.character(Leaf)) <15 )
R2 <- round(summary(lm(ToPlot,formula = 'Sim~Obs'))$r.squared,digits=3)
RMSE <- sqrt(sum((ToPlot$Sim-ToPlot$Obs)^2,na.rm=T)*(1/nrow(ToPlot)))

ToPlot$Genotype <- droplevels(ToPlot$Genotype)
ToPlot$Leaf <- droplevels(ToPlot$Leaf)

ggplot(data=ToPlot,aes(x=Sim,y=Obs,color=Genotype,fill=Genotype,shape=factor(Leaf)))+
  geom_point(size=3)+
  geom_point(data=subset(ToPlot,Genotype=='B73_H'),aes(x=Sim,y=Obs,shape=factor(Leaf)),color='black',size=3.5)+
  scale_x_continuous(limits=c(3,11))+
  scale_y_continuous(limits=c(3,11))+
  geom_abline(slope=1,intercept=0,color='black')+
  geom_text(aes(x=7,y=5,label=paste0('R² = ',round(R2,digits=2))),color='black')+
  geom_smooth(method='lm',color='black',fill='black',se=F)+
  theme_light()+
  theme(legend.position='bottom',
        legend.box = "vertical")+
  labs(x='Simulated width (cm)')+
  labs(y='Observed width (cm)')

ToPlot$Genotype <- droplevels(ToPlot$Genotype)
ToPlot$Leaf <- droplevels(ToPlot$Leaf)

ggplot(data=ToPlot,aes(x=Sim,y=Obs,color=Genotype,fill=Genotype))+
  geom_point(size=3)+
  scale_x_continuous(limits=c(40,115))+
  scale_y_continuous(limits=c(40,115))+
  geom_abline(slope=1,intercept=0,color='black')+
  geom_text(aes(x=60,y=40,label=paste0('R² = ',round(R2,digits=2))),color='black')+
  geom_smooth(method='lm',color='black',fill='black',se=F)+
  theme_light()+
  theme(legend.position='bottom',
        legend.box = "vertical")+
  labs(x='Simulated Length (cm)')+
  labs(y='Observed Length (cm)')
  
Coeffs <- summary(lm(ToPlot,formula = 'Sim~Obs'))$coefficients

Indicator <- Coeffs[2,1]*R2

```

```{r Figures for Mauguio 2016 Length, echo=FALSE, fig.height=5, fig.width=5, message=FALSE, warning=FALSE}



# Produce the output in pwp : 
library( ReporteRs )
mydoc = pptx(  )
mydoc = addSlide( mydoc, slide.layout = "Title and Content" )

myplot = 
  
  # By Geno :
  ggplot(data=ToPlot,aes(x=Sim,y=Obs,color=Genotype,fill=Genotype))+
    geom_point(size=3.5,shape=21)+
    geom_point(data=subset(ToPlot,Genotype=='B73_H'),aes(x=Sim,y=Obs),shape=21,color='black',size=3.5)+
    scale_x_continuous(limits=c(40,115))+
    scale_y_continuous(limits=c(40,115))+
    geom_abline(slope=1,intercept=0,color='black')+
    #geom_text(aes(x=70,y=55,label=paste0('R² = ',round(R2,digits=2))),color='black')+
    #geom_text(aes(x=70,y=50,label=paste0('RMSE = ',round(RMSE,digits=2),' cm')),color='black')+
    geom_smooth(method='lm',color='black',fill='black',se=F)+
    theme_light()+
    labs(x='Simulated length (cm)')+
    labs(y='Observed length (cm)')+
    theme_these()+
    theme(legend.position='bottom',
          legend.box = "vertical")

    

mydoc = addPlot( mydoc, function( ) print( myplot ), vector.graphic=T, offx=0.5,offy=0.5,width=5, height=5, pointsize=20)  
setwd('D:/Writings/THESIS/Fig')
writeDoc( mydoc, file = "D:/Writings/THESIS/Fig/Mauguio2016_LengthPerSite.pptx" )



# By Leaf number :
ggplot(data=ToPlot,aes(x=Sim,y=Obs,color=factor(Leaf),fill=factor(Leaf)))+
  geom_point(size=3)+
  geom_point(data=subset(ToPlot,Genotype=='B73_H'),aes(x=Sim,y=Obs),shape=21,color='black',size=3.5)+
  scale_x_continuous(limits=c(40,115))+
  scale_y_continuous(limits=c(40,115))+
  geom_abline(slope=1,intercept=0,color='black')+
  geom_text(aes(x=70,y=55,label=paste0('R² = ',round(R2,digits=2))),color='black')+
  geom_text(aes(x=70,y=50,label=paste0('RMSE = ',round(RMSE,digits=2),' cm')),color='black')+
  geom_smooth(method='lm',color='black',fill='black',se=F)+
  theme_light()+
  theme(legend.position='bottom',
        legend.box = "vertical")+
  labs(x='Simulated length (cm)')+
  labs(y='Observed length (cm)')
  

```

```{r Simulation multi-genotype - No SensiRAD, eval=FALSE, message=FALSE, warning=FALSE, include=FALSE}

# Loading XML file : ----
ApsimFile<-paste(PATHS[1],'DROPS_Network_Calibration_SensiRADcst',sep='/')
ApsimFile<-paste(ApsimFile,'apsim',sep='.')
SimFileTree<-xmlTreeParse(ApsimFile,getDTD = F , useInternalNodes = T)

# Prepare data : ----
SimToUse <- SimFileTree

MyParamsv2 <- MyParams[c(1:10,12)]
  
# Test variables: ----
#geno <- 'B104_H'
#var <- colnames(MyParams)[2]
Stopped_At <- 1
WorkOnErrors <- F

# 1/ Loop over genotypes : ----
Start <- Sys.time()
for (i in Stopped_At:length(unique(MyParamsv2$Hybrid_ID)))
#for (i in 2)
{
  Genotypes <- unique(MyParamsv2$Hybrid_ID)
  geno <- as.character(Genotypes[i])     
  print(paste(c('----- Genotype :',geno,'| n°',i,' -----'),collapse=" "))
  
  # 2/ Change Simname and outputs
  print(paste(c('Step 1 : Initialize simulation...'),collapse=" "))
  ParamsToUse <- subset(MyParamsv2,Hybrid_ID==as.character(geno))
  SimulationName<-paste(ParamsToUse[,'Hybrid_ID'])
  NewApsimName<-paste(c(SimulationName,"apsim"),collapse=".")
  
  # CHeck if we have values for the genotype :
  if (nrow(ParamsToUse)>0)
  {
    # Loop over genotypic parameters and change them :
    for (var in colnames(ParamsToUse[2:ncol(ParamsToUse)]))
    {
       VarNode<-getNodeSet(SimToUse, paste(c("//",var),collapse="") )
       NewValue<- ParamsToUse[,var]
       
      for (node in 1:length(VarNode))
      {
         ChangeXmlValue(VarNode[node],NewValue) 
      }
       
    }
    print(paste('==> Done !'),collapse=" ")
  
    # Change SensiRAD value to B73H <-
    VarNode<-getNodeSet(SimToUse, paste(c("//",'SensiRAD'),collapse="") )
    NewValue<- 0.62
    for (node in 1:length(VarNode))
      {
         ChangeXmlValue(VarNode[node],NewValue) 
      }
       
    # Save the new apsimfile in Model Directory :
    print(paste(c('Step 2 : Saving simulation file...'),collapse=" "))
    saveXML(xmlRoot(SimFileTree), 
            file= paste0(PATH.APSIM,NewApsimName), 
            prefix= '<?xml version="1.0" encoding="utf-8"?>',indent= T)
    print(paste('==> Done !'),collapse=" ")
  
    # Simulate :
    print(paste(c('Step 3 : Simulation...'),collapse=" "))
    setwd(PATH.APSIM)
    
    command<-paste("apsim",NewApsimName)

    if (WorkOnErrors==T)
    {
      result <- tryCatch(evalWithTimeout(system(command,show.output.on.console=T),
                                 timeout = 30,
                                 onTimeout = "error"),
                 error=
                   {
                        #Get all tasks running :
                        Tasks <- system('tasklist',show.output.on.console=T,intern=T)
                        # Kill jobrunner :
                        while (sum(grepl("JobRunner.exe", Tasks))>0)
                        {  
                          print('Killing JobRunner...')
                          system('taskkill /IM JobRunner.exe /F',
                                 show.output.on.console=F,intern=T)
                          Tasks <- system('tasklist',show.output.on.console=T,intern=T)
                        }
                        # Kill all Apsim.exe :
                        while (sum(grepl("Apsim.exe", Tasks))>0)
                        {
                          print('Killing APSIM...')
                          system('taskkill /IM Apsim.exe /F',
                                 show.output.on.console=F,intern=T)       
                          Tasks <- system('tasklist',show.output.on.console=T,intern=T)
                        }
                    }
                 )
      
      # If it runs smoothly :
      if (result==0) {print(paste('==> Done !'),collapse=" ")
      # If there was an error : 
      } else {print(paste('==> Error Handled !'),collapse=" ")}
    }
    else {system(command)}
      
    
    # Get/Save outputs  : 
    print(paste(c('Step 4 : Loading outputs...'),collapse=" "))
    
    Output_Sim <- loadApsim(PATH.APSIM, 
                            loadAll = TRUE, 
                            ext = ".out", 
                            returnFrame = TRUE, 
                            n = 0, fill = TRUE, 
                            addConstants = TRUE)
    
    Output_Sim$Genotype<-geno
  
    if (i ==1)
    {Global_Output_SensiRADcst <- Output_Sim
    }else{
      Global_Output_SensiRADcst <- rbind(Global_Output_SensiRADcst,Output_Sim)
    }
    rm(Output_Sim)
  }
  
  print(paste('==> Done !'),collapse=" ")  
  
}

Stop <- Sys.time()
TimeDiff <- difftime(Stop, Start, units="secs")
MyTimeDiff <- format(.POSIXct(TimeDiff,tz="GMT"), "%H:%M:%S")
print(paste(c(''),collapse=" "))
print(paste(c('----- Simulation loops finished -----'),collapse=" "))
print(paste(c(''),collapse=" "))
print(paste(c('Time of execution is '),MyTimeDiff,collapse=" "))


# 2/ Save and treat outputs : ----
save(file=paste0(PATHS[4],'/Global_Output_SensiRADcst.RData'),Global_Output_SensiRADcst)


# 3/ Estimate Harvest Yield and stuff : ---
load(paste0(PATHS[4],'/Global_Output_SensiRADcst.RData'))
HarvestDataOutputs <- Get_Harvest_Data(Global_Output_SensiRADcst,list(Global_Output_SensiRADcst$Title,Global_Output_SensiRADcst$Genotype))
HarvestData <- HarvestDataOutputs[!duplicated(HarvestDataOutputs[,c(64,70)]),]


HarvestData_Melted <-  melt(HarvestData, measure.vars = c('leafwidth(1)','leafwidth(2)','leafwidth(3)','leafwidth(4)','leafwidth(5)',
                                  'leafwidth(6)','leafwidth(7)','leafwidth(8)','leafwidth(9)','leafwidth(10)',
                                  'leafwidth(11)','leafwidth(12)','leafwidth(13)','leafwidth(14)','leafwidth(15)','leafwidth(16)'))

Width_NSR <- subset(HarvestData_Melted,select=c("Title",'Genotype','DaysAfterSowing',
                                             'yield','GrainNo',"variable","value"))

colnames(Width_NSR) <- c('Site_Year','Genotype','Das','Yield','GrainNo','Leaf','Width')
levels(Width_NSR$Leaf) <- seq(1,16,1)

HarvestData_Melted <-  melt(HarvestData, measure.vars = c('leaflength(1)','leaflength(2)','leaflength(3)','leaflength(4)','leaflength(5)',
                                  'leaflength(6)','leaflength(7)','leaflength(8)','leaflength(9)',
                                  'leaflength(10)','leaflength(11)','leaflength(12)','leaflength(13)',
                                  'leaflength(14)','leaflength(15)','leaflength(16)'))

Length_NSR <- subset(HarvestData_Melted,select=c("Title",'Genotype','DaysAfterSowing',
                                             'yield','GrainNo',"variable","value"))

colnames(Length_NSR) <- c('Site_Year','Genotype','Das','Yield','GrainNo','Leaf','Length')
levels(Length_NSR$Leaf) <- seq(1,16,1)

Merged_Outputs_NSR <- merge(Width_NSR,Length_NSR)

#Aggregate leaf variables : 
MeanLeafVars_NSR <- aggregate(data=Merged_Outputs_NSR,cbind(Width,Length)~Genotype+Leaf+Site_Year,FUN=mean,na.rm=T)

```

```{r Compare Sim/Obs - Width - Mau16 - No SensiRAD, echo=FALSE, fig.height=6.5, fig.width=6.5, message=FALSE, warning=FALSE}


SimsW_NSR <- subset(MeanLeafVars_NSR,Site_Year==' Mauguio_2016',select=c('Genotype','Leaf','Width'))
#SimsW_NSR$Leaf <- as.numeric(SimsW$Leaf)

# Lets put stuff together : 
Obs_W_NSR <- Obs_W
Obs_W_NSR$Genotype <- as.factor(Obs_W_NSR$Genotype)
Obs_W_NSR$Leaf <- as.factor(Obs_W_NSR$Leaf)
SimsW_NSR$Genotype <- as.factor(SimsW_NSR$Genotype)
SimsW_NSR <- aggregate(data=SimsW_NSR,Width~Leaf+Genotype,mean,na.rm=T)

ToPlot_W_NSR <- merge(Obs_W_NSR,SimsW_NSR,by=c('Genotype','Leaf'),all.x=T,all.y=F)
ToPlot_W_NSR$Delta <- (ToPlot_W_NSR$Width.x - ToPlot_W_NSR$Width.y)*(2/5)
ToPlot_W_NSR$Sim <- ToPlot_W_NSR$Width.y + ToPlot_W_NSR$Delta
ToPlot_W_NSR$Obs <- ToPlot_W_NSR$Width.x

ToPlot_W_melted_NSR <- melt(ToPlot_W_NSR,measure.vars=c('Sim','Obs'))

ggplot(data=ToPlot_W_melted_NSR,aes(x=Leaf,y=value,color=variable))+
  geom_point()+
  facet_wrap(~Genotype)


ToPlot_NSR <- subset(ToPlot_W_NSR,as.numeric(as.character(Leaf)) >6  & as.numeric(as.character(Leaf)) <15 )
R2 <- round(summary(lm(ToPlot_NSR,formula = 'Sim~Obs'))$r.squared,digits=3)
RMSE <- sqrt(sum((ToPlot_NSR$Sim-ToPlot_NSR$Obs)^2,na.rm=T)*(1/nrow(ToPlot_NSR)))

ToPlot_NSR$Genotype <- droplevels(ToPlot_NSR$Genotype)
ToPlot_NSR$Leaf <- droplevels(ToPlot_NSR$Leaf)

ggplot(data=ToPlot_NSR,aes(x=Sim,y=Obs,color=Genotype,fill=Genotype))+
  geom_point(size=3)+
  scale_x_continuous(limits=c(4,10))+
  scale_y_continuous(limits=c(4,10))+
  geom_abline(slope=1,intercept=0,color='black')+
  geom_text(aes(x=7,y=5,label=paste0('R² = ',round(R2,digits=2))),color='black')+
  geom_smooth(method='lm',color='black',fill='black',se=F)+
  theme_light()+
  theme(legend.position='bottom',
        legend.box = "vertical")+
  labs(x='Simulated width (cm)')+
  labs(y='Observed width (cm)')
  

# By Geno :
ggplot(data=subset(ToPlot_NSR,Genotype!='B73_H'),aes(x=Sim,y=Obs,color=Genotype,fill=Genotype))+
  geom_point(size=3)+
  geom_point(data=subset(ToPlot,Genotype=='B73_H'),aes(x=Sim,y=Obs),shape=21,color='black',size=3.5)+
  scale_x_continuous(limits=c(3,11))+
  scale_y_continuous(limits=c(3,11))+
  geom_abline(slope=1,intercept=0,color='black')+
  geom_text(aes(x=7,y=4.5,label=paste0('R² = ',round(R2,digits=2))),color='black')+
  geom_text(aes(x=7,y=4,label=paste0('RMSE = ',round(RMSE,digits=2),' cm')),color='black')+
  geom_smooth(method='lm',color='black',fill='black',se=F)+
  theme_light()+
  theme(legend.position='bottom',
        legend.box = "vertical")+
  labs(x='Simulated width (cm)')+
  labs(y='Observed width (cm)')+
  guides(color=FALSE)
  
# By Leaf number :
ggplot(data=subset(ToPlot_NSR,Genotype!='B73_H'),aes(x=Sim,y=Obs,color=factor(Leaf),fill=factor(Leaf)))+
  geom_point(size=3)+
  geom_point(data=subset(ToPlot,Genotype=='B73_H'),aes(x=Sim,y=Obs,fill=factor(Leaf)),shape=21,color='black',size=3.5)+
  scale_x_continuous(limits=c(3,11))+
  scale_y_continuous(limits=c(3,11))+
  geom_abline(slope=1,intercept=0,color='black')+
  geom_text(aes(x=7,y=4.5,label=paste0('R² = ',round(R2,digits=2))),color='black')+
  geom_text(aes(x=7,y=4,label=paste0('RMSE = ',round(RMSE,digits=2),' cm')),color='black')+
  geom_smooth(method='lm',color='black',fill='black',se=F)+
  theme_light()+
  theme(legend.position='bottom',
        legend.box = "vertical")+
  labs(x='Simulated width (cm)')+
  labs(y='Observed width (cm)')+
  guides(color=FALSE)
  

```

```{r Better with or without Genotypic variability ?, echo=FALSE, fig.height=5, fig.width=6, message=FALSE, warning=FALSE}


WithSensiRAD <- subset(ToPlot,select=c('Genotype','Leaf','Sim','Obs'))
colnames(WithSensiRAD) <- c('Genotype','Leaf','Sim_SR','Obs')
WithoutSensiRAD <- subset(ToPlot_NSR,select=c('Genotype','Leaf','Sim','Obs'))
colnames(WithoutSensiRAD) <- c('Genotype','Leaf','Sim_NSR','Obs')

ToPlot_NSRvsSR <- merge(WithoutSensiRAD,WithSensiRAD)
R2_SR <- round(summary(lm(ToPlot_NSRvsSR,formula = 'Sim_SR~Obs'))$r.squared,digits=3)
R2_NSR <- round(summary(lm(ToPlot_NSRvsSR,formula = 'Sim_NSR~Obs'))$r.squared,digits=3)
RMSE_SR <- sqrt(sum((ToPlot_NSRvsSR$Sim_SR-ToPlot_NSRvsSR$Obs)^2,na.rm=T)*(1/nrow(ToPlot_NSRvsSR)))
RMSE_NSR <- sqrt(sum((ToPlot_NSRvsSR$Sim_NSR-ToPlot_NSRvsSR$Obs)^2,na.rm=T)*(1/nrow(ToPlot_NSRvsSR)))

ggplot()+
  geom_point(data=subset(ToPlot_NSRvsSR,Genotype!='B73_H'),aes(x=Sim_NSR,y=Obs),color='grey',size=3)+
  geom_segment(data=subset(ToPlot_NSRvsSR,Genotype!='B73_H'),aes(x=Sim_NSR,y=Obs,xend=Sim_SR,yend=Obs),color='grey',
               arrow = arrow(length = unit(0.03, "npc")))+
  geom_smooth(data=subset(ToPlot_NSRvsSR,Genotype!='B73_H'),aes(x=Sim_NSR,y=Obs),method='lm',color='grey',fill='grey',se=F)+
  geom_point(data=subset(ToPlot_NSRvsSR,Genotype!='B73_H'),aes(x=Sim_SR,y=Obs),color='black',size=3)+
  geom_point(data=subset(ToPlot_NSRvsSR,Genotype!='B73_H'),aes(x=Sim_SR,y=Obs),color='black',size=3)+
  geom_point(data=subset(ToPlot_NSRvsSR,Genotype=='B73_H'),aes(x=Sim_SR,y=Obs),color='black',size=3)+
  geom_smooth(data=subset(ToPlot_NSRvsSR,Genotype!='B73_H'),aes(x=Sim_SR,y=Obs),method='lm',color='black',fill='black',se=F)+
  scale_x_continuous(limits=c(4,10))+
  scale_y_continuous(limits=c(4,10))+
  geom_abline(slope=1,intercept=0,color='black')+
  theme_light()+
  labs(x='Simulated width (cm)')+
  labs(y='Observed width (cm)')+ 
  geom_text(aes(x=7,y=4.25,label=paste0('R² = ',round(R2_NSR,digits=2))),color='grey')+
  geom_text(aes(x=7,y=4.75,label=paste0('R² = ',round(R2_SR,digits=2))),color='black')+
  geom_text(aes(x=9,y=4.75,label=paste0('RMSE = ',round(RMSE_SR,digits=2),' cm')),color='black')+
  geom_text(aes(x=9,y=4.25,label=paste0('RMSE = ',round(RMSE_NSR,digits=2),' cm')),color='grey')


```

```{r Simulation multi-genotype - No SensiVPD, eval=FALSE, message=FALSE, warning=FALSE, include=FALSE}

# Loading XML file : ----
ApsimFile<-paste(PATHS[1],'DROPS_Network_Calibration_SensiVPDcst',sep='/')
ApsimFile<-paste(ApsimFile,'apsim',sep='.')
SimFileTree<-xmlTreeParse(ApsimFile,getDTD = F , useInternalNodes = T)

# Prepare data : ----
SimToUse <- SimFileTree

MyParamsv2 <- MyParams[c(1:12)]
  
# Test variables: ----
#geno <- 'B104_H'
#var <- colnames(MyParams)[2]
Stopped_At <- 1
WorkOnErrors <- F

# 1/ Loop over genotypes : ----
Start <- Sys.time()
for (i in Stopped_At:length(unique(MyParamsv2$Hybrid_ID)))
#for (i in 2)
{
  Genotypes <- unique(MyParamsv2$Hybrid_ID)
  geno <- as.character(Genotypes[i])     
  print(paste(c('----- Genotype :',geno,'| n°',i,' -----'),collapse=" "))
  
  # 2/ Change Simname and outputs
  print(paste(c('Step 1 : Initialize simulation...'),collapse=" "))
  ParamsToUse <- subset(MyParamsv2,Hybrid_ID==as.character(geno))
  SimulationName<-paste(ParamsToUse[,'Hybrid_ID'])
  NewApsimName<-paste(c(SimulationName,"apsim"),collapse=".")
  
  # CHeck if we have values for the genotype :
  if (nrow(ParamsToUse)>0)
  {
    # Loop over genotypic parameters and change them :
    for (var in colnames(ParamsToUse[2:ncol(ParamsToUse)]))
    {
       VarNode<-getNodeSet(SimToUse, paste(c("//",var),collapse="") )
       NewValue<- ParamsToUse[,var]
       
      for (node in 1:length(VarNode))
      {
         ChangeXmlValue(VarNode[node],NewValue) 
      }
       
    }
    print(paste('==> Done !'),collapse=" ")
  
    # Change SensiRAD value to B73H <-
    VarNode<-getNodeSet(SimToUse, paste(c("//",'SensiRAD'),collapse="") )
    NewValue<- 0.62
    for (node in 1:length(VarNode))
      {
         ChangeXmlValue(VarNode[node],NewValue) 
      }
       
    # Save the new apsimfile in Model Directory :
    print(paste(c('Step 2 : Saving simulation file...'),collapse=" "))
    saveXML(xmlRoot(SimFileTree), 
            file= paste0(PATH.APSIM,NewApsimName), 
            prefix= '<?xml version="1.0" encoding="utf-8"?>',indent= T)
    print(paste('==> Done !'),collapse=" ")
  
    # Simulate :
    print(paste(c('Step 3 : Simulation...'),collapse=" "))
    setwd(PATH.APSIM)
    
    command<-paste("apsim",NewApsimName)

    if (WorkOnErrors==T)
    {
      result <- tryCatch(evalWithTimeout(system(command,show.output.on.console=T),
                                 timeout = 30,
                                 onTimeout = "error"),
                 error=
                   {
                        #Get all tasks running :
                        Tasks <- system('tasklist',show.output.on.console=T,intern=T)
                        # Kill jobrunner :
                        while (sum(grepl("JobRunner.exe", Tasks))>0)
                        {  
                          print('Killing JobRunner...')
                          system('taskkill /IM JobRunner.exe /F',
                                 show.output.on.console=F,intern=T)
                          Tasks <- system('tasklist',show.output.on.console=T,intern=T)
                        }
                        # Kill all Apsim.exe :
                        while (sum(grepl("Apsim.exe", Tasks))>0)
                        {
                          print('Killing APSIM...')
                          system('taskkill /IM Apsim.exe /F',
                                 show.output.on.console=F,intern=T)       
                          Tasks <- system('tasklist',show.output.on.console=T,intern=T)
                        }
                    }
                 )
      
      # If it runs smoothly :
      if (result==0) {print(paste('==> Done !'),collapse=" ")
      # If there was an error : 
      } else {print(paste('==> Error Handled !'),collapse=" ")}
    }
    else {system(command)}
      
    
    # Get/Save outputs  : 
    print(paste(c('Step 4 : Loading outputs...'),collapse=" "))
    
    Output_Sim <- loadApsim(PATH.APSIM, 
                            loadAll = TRUE, 
                            ext = ".out", 
                            returnFrame = TRUE, 
                            n = 0, fill = TRUE, 
                            addConstants = TRUE)
    
    Output_Sim$Genotype<-geno
  
    if (i ==1)
    {Global_Output_SensiVPDcst <- Output_Sim
    }else{
      Global_Output_SensiVPDcst <- rbind(Global_Output_SensiVPDcst,Output_Sim)
    }
    rm(Output_Sim)
  }
  
  print(paste('==> Done !'),collapse=" ")  
  
}

Stop <- Sys.time()
TimeDiff <- difftime(Stop, Start, units="secs")
MyTimeDiff <- format(.POSIXct(TimeDiff,tz="GMT"), "%H:%M:%S")
print(paste(c(''),collapse=" "))
print(paste(c('----- Simulation loops finished -----'),collapse=" "))
print(paste(c(''),collapse=" "))
print(paste(c('Time of execution is '),MyTimeDiff,collapse=" "))


# 2/ Save and treat outputs : ----
save(file=paste0(PATHS[4],'/Global_Output_SensiVPDcst.RData'),Global_Output_SensiVPDcst)


# 3/ Estimate Harvest Yield and stuff : ---
load(paste0(PATHS[4],'/Global_Output_SensiVPDcst.RData'))
HarvestDataOutputs <- Get_Harvest_Data(Global_Output_SensiVPDcst,list(Global_Output_SensiVPDcst$Title,Global_Output_SensiVPDcst$Genotype))
HarvestData <- HarvestDataOutputs[!duplicated(HarvestDataOutputs[,c(64,70)]),]


HarvestData_Melted <-  melt(HarvestData, measure.vars = c('leafwidth(1)','leafwidth(2)','leafwidth(3)','leafwidth(4)','leafwidth(5)',
                                  'leafwidth(6)','leafwidth(7)','leafwidth(8)','leafwidth(9)','leafwidth(10)',
                                  'leafwidth(11)','leafwidth(12)','leafwidth(13)','leafwidth(14)','leafwidth(15)','leafwidth(16)'))

Width_NSV <- subset(HarvestData_Melted,select=c("Title",'Genotype','DaysAfterSowing',
                                             'yield','GrainNo',"variable","value"))

colnames(Width_NSV) <- c('Site_Year','Genotype','Das','Yield','GrainNo','Leaf','Width')
levels(Width_NSV$Leaf) <- seq(1,16,1)

HarvestData_Melted <-  melt(HarvestData, measure.vars = c('leaflength(1)','leaflength(2)','leaflength(3)','leaflength(4)','leaflength(5)',
                                  'leaflength(6)','leaflength(7)','leaflength(8)','leaflength(9)',
                                  'leaflength(10)','leaflength(11)','leaflength(12)','leaflength(13)',
                                  'leaflength(14)','leaflength(15)','leaflength(16)'))

Length_NSV <- subset(HarvestData_Melted,select=c("Title",'Genotype','DaysAfterSowing',
                                             'yield','GrainNo',"variable","value"))

colnames(Length_NSV) <- c('Site_Year','Genotype','Das','Yield','GrainNo','Leaf','Length')
levels(Length_NSV$Leaf) <- seq(1,16,1)

Merged_Outputs_NSV <- merge(Width_NSV,Length_NSV)

#Aggregate leaf variables : 
MeanLeafVars_NSV <- aggregate(data=Merged_Outputs_NSV,cbind(Width,Length)~Genotype+Leaf+Site_Year,FUN=mean,na.rm=T)

```

```{r Compare Sim/Obs - Length - Mau16 - No SensiVPD, echo=FALSE, fig.height=6.5, fig.width=6.5, message=FALSE, warning=FALSE}


SimsL_NSV <- subset(MeanLeafVars_NSV,Site_Year==' Mauguio_2016',select=c('Genotype','Leaf','Length'))
#SimsL_NSV$Leaf <- as.numeric(SimsL$Leaf)

# Lets put stuff together : 
Obs_L_NSV <- Obs_L
Obs_L_NSV$Genotype <- as.factor(Obs_L_NSV$Genotype)
Obs_L_NSV$Leaf <- as.factor(Obs_L_NSV$Leaf)
SimsL_NSV$Genotype <- as.factor(SimsL_NSV$Genotype)
SimsL_NSV <- aggregate(data=SimsL_NSV,Length~Leaf+Genotype,mean,na.rm=T)

ToPlot_L_NSV <- merge(Obs_L_NSV,SimsL_NSV,by=c('Genotype','Leaf'),all.x=T,all.y=F)
ToPlot_L_NSV$Delta <- (ToPlot_L_NSV$Length.x - ToPlot_L_NSV$Length.y)*(2/5)
ToPlot_L_NSV$Sim <- ToPlot_L_NSV$Length.y + ToPlot_L_NSV$Delta
ToPlot_L_NSV$Obs <- ToPlot_L_NSV$Length.x

ggplot(data=ToPlot_W_melted_NSV,aes(x=Leaf,y=value,color=variable))+
  geom_point()+
  facet_wrap(~Genotype)


ToPlot_NSV <- subset(ToPlot_L_NSV,as.numeric(as.character(Leaf)) >6  & as.numeric(as.character(Leaf)) <15 )
R2 <- round(summary(lm(ToPlot_NSV,formula = 'Sim~Obs'))$r.squared,digits=3)
RMSE <- sqrt(sum((ToPlot_NSV$Sim-ToPlot_NSV$Obs)^2,na.rm=T)*(1/nrow(ToPlot_NSV)))

ToPlot_NSV$Genotype <- droplevels(ToPlot_NSV$Genotype)
ToPlot_NSV$Leaf <- droplevels(ToPlot_NSV$Leaf)

ggplot(data=ToPlot_NSV,aes(x=Sim,y=Obs,color=Genotype,fill=Genotype))+
  geom_point(size=3)+
  scale_x_continuous(limits=c(40,115))+
  scale_y_continuous(limits=c(40,115))+
  geom_abline(slope=1,intercept=0,color='black')+
  geom_text(aes(x=7,y=5,label=paste0('R² = ',round(R2,digits=2))),color='black')+
  geom_smooth(method='lm',color='black',fill='black',se=F)+
  theme_light()+
  theme(legend.position='bottom',
        legend.box = "vertical")+
  labs(x='Simulated width (cm)')+
  labs(y='Observed width (cm)')
  

# By Geno :
ggplot(data=subset(ToPlot_NSV,Genotype!='B73_H'),aes(x=Sim,y=Obs,color=Genotype,fill=Genotype))+
  geom_point(size=3)+
  geom_point(data=subset(ToPlot,Genotype=='B73_H'),aes(x=Sim,y=Obs),shape=21,color='black',size=3.5)+
  scale_x_continuous(limits=c(40,115))+
  scale_y_continuous(limits=c(40,115))+
  geom_abline(slope=1,intercept=0,color='black')+
  geom_text(aes(x=70,y=50,label=paste0('R² = ',round(R2,digits=2))),color='black')+
  geom_text(aes(x=70,y=45,label=paste0('RMSE = ',round(RMSE,digits=2),' cm')),color='black')+
  geom_smooth(method='lm',color='black',fill='black',se=F)+
  theme_light()+
  theme(legend.position='bottom',
        legend.box = "vertical")+
  labs(x='Simulated length (cm)')+
  labs(y='Observed length (cm)')+
  guides(color=FALSE)
  
# By Leaf number :
levels(ToPlot_NSV$Leaf) <- levels(ToPlot$Leaf)
ggplot(data=subset(ToPlot_NSV,Genotype!='B73_H'),aes(x=Sim,y=Obs,color=factor(Leaf),fill=factor(Leaf)))+
  geom_point(size=3)+
  geom_point(data=subset(ToPlot,Genotype=='B73_H'),aes(x=Sim,y=Obs,fill=factor(Leaf)),shape=21,color='black',size=3.5)+
  scale_x_continuous(limits=c(40,115))+
  scale_y_continuous(limits=c(40,115))+
  geom_abline(slope=1,intercept=0,color='black')+
  geom_text(aes(x=70,y=50,label=paste0('R² = ',round(R2,digits=2))),color='black')+
  geom_text(aes(x=70,y=45,label=paste0('RMSE = ',round(RMSE,digits=2),' cm')),color='black')+
  geom_smooth(method='lm',color='black',fill='black',se=F)+
  theme_light()+
  theme(legend.position='bottom',
        legend.box = "vertical")+
  labs(x='Simulated length (cm)')+
  labs(y='Observed length (cm)')+
  guides(color=FALSE)
  

```

```{r Better with or without Genotypic variability ?, echo=FALSE, fig.height=5, fig.width=6, message=FALSE, warning=FALSE}


WithSensiVPD <- subset(ToPlot,select=c('Genotype','Leaf','Sim','Obs'))
colnames(WithSensiVPD) <- c('Genotype','Leaf','Sim_SV','Obs')
WithoutSensiVPD <- subset(ToPlot_NSV,select=c('Genotype','Leaf','Sim','Obs'))
colnames(WithoutSensiVPD) <- c('Genotype','Leaf','Sim_NSV','Obs')

ToPlot_NSVvsSV <- merge(WithoutSensiVPD,WithSensiVPD)
R2_SV <- round(summary(lm(ToPlot_NSVvsSV,formula = 'Sim_SV~Obs'))$r.squared,digits=3)
R2_NSV <- round(summary(lm(ToPlot_NSVvsSV,formula = 'Sim_NSV~Obs'))$r.squared,digits=3)
RMSE_SV <- sqrt(sum((ToPlot_NSVvsSV$Sim_SV-ToPlot_NSVvsSV$Obs)^2,na.rm=T)*(1/nrow(ToPlot_NSVvsSV)))
RMSE_NSV <- sqrt(sum((ToPlot_NSVvsSV$Sim_NSV-ToPlot_NSVvsSV$Obs)^2,na.rm=T)*(1/nrow(ToPlot_NSVvsSV)))

ggplot()+
  geom_point(data=subset(ToPlot_NSVvsSV,Genotype!='B73_H'),aes(x=Sim_NSV,y=Obs),color='grey',size=3)+
  geom_segment(data=subset(ToPlot_NSVvsSV,Genotype!='B73_H'),aes(x=Sim_NSV,y=Obs,xend=Sim_SV,yend=Obs),color='grey',
               arrow = arrow(length = unit(0.03, "npc")))+
  geom_smooth(data=subset(ToPlot_NSVvsSV,Genotype!='B73_H'),aes(x=Sim_NSV,y=Obs),method='lm',color='grey',fill='grey',se=F)+
  geom_point(data=subset(ToPlot_NSVvsSV,Genotype!='B73_H'),aes(x=Sim_SV,y=Obs),color='black',size=3)+
  geom_point(data=subset(ToPlot_NSVvsSV,Genotype!='B73_H'),aes(x=Sim_SV,y=Obs),color='black',size=3)+
  geom_point(data=subset(ToPlot_NSVvsSV,Genotype=='B73_H'),aes(x=Sim_SV,y=Obs),color='black',size=3)+
  geom_smooth(data=subset(ToPlot_NSVvsSV,Genotype!='B73_H'),aes(x=Sim_SV,y=Obs),method='lm',color='black',fill='black',se=F)+
  scale_x_continuous(limits=c(40,115))+
  scale_y_continuous(limits=c(40,115))+
  geom_abline(slope=1,intercept=0,color='black')+
  theme_light()+
  labs(x='Simulated length (cm)')+
  labs(y='Observed length (cm)')+ 
  geom_text(aes(x=70,y=42.5,label=paste0('R² = ',round(R2_NSV,digits=2))),color='grey')+
  geom_text(aes(x=70,y=47.5,label=paste0('R² = ',round(R2_SV,digits=2))),color='black')+
  geom_text(aes(x=90,y=47.5,label=paste0('RMSE = ',round(RMSE_SV,digits=2),' cm')),color='black')+
  geom_text(aes(x=90,y=42.5,label=paste0('RMSE = ',round(RMSE_NSV,digits=2),' cm')),color='grey')


```

## Test Data in WD conditions : 

```{r Simulation multi-genotype / WD Mau16, eval=FALSE, message=FALSE, warning=FALSE, include=FALSE}

# Loading XML file : ----
ApsimFile<-paste(PATHS[1],'DROPS_Network_Calibration_WD_MA16',sep='/')
ApsimFile<-paste(ApsimFile,'apsim',sep='.')
SimFileTree<-xmlTreeParse(ApsimFile,getDTD = F , useInternalNodes = T)

# Prepare data : ----
SimToUse <- SimFileTree

# Test variables: ----
#geno <- 'B104_H'
#var <- colnames(MyParams)[2]
Stopped_At <- 1
WorkOnErrors <- F

# 1/ Loop over genotypes : ----
Start <- Sys.time()
for (i in Stopped_At:length(unique(MyParams$Hybrid_ID)))
#for (i in 2)
{
  Genotypes <- unique(MyParams$Hybrid_ID)
  geno <- as.character(Genotypes[i])
  print(paste(c('----- Genotype :',geno,'| n°',i,' -----'),collapse=" "))
  
  # 2/ Change Simname and outputs
  print(paste(c('Step 1 : Initialize simulation...'),collapse=" "))
  ParamsToUse <- subset(MyParams,Hybrid_ID==as.character(geno))
  SimulationName<-paste(ParamsToUse[,'Hybrid_ID'])
  NewApsimName<-paste(c(SimulationName,"apsim"),collapse=".")
  
  # CHeck if we have values for the genotype :
  if (nrow(ParamsToUse)>0)
  {
    # Loop over genotypic parameters and change them :
    for (var in colnames(ParamsToUse[2:ncol(ParamsToUse)]))
    {
       VarNode<-getNodeSet(SimToUse, paste(c("//",var),collapse="") )
       NewValue<- ParamsToUse[,var]
       
      for (node in 1:length(VarNode))
      {
         ChangeXmlValue(VarNode[node],NewValue) 
      }
       
    }
    print(paste('==> Done !'),collapse=" ")
  
    # Save the new apsimfile in Model Directory :
    print(paste(c('Step 2 : Saving simulation file...'),collapse=" "))
    saveXML(xmlRoot(SimFileTree), 
            file= paste0(PATH.APSIM,NewApsimName), 
            prefix= '<?xml version="1.0" encoding="utf-8"?>',indent= T)
    print(paste('==> Done !'),collapse=" ")
  
    # Simulate :
    print(paste(c('Step 3 : Simulation...'),collapse=" "))
    setwd(PATH.APSIM)
    
    command<-paste("apsim",NewApsimName)

    if (WorkOnErrors==T)
    {
      result <- tryCatch(evalWithTimeout(system(command,show.output.on.console=T),
                                 timeout = 30,
                                 onTimeout = "error"),
                 error=
                   {
                        #Get all tasks running :
                        Tasks <- system('tasklist',show.output.on.console=T,intern=T)
                        # Kill jobrunner :
                        while (sum(grepl("JobRunner.exe", Tasks))>0)
                        {  
                          print('Killing JobRunner...')
                          system('taskkill /IM JobRunner.exe /F',
                                 show.output.on.console=F,intern=T)
                          Tasks <- system('tasklist',show.output.on.console=T,intern=T)
                        }
                        # Kill all Apsim.exe :
                        while (sum(grepl("Apsim.exe", Tasks))>0)
                        {
                          print('Killing APSIM...')
                          system('taskkill /IM Apsim.exe /F',
                                 show.output.on.console=F,intern=T)       
                          Tasks <- system('tasklist',show.output.on.console=T,intern=T)
                        }
                    }
                 )
      
      # If it runs smoothly :
      if (result==0) {print(paste('==> Done !'),collapse=" ")
      # If there was an error : 
      } else {print(paste('==> Error Handled !'),collapse=" ")}
    }
    else {system(command)}
      
    
    # Get/Save outputs  : 
    print(paste(c('Step 4 : Loading outputs...'),collapse=" "))
    
    Output_Sim <- loadApsim(PATH.APSIM, 
                            loadAll = TRUE, 
                            ext = ".out", 
                            returnFrame = TRUE, 
                            n = 0, fill = TRUE, 
                            addConstants = TRUE)
    
    Output_Sim$Genotype<-geno
  
    if (i ==1)
    {Global_Output_WD <- Output_Sim
    }else{
      Global_Output_WD <- rbind(Global_Output_WD,Output_Sim)
    }
    rm(Output_Sim)
  }
  
  print(paste('==> Done !'),collapse=" ")  
  
}

Stop <- Sys.time()
TimeDiff <- difftime(Stop, Start, units="secs")
MyTimeDiff <- format(.POSIXct(TimeDiff,tz="GMT"), "%H:%M:%S")
print(paste(c(''),collapse=" "))
print(paste(c('----- Simulation loops finished -----'),collapse=" "))
print(paste(c(''),collapse=" "))
print(paste(c('Time of execution is '),MyTimeDiff,collapse=" "))

```

```{r Treat sim / WD Mau16, eval=FALSE, message=FALSE, warning=FALSE, include=FALSE}

# 2/ Save/load and treat outputs : ----

#save(file=paste0(PATHS[4],'/Global_Output_WD.RData'),Global_Output_WD)

Obs_L_WD <- subset(Mauguio_2016_PerLeaf,Treatment=='WD',select=c('Genotype','Leaf','Length'))
Obs_W_WD <- subset(Mauguio_2016_PerLeaf,Treatment=='WD',select=c('Genotype','Leaf','Width'))
Obs_W_WD$Width <- Obs_W_WD$Width/10


# 3/ Estimate Harvest Yield and stuff : ---
load(paste0(PATHS[4],'/Global_Output_WD.RData'))
HarvestDataOutputs <- Get_Harvest_Data(Global_Output_WD,list(Global_Output_WD$Title,Global_Output_WD$Genotype))

HarvestData_Melted <-  melt(HarvestDataOutputs, measure.vars = c('leafwidth(1)','leafwidth(2)','leafwidth(3)','leafwidth(4)','leafwidth(5)',
                                  'leafwidth(6)','leafwidth(7)','leafwidth(8)','leafwidth(9)','leafwidth(10)',
                                  'leafwidth(11)','leafwidth(12)','leafwidth(13)','leafwidth(14)','leafwidth(15)','leafwidth(16)'))

Width <- subset(HarvestData_Melted,select=c("Title",'Genotype','DaysAfterSowing',
                                             'yield','GrainNo',"variable","value"))

colnames(Width) <- c('Site_Year','Genotype','Das','Yield','GrainNo','Leaf','Width')
levels(Width$Leaf) <- seq(1,16,1)

HarvestData_Melted <-  melt(HarvestDataOutputs, measure.vars = c('leaflength(1)','leaflength(2)','leaflength(3)','leaflength(4)','leaflength(5)',
                                  'leaflength(6)','leaflength(7)','leaflength(8)','leaflength(9)',
                                  'leaflength(10)','leaflength(11)','leaflength(12)','leaflength(13)',
                                  'leaflength(14)','leaflength(15)','leafwidth(16)'))

Length <- subset(HarvestData_Melted,select=c("Title",'Genotype','DaysAfterSowing',
                                             'yield','GrainNo',"variable","value"))

colnames(Length) <- c('Site_Year','Genotype','Das','Yield','GrainNo','Leaf','Length')
levels(Length$Leaf) <- seq(1,16,1)

Merged_Outputs <- merge(Width,Length)


#Aggregate leaf variables : 
MeanLeafVars <- aggregate(data=Merged_Outputs,cbind(Width,Length)~Genotype+Leaf+Site_Year,FUN=mean,na.rm=T)
  
SimsW_WD <- subset(MeanLeafVars,select=c('Genotype','Leaf','Width'))
SimsW_WD$Leaf <- as.numeric(SimsW_WD$Leaf)
SimsL_WD <- subset(MeanLeafVars,select=c('Genotype','Leaf','Length'))
SimsL_WD$Leaf <- as.numeric(SimsL_WD$Leaf)


```

```{r Show sim vs obs for WD Mau16, eval=FALSE, message=FALSE, warning=FALSE, include=FALSE}


# Lets put stuff together : 
Obs_W_WD$Genotype <- as.factor(Obs_W_WD$Genotype)
Obs_W_WD$Leaf <- as.factor(Obs_W_WD$Leaf)
SimsW_WD$Genotype <- as.factor(SimsW_WD$Genotype)
SimsW_WD$Leaf <- as.factor(SimsW_WD$Leaf)
SimsW_WD <- aggregate(data=SimsW_WD,Width~Leaf+Genotype,mean,na.rm=T)

ToPlot_W <- merge(Obs_W_WD,SimsW_WD,by=c('Genotype','Leaf'),all.x=T,all.y=F)

ToPlot_W$Delta <- (ToPlot_W$Width.x - ToPlot_W$Width.y)*1/5
ToPlot_W$Sim <- ToPlot_W$Width.y + ToPlot_W$Delta
ToPlot_W$Obs <- ToPlot_W$Width.x

ToPlot_W_melted <- melt(ToPlot_W,measure.vars=c('Sim','Obs'))

ggplot(data=ToPlot_W_melted,aes(x=Leaf,y=value,color=variable))+
  geom_point()+
  facet_wrap(~Genotype)

```

## Yield and Grain number simulation :

```{r Treat Data -  GN / GY, echo=FALSE, fig.height=5, fig.width=5, message=FALSE, warning=FALSE}


# Obs : 
# ------

## GN :
Measured_GN15 <- read.table(paste0(PATHS[3],'/BLUES_KNO.csv'), 
                       sep=";",                                           
                       header=TRUE,dec=".",
                       row.names=NULL,
                       as.is=FALSE, 
                       fill=TRUE)

# Separate data to fit the one we already have (in lines not columns...)
Measured_GN <- gather(Measured_GN15,one_of(colnames(Measured_GN15)[5:33]),key = 'Site',value = 'KNO')
Measured_GN_v2 <- separate(Measured_GN,Site,sep="_",into=c('Site','Year_Scenario'),extra='merge')
Measured_GN_v3 <- separate(Measured_GN_v2,Year_Scenario,sep="_",into=c('Year','Scenario'),extra='merge')
Measured_GN_v4 <- separate(Measured_GN_v3,Site,sep="[.]",into=c('Var','Site'),extra='merge')
Measured_GN_v5 <- subset(Measured_GN_v4,Scenario=="watered",select=c('Hybrid_ID','Site','Year','KNO'))
Measured_Data_GN <- unite(Measured_GN_v5,Site,Year,col="Title",sep='_')
Measured_Data_GN$Title <- factor(Measured_Data_GN$Title)
levels(Measured_Data_GN$Title) <- c('Martonvasar_2013','Gaillac_2012','Gaillac_2013','Graneros_2013',
                                 'Nerac_2012','Nerac_2013','Caracal_2012','Karlsruhe_2012',
                                 'Karlsruhe_2013','Murony_2013','Karacal_2012',
                                 'Campagnola_2013','Cadriano_2012')
colnames(Measured_Data_GN) <- c("Genotype","Title","M.GrainNo")  

## GN :
Measured_GY15 <- read.table(paste0(PATHS[3],'/BLUES_GY15.csv'), 
                       sep=",",                                           
                       header=TRUE,dec=".",
                       row.names=NULL,  
                       as.is=FALSE, 
                       fill=TRUE)

# Separate data to fit the one we already have (in lines not columns...)
Measured_GY <- unite(
                   subset(
                     separate(
                        separate(
                          separate(
                            gather(Measured_GY15,
                                   one_of(colnames(Measured_GY15)[5:33]),key = 'Site',value ='Yield'),
                            Site,sep="_",into=c('Site','Year_Scenario'),extra='merge'),
                        Year_Scenario,sep="_",into=c('Year','Scenario'),extra='merge'),
                     Site,sep="[.]",into=c('Var','Site'),extra='merge'),
                   Scenario=="watered",select=c('Hybrid_ID','Site','Year','Yield')),
               Site,Year,col="Title",sep='_')

Measured_GY$Title <- factor(Measured_GY$Title)


levels(Measured_GY$Title) <- c('Martonvasar_2013','Gaillac_2012','Gaillac_2013','Graneros_2013',
                                 'Nerac_2012',' Nerac_2013','Caracal_2012','Karlsruhe_2012',
                                 'Karlsruhe_2013','Murony_2013','Karacal_2012',
                                 'Campagnola_2012','Campagnola_2013','Cadriano_2012')
colnames(Measured_GY) <- c("Genotype","Title","M.Yield") 



MeasuredYield <- merge(Measured_Data_GN,Measured_GY,by=c('Genotype','Title'))
MeasuredYield$M.Yield.kgha <- MeasuredYield$M.Yield *1000

# -----------------------------------

# Sim :
# ------
load(paste0(PATHS[4],'/Global_Output.RData'))
HarvestDataOutputs <- Get_Harvest_Data(Global_Output,list(Global_Output$Title,Global_Output$Genotype))
HarvestData <- HarvestDataOutputs[!duplicated(HarvestDataOutputs[,c(64,70)]),]
Network_GN_GY <- select(HarvestData,Title,Genotype,GrainNo,yield)
Network_GN_GY$Title  <- gsub(' ', '', Network_GN_GY$Title)

MeasuredYield$Title <- as.character(MeasuredYield$Title)
MeasuredYield$Titlev2 <- gsub(' ', '', MeasuredYield$Title)

MeasuredYield$Genotype <- as.character(MeasuredYield$Genotype)

Data_GY_GN <- merge(MeasuredYield,Network_GN_GY,by=c('Genotype','Title'))



  

```

```{r Show Sim/Obs GN/GY - B73, echo=FALSE, fig.height=5, fig.width=5, message=FALSE, warning=FALSE}

ToPlot <- subset(Data_GY_GN,Genotype=='B73_H')


R2 <- round(summary(lm(ToPlot,formula = 'M.GrainNo~GrainNo'))$r.squared,digits=3)
RMSE <- sqrt(sum((ToPlot$M.GrainNo-ToPlot$GrainNo)^2,na.rm=T)*(1/nrow(ToPlot)))

ggplot(data=ToPlot,aes(x=M.GrainNo,y=GrainNo,label=Title))+
  geom_point(size=2)+
  geom_text_repel(size=3.5)+
  geom_smooth(method='lm',se=F)+
  geom_abline(slope=1,intercept=0,color='black')+
  scale_x_continuous(limits=c(1000,6000))+
  scale_y_continuous(limits=c(1000,6000))+
  geom_text(aes(x=2500,y=2000,label=paste0('R² = ',round(R2,digits=2))),color='black')+
  geom_text(aes(x=2500,y=1500,label=paste0('RMSE = ',round(RMSE,digits=2),' gr/m²')),color='black')+
  theme_light()+
  labs(x='Measured grain number (gr/m²)',y='Simulated Grain number (gr/m²)')+
  ggtitle('Genotype B73 - DROPS network')


R2 <- round(summary(lm(ToPlot,formula = 'M.Yield~yield'))$r.squared,digits=3)
RMSE <- sqrt(sum((ToPlot$M.Yield-(ToPlot$yield/1000))^2,na.rm=T)*(1/nrow(ToPlot)))

ggplot(data=ToPlot,aes(x=M.Yield,y=yield/1000,label=Title))+
  geom_point(size=2)+
  geom_text_repel(size=3.5)+
  geom_smooth(method='lm',se=F)+
  geom_abline(slope=1,intercept=0,color='black')+
  scale_x_continuous(limits=c(0,20))+
  scale_y_continuous(limits=c(0,20))+
  geom_text(aes(x=6,y=6,label=paste0('R² = ',round(R2,digits=2))),color='black')+
  geom_text(aes(x=6,y=5,label=paste0('RMSE = ',round(RMSE,digits=2),' t/ha')),color='black')+
  theme_light()+
  labs(x='Measured grain yield (t/ha)',y='Simulated grain yield (t/ha)')+
  ggtitle('Genotype B73 - DROPS network')

R2 <- round(summary(lm(ToPlot,formula = 'Sim~Obs'))$r.squared,digits=3)
RMSE <- sqrt(sum((ToPlot$Sim-ToPlot$Obs)^2,na.rm=T)*(1/nrow(ToPlot)))

```

```{r Show Sim/Obs GN/GY - 16 Hybrids, echo=FALSE, fig.height=5, fig.width=5, message=FALSE, warning=FALSE}


ggplot(data=Data_GY_GN,aes(x=M.GrainNo,y=GrainNo,label=Title))+
  geom_point(size=2)+
  geom_smooth(method='lm',se=F)+
  geom_abline(slope=1,intercept=0,color='black')+
  scale_x_continuous(limits=c(1000,6000))+
  scale_y_continuous(limits=c(1000,6000))+
  theme_light()+
  facet_wrap(~Genotype)+
  labs(x='Measured grain number (gr/m²)',y='Simulated grain number (gr/m²)')+
  ggtitle('16 hybrids - DROPS network')



```

------------' RAD effect '------------

## Calculate RAD effect DROPS panel  

```{r Load Field Data, echo=FALSE, fig.height=4, fig.width=5, message=FALSE, warning=FALSE}

# St Martin :

GWAS_template <- read.table('D:/Work/Widths/3.Data/St MArtin-ZA16/gwas_template.csv',
                            sep=";",                                           
                            header=TRUE,dec=".",
                            row.names=NULL)


Meas_StMartin <- read.table('D:/Work/Widths/3.Data/St MArtin-ZA16/Measurements_LW_StMartin_vCorr.csv',
                            sep=";",                                           
                            header=TRUE,dec=".",
                            row.names=NULL)

Meas_StMartin$Length <- as.numeric(as.character(Meas_StMartin$Length))
Meas_StMartin$Width <- as.numeric(as.character(Meas_StMartin$Width))


FinalDataset_StMartin <- merge(GWAS_template,Meas_StMartin,by=c('Code_ID'),all.x=T,all.y=F)
ToTreat_St_Martin <- aggregate(cbind(Length,Width)~Leaf+Hybrid_ID,
                                             FinalDataset_StMartin,
                                             FUN = mean,
                                             na.rm=T)
ToTreat_St_Martin$Scenario <- 'WW'
ToTreat_St_Martin$Environment <- 'Platform'

```

```{r Load Platform Data, echo=FALSE, fig.height=4, fig.width=5, message=FALSE, warning=FALSE}

# ZA16 : 
ZA16 <- read.table('D:/Work/Widths/3.Data/St MArtin-ZA16/For GWAS/Platform_For_GWAS_Seb.csv',
            sep=";",
            header=TRUE,dec=".",
            row.names=NULL)

ToTreat_ZA16 <- melt(ZA16,measure.vars =c("L7.WD","L8.WD","L9.WD","L10.WD","Lmean_78.WD",
                                                  "W7.WD","W8.WD","W9.WD","W10.WD","Wmean_78.WD",
                                                  "L7.WW","L8.WW","L9.WW","L10.WW","Lmean_78.WW",
                                                  "W7.WW","W8.WW","W9.WW","W10.WW","Wmean_78.WW"))

ToTreat_ZA16$Variable <- apply(ToTreat_ZA16,1, function (x) {strsplit(as.character(x[[19]]),".",fixed=T)[[1]][1]})
ToTreat_ZA16$Scenario <- apply(ToTreat_ZA16,1, function (x) {strsplit(as.character(x[[19]]),".",fixed=T)[[1]][2]})
ToTreat_ZA16_NoMean <- subset(ToTreat_ZA16,Variable!="Wmean_78" & Variable!="Lmean_78")
ToTreat_ZA16_NoMean$Leaf <- substr(as.character(ToTreat_ZA16_NoMean$Variable), 2, nchar(ToTreat_ZA16_NoMean$Variable))
ToTreat_ZA16_NoMean$Variable <- substr(as.character(ToTreat_ZA16_NoMean$Variable), 1, 1)
ToTreat_ZA16_NoMean$Variable <- as.factor(ToTreat_ZA16_NoMean$Variable)
levels(ToTreat_ZA16_NoMean$Variable) <- c('Length','Width')
ToTreat_ZA16_NoMean$Leaf <- as.numeric(ToTreat_ZA16_NoMean$Leaf)

ZA16_Data <- subset(ToTreat_ZA16_NoMean,select=c('Hybrid_ID','value','Variable','Scenario','Leaf'))
colnames(ZA16_Data) <- c('Hybrid_ID','value','Variable','Scenario','Leaf')

ZA16_Width <- subset(ZA16_Data,Variable=='Width')
ZA16_Length <- subset(ZA16_Data,Variable=='Length')

ZA16_1 <- merge(ZA16_Width,ZA16_Length,by=c('Hybrid_ID','Leaf','Scenario'))
colnames(ZA16_1) <- c('Hybrid_ID','Leaf','Scenario','Width','Variable.x','Length','Variable.y')
ZA16_ToTreat <- subset(ZA16_1,select=c('Leaf','Hybrid_ID','Length','Width','Scenario'))
ZA16_ToTreat$Environment <- 'Platform'
ZA16_ToTreat_Mean <- aggregate(cbind(Length,Width)~Leaf+Hybrid_ID,ZA16_ToTreat,FUN = mean,na.rm=T)
ZA16_ToTreat_Mean$Scenario <- 'WW'
ZA16_ToTreat_Mean$Environment <- 'Platform'

```

```{r Load Platform Meteo, echo=FALSE, fig.height=4, fig.width=5, message=FALSE, warning=FALSE}

## Get the meteo data : 


PlatMeteo <- read.table('D:/Work/Widths/3.Data/ARCH measurements/Meteo/Meteo_ZA16.csv',
                            sep=";",                                           
                            header=TRUE,dec=".",
                            row.names=NULL)


PlatMeteo$RADi <-  ( (PlatMeteo$par.moy*15*60)/(4.57*0.48) ) * ( 10^-6 )
MyYear <-year(as.Date(PlatMeteo$aaaammdd[1],format='%d/%m/%Y'))
FirstDate <- unique(PlatMeteo$date)[1]
LastDate <- unique(PlatMeteo$date)[length( unique(PlatMeteo$date))]

OutTHT <- ThermalTime('Apsim',PlatMeteo$date,PlatMeteo$tc.moy)
  
DailySum <- aggregate(RADi~date,PlatMeteo,FUN = sum)
DailyMean <- aggregate(cbind(vpd.air,hr.moy,tc.moy,t.air.moy,par.moy)~date,PlatMeteo,FUN = mean)
DailyMax <- aggregate(cbind(vpd.air,hr.moy,tc.moy,t.air.moy,par.moy)~date,PlatMeteo,FUN = max)
DailyMin <- aggregate(cbind(vpd.air,hr.moy,tc.moy,t.air.moy,par.moy)~date,PlatMeteo,FUN = min)
  
PlatMeteoDaily <- data.frame(date=DailySum$date,
                                    cumTht=OutTHT$cumtht,Tht=OutTHT$tht,
                                    Tmean=DailyMean$tc.moy,Tmax=DailyMax$tc.moy,Tmin=DailyMin$tc.moy,
                                    VPDmean=DailyMean$vpd.air,VPDmax=DailyMax$vpd.air,
                                    RADinc=DailySum$RADi,
                                    RHmin=DailyMin$hr.moy,RHmax=DailyMax$hr.moy)

LoessModel <- loess(InterceptedRAD,formula = 'Epsilon~cumTht')
PlatMeteoDaily$Predicted_Epsilon <- predict(LoessModel,data.frame(cumTht=PlatMeteoDaily$cumTht))
PlatMeteoDaily$RADint <- PlatMeteoDaily$RADinc*PlatMeteoDaily$Predicted_Epsilon
PlatMeteoDaily$RADint[which(is.na(PlatMeteoDaily$RADint))] <- 0
PlatMeteoDaily$RADrollmean <- rollmean(PlatMeteoDaily$RADint,k = 7,fill=NA,align = 'right')
PlatMeteoDaily$RADrollmean[which(PlatMeteoDaily$RADrollmean<0)] <- 0

```

```{r Load Field Meteo, echo=FALSE, fig.height=4, fig.width=5, message=FALSE, warning=FALSE}

FieldMeteo <- read.table('D:/Work/Widths/3.Data/St Martin-ZA16/MeteoClean_SMH2016-1.csv',
                            sep=",",                                           
                            header=TRUE,dec=".",
                            row.names=NULL)

FieldMeteo$RADinc <-  (FieldMeteo$RG* 10^-2)
DailySum <- aggregate(RADinc~Jour,FieldMeteo,FUN = sum)
OutTHT <- ThermalTime('Apsim',FieldMeteo$Jour,FieldMeteo$T)

FieldMeteoDaily <- data.frame(date=DailySum$Jour,
                                    cumTht=OutTHT$cumtht,Tht=OutTHT$tht,
                                    Tmean=OutTHT$MeanTemp,
                                    RADinc=DailySum$RADinc)

LoessModel <- loess(InterceptedRAD,formula = 'Epsilon~cumTht')
FieldMeteoDaily$Predicted_Epsilon <- predict(LoessModel,data.frame(cumTht=FieldMeteoDaily$cumTht))
FieldMeteoDaily$RADint <- FieldMeteoDaily$RADinc*FieldMeteoDaily$Predicted_Epsilon
FieldMeteoDaily$RADint[which(is.na(FieldMeteoDaily$RADint))] <- 0
FieldMeteoDaily$RADrollmean <- rollmean(FieldMeteoDaily$RADint,k = 7,fill=NA,align = 'right')
FieldMeteoDaily$RADrollmean[which(FieldMeteoDaily$RADrollmean<0)] <- 0


```

```{r Treat data, echo=FALSE, fig.height=4, fig.width=5, message=FALSE, warning=FALSE}

FieldData <- ToTreat_St_Martin
PlatformData <- ZA16_ToTreat_Mean
APSIM_Params <- read.table('D:/Work/Sims - DROPS/4.Out/Parameters_DROPS_Panel_ForAPSIM.csv',
                            sep=",",                                           
                            header=TRUE,dec=".",
                            row.names=NULL)

RADvsWidth <- data.frame(Geno=as.character(),
                         Leaf=as.numeric(),
                         RADint=as.numeric(),
                         Width=as.numeric(),
                         Environment=as.character())
SensiRAD <- data.frame(Geno=as.character(),
                         Leaf=as.numeric(),
                         Sensi=as.numeric(),
                         Oo=as.numeric())

for (geno in unique(APSIM_Params$Hybrid_ID))
{
  
  # Get Params for this genotype (LER and WidthMin):
  Params <- subset(APSIM_Params,Hybrid_ID==geno)
  # Simulate LER model for leaf growth periods : 
  df.Dev_Stages <- Leaf_dev_stages(Params$Nfinal,Params$Leaf_tip_emerg,
                                   Params$Phyllochron,Params$tt_firstligule_sinceapp,
                                   Params$First_LiguloChrone,5.4,2*Params$Phyllochron,4.5)
  
  df.Dev_Stages$Geno <- geno
  HisWmin <- SimulateWidthAtLowRad(Params$Nfinal)

  # Platform :
  GenoPlat <- subset(PlatformData,Hybrid_ID==geno & Leaf<11 & Leaf >6)
  for (leaf in unique(GenoPlat$Leaf))
  {
    StagesTT <- subset(df.Dev_Stages,Leaf==leaf)
    ToSummarizeRAD <- subset(PlatMeteoDaily,cumTht< StagesTT$nLeafFullyExp & cumTht> StagesTT$nLeafExp)
    RADperplant <- tail(ToSummarizeRAD$RADinc, n=1)/13
    WidthData <- subset(GenoPlat,Leaf==leaf)$Width #- subset(HisWmin,Leaf==leaf)$SimWidth
    
    RADvsWidth <-rbind(RADvsWidth,data.frame(Geno=geno,
                                             Leaf=leaf,
                                             RADint=RADperplant,
                                             Width=WidthData,
                                             Environment='Platform'))
              
  }

  # Field :
  GenoField <- subset(FieldData,Hybrid_ID==geno & Leaf<11 & Leaf >6)

  for (leaf in unique(GenoField$Leaf))
  {
    StagesTT <- subset(df.Dev_Stages,Leaf==leaf)
    ToSummarizeRAD <- subset(FieldMeteoDaily,cumTht< StagesTT$nLeafFullyExp & cumTht> StagesTT$nLeafExp)
    RADperplant <- tail(ToSummarizeRAD$RADinc, n=1)/8
    WidthData <- subset(GenoField,Leaf==leaf)$Width #- subset(HisWmin,Leaf==leaf)$SimWidth
    
    RADvsWidth <-rbind(RADvsWidth,data.frame(Geno=geno,
                                             Leaf=leaf,
                                             RADint=RADperplant,
                                             Width=WidthData,
                                             Environment='Field'))


  }
  
  # Calculate sensitivity to RAD :
  LmFit <- lm('Width~RADint',subset(RADvsWidth,Geno==geno))
  SensiRAD <- rbind(SensiRAD,data.frame(Geno=geno,
                                       Leaf=leaf,
                                       Sensi=round(summary(LmFit)$coefficients[2,1],digits=2),
                                       Oo=round(summary(LmFit)$coefficients[1,1],digits=2) ))
}
 
```

```{r Genotype sensitivity, echo=FALSE, fig.height=4, fig.width=5, message=FALSE, warning=FALSE}


SensiRADgeno <- SensiRAD
SensiRADgeno <- subset(SensiRADgeno,Sensi>0)

ggplot(data=SensiRADgeno,aes(x=Sensi))+
  geom_histogram(binwidth = 0.1)+
  theme_light()+
  theme(legend.position='none')+
  labs(x='Sensitivity to Radiation (cm/MJ)',y='Frequency in panel')+
  geom_vline(xintercept = 0.24 ,color='blue')
  
ToPlot <- merge(SensiRADgeno,RADvsWidth)
WvsRAD <- ToPlot

MaxGeno <- SensiRADgeno[which(SensiRADgeno$Sensi==max(SensiRADgeno$Sensi)),][1,]
MinGeno <-  SensiRADgeno[which(SensiRADgeno$Sensi==min(SensiRADgeno$Sensi)),][1,]

ggplot(data=ToPlot,aes(x=RADint,y=Width,group=Geno))+
  #geom_point(se=F,color='darkgrey')+
  geom_smooth(data=subset(ToPlot,Geno=='B73_H'),aes(x=RADint,y=Width),color='blue',method='lm',se=F)+
  geom_abline(data=ToPlot,aes(slope=Sensi,intercept=5),color='grey',size=2.5)+
  geom_abline(data=subset(ToPlot,Geno=='B73_H'),aes(slope=Sensi,intercept=5),color='blue',size=2)+
  #geom_smooth(data=subset(ToPlot,Geno==MaxGeno$Geno),aes(x=RADint,y=Width),color='green',method='lm',se=F)+
  #geom_smooth(data=subset(ToPlot,Geno==MinGeno$Geno),aes(x=RADint,y=Width),color='red',method='lm',se=F)+
  theme_light()+
  theme(legend.position='none')+
  scale_x_continuous(expand = c(0, 0),limits=c(0,3))+
  scale_y_continuous(limits=c(0,12))+
  labs(x='Intercepted radiation (MJ/plant)',y='Width (cm)')
  

## Write results : 
write.csv(SensiRADgeno,file ='D:/Work/Sims - DROPS/3.Data/SensiRAD.csv',row.names = FALSE)


```

```{r Calculate Width vs RAD simulated, echo=FALSE, fig.height=8, fig.width=8, message=FALSE, warning=FALSE}

## Get the 15 genotypes :
load(paste0(PATHS[4],'/Global_Output.RData'))


WvsRAD_Mau16 <- data.frame(Geno=character(),Leaf=numeric(),Width=numeric(),RADi=numeric())
for (g in unique(Global_Output$Genotype))
{
  
  GenoParams <- subset(APSIM_Params,Hybrid_ID==g)
  HisWmin <- SimulateWidthAtLowRad(GenoParams$Nfinal)
  df.Dev_Stages <- Leaf_dev_stages(GenoParams$Nfinal,GenoParams$Leaf_tip_emerg,
                                   GenoParams$Phyllochron,GenoParams$tt_firstligule_sinceapp,
                                   GenoParams$First_LiguloChrone,5.4,2*GenoParams$Phyllochron,4.5)
  DataToTreat <- subset(Global_Output,Genotype==g)
  DataToTreat_Melted <-  melt(DataToTreat, measure.vars = c('leafwidth(1)','leafwidth(2)','leafwidth(3)','leafwidth(4)','leafwidth(5)',
                                    'leafwidth(6)','leafwidth(7)','leafwidth(8)','leafwidth(9)','leafwidth(10)',
                                    'leafwidth(11)','leafwidth(12)','leafwidth(13)','leafwidth(14)','leafwidth(15)','leafwidth(16)'))
  
  WidthToTreat <- subset(DataToTreat_Melted,select=c("Title",'Genotype','TT',
                                               'yield','GrainNo',"variable","value",'RADi'))
  
  colnames(WidthToTreat) <- c('Site_Year','Genotype','TT','Yield','GrainNo','Leaf','Width','RADi')
  levels(WidthToTreat$Leaf) <- seq(1,16,1)
  
  Mau16_ToTreat <- subset(WidthToTreat,Site_Year==" Mauguio_2016" & !is.na(RADi) & RADi != 0 & !is.na(Width))

  
  for (leaf in 1:GenoParams$Nfinal)
  {
    LeafTiming <- subset(df.Dev_Stages,Leaf==leaf)
    Df_ToTreat <- subset(Mau16_ToTreat,Leaf==leaf & RADi != 0 & TT< LeafTiming$nLeafFullyExp & TT> LeafTiming$nLeafExp )
    RADi <- mean(Df_ToTreat$RADi)/7.5
    Width <- max(Df_ToTreat$Width)
    Wmin <- subset(HisWmin,Leaf==leaf)$SimWidth
    WvsRAD_Mau16 <- rbind(WvsRAD_Mau16,data.frame(Geno=g,
                                                  Leaf=leaf,
                                                  Width=Width,#-Wmin,
                                                  RADi=RADi))
    
  }

}

WvsRAD_Mau16 <- subset(WvsRAD_Mau16,Width>0)

```

```{r Add Mauguio in sensitivity, echo=FALSE, fig.height=8, fig.width=8, message=FALSE, warning=FALSE}


Points_PvsF <- subset(RADvsWidth, Geno %in% unique(WvsRAD_Mau16$Geno )) 
Points_Mau16 <- subset(WvsRAD_Mau16,Leaf>9 & Leaf <13)

Points <- merge(Points_Mau16,SensiRADgeno,by=c('Geno'))
Points$SimWidth <- Points$Sensi*Points$RADi+Points$Oo
Points$Delta <- Points$SimWidth - Points$Width
Points$WidthCor <- Points$Width + (Points$Delta)*0.5
  
ggplot()+
  geom_point(data=Points_PvsF,aes(x=RADint,y=Width),color='black')+
  geom_point(data=Points,aes(x=RADi+0.5,y=WidthCor),color='red')+
  geom_abline(slope=0.75,intercept=6.23,color='grey')+
  geom_smooth(method='lm',data=Points_PvsF,aes(x=RADint,y=Width),color='blue',se=F)+
  facet_wrap(~Geno)+
  theme_light()+
  theme(legend.position='none')+
  scale_y_continuous(expand = c(0, 0),limits=c(0,12))+
  scale_x_continuous(limits=c(0,4))+
  labs(x='Intercepted radiation (MJ/plant)',y='Width (cm)')
  

```

```{r Add Mauguio in sensitivity v2, echo=FALSE, fig.height=8, fig.width=8, message=FALSE, warning=FALSE}


# Calculate var for each genotypes :
SensiRADgeno <- subset(SensiRAD,Sensi > 0 & Sensi < 1.5)
SensiRADgeno <- aggregate(cbind(Sensi,Oo)~Geno,SensiRADgeno,mean,na.rm=T)
Abline <- subset(SensiRADgeno, Geno %in% unique(WvsRAD_Mau16$Geno)) 
Abline_B73 <- subset(Abline, Geno=='B73_H') 
ToPlot <- merge(SensiRADgeno,RADvsWidth)
ToPlot <- subset(RADvsWidth, Geno %in% unique(WvsRAD_Mau16$Geno)) 
WvsRAD_Mau16$Leaf <- as.numeric(as.character(WvsRAD_Mau16$Leaf))
ToPlot_Mau16 <- subset(WvsRAD_Mau16, Leaf <= 11 & Leaf >= 9) 

# Calculate Mau16 points :
Merged <- merge(ToPlot_Mau16,SensiRADgeno)
Merged$WidthPredB73 <- Merged$RADi*Abline_B73$Sensi + Merged$Leaf*Abline_B73$Oo
Merged$WidthPredGeno <-  Merged$RADi*Merged$Sensi + Merged$Leaf*Merged$Oo
Merged$Delta <-Merged$WidthPredGeno -  Merged$Width
Merged$WidthSim <- Merged$Width + Merged$Delta*1/2

# Calculate point from Field / Platform :
Merged_FvsP <- merge(ToPlot,SensiRADgeno)
Merged_FvsP$WidthPredGeno <-  Merged_FvsP$RADi*Merged_FvsP$Sensi + Merged_FvsP$Leaf*Merged_FvsP$Oo
Merged_FvsP$Delta <-Merged_FvsP$WidthPredGeno -  Merged_FvsP$Width
Merged_FvsP$Widthv2 <- Merged_FvsP$Width + Merged_FvsP$Delta

ggplot(data=Merged_FvsP,aes(x=RADint,y=Widthv2,group=Geno))0+
  geom_point()+
  geom_point(data=Merged,aes(x=RADi,y=WidthSim),color='red')+
  geom_abline(data=Abline,aes(slope=Sensi,intercept=Oo*10),color='blue')+
  #geom_abline(data=Abline,aes(slope=Sensi,intercept = Oo),color='grey')+
  geom_abline(slope=Abline_B73$Sensi,intercept = Abline_B73$Oo*10,color='grey')+
  facet_wrap(~Geno)+
  theme_light()+
  theme(legend.position='none')+
  scale_x_continuous(limits=c(0,3.5))+
  scale_y_continuous(limits=c(0,12.5))+
  labs(x='Intercepted radiation (MJ/plant)',y='Width (cm)')


```


------------' Tnight effect '------------

## Estimate Tnight effect :

```{r Estimate TNight and GN in all environments and 16 Hybrids, echo=FALSE, fig.height=8, fig.width=8, message=FALSE, warning=FALSE}

# Get grain number for each site for B73_H :
load(paste0(PATHS[4],'/Global_Output.RData'))
HarvestDataOutputs <- Get_Harvest_Data(Global_Output,list(Global_Output$Title,Global_Output$Genotype))
HarvestData <- HarvestDataOutputs[!duplicated(HarvestDataOutputs[,c(64,70)]),]

B73_Network <- select(subset(HarvestData,Genotype=='B73_H'),Title,GrainNo)
PATH.Met <- "D:/R-epositery/Widths/2.Met/Origin"


# Load Met data ----
MetFilesName <- list.files(path=PATH.Met,pattern = "\\.csv$",full.names = T)

for ( file in 1:length(MetFilesName) )
{

  DATA <- read.table(MetFilesName[file],
                       sep=";",                                           
                       header=TRUE,dec=".",
                       row.names=NULL,
                       as.is=FALSE, 
                       fill=TRUE)
  DATA <- subset(DATA,select=c('Tmoy','Tmin','Tmax','Jour','Light','VPDair','ET0'))
  DATA$Site_Year <- strsplit(strsplit(MetFilesName[file],"/",fixed=TRUE)[[1]][6],'.',fixed=TRUE)[[1]][1]
  DATA <- cbind(DATA,ThermalTime('Apsim',DATA$Jour,DATA$Tmoy))
  
   if (file ==1)
  {
    MetData <-  DATA
  }
  else {
    MetData <- rbind(MetData,DATA)    
  }
  
}

# Load Met data ----
Outputs <- data.frame(Site=character(),
                      Geno=character(),
                      Tnight=numeric(),
                      Yield=numeric())

library(lubridate)

for (geno in unique(HarvestData$Genotype))
{
  
  NetworkData <- select(subset(HarvestData,Genotype==geno),Title,GrainNo)
  
  for (site in unique(MetData$Site_Year))
  {
    DataYield <- subset(Global_Output,Title==paste0(' ',site) & Genotype == 'B73_H' & Stage > 5 & Stage < 7 )
    FromDay <- min(yday(DataYield$Date))
    ToDay <-  max(yday(DataYield$Date))
    Tnight <- mean(subset(MetData,Site_Year==site & Jour<ToDay & Jour >FromDay)$Tmin,na.rm=T)
    KNO <- subset(NetworkData,Title==paste0(' ',site))$GrainNo
    
    if (nrow(DataTnight)>0 & nrow(DataYield)>0)
    {
      Outputs <- rbind(Outputs,data.frame(site=site,
                                          Geno=geno,
                                          Tnight=Tnight,
                                          KNO=KNO))
    }
  
    
  }

}
```

```{r  Sensitivity of GN to Tnight, echo=FALSE, fig.height=8, fig.width=8, message=FALSE, warning=FALSE}

## B73_H :
ggplot(subset(Outputs,Geno=='B73_H'),aes(x=Tnight,y=KNO,label=site))+
  geom_point(size=3)+
  #geom_text_repel(size=5)+
  geom_smooth(method='lm',color='black',se=F)+
  scale_x_continuous(limits=c(5,25))+
  scale_y_continuous(limits=c(2000,6000))+
  theme_light()

# 16 Genotypes
ggplot(Outputs,aes(x=Tnight,y=KNO,label=site))+
  geom_point()+
  geom_smooth(method='lm',color='black')+
  scale_x_continuous(limits=c(5,25))+
  scale_y_continuous(limits=c(2000,6000))+
  theme_light()+
  facet_wrap(~Geno)

## Load Emilie Data :
M.Tnight <- read.table(paste0(PATHS[3],'/RespTnight.csv'), 
                       sep=",",                                           
                       header=TRUE,dec=".",
                       row.names=NULL,
                       as.is=FALSE, 
                       fill=TRUE)

M.Tnight$GN15cor <-  M.Tnight$grain.nb15.seq + ((300-M.Tnight$ri.zero.seq)*M.Tnight$resp.ri.sequ)
B73_TnResp <- subset(M.Tnight,Hybrid_ID=='B73_H')
Hybrids_TnResp <- select(subset(M.Tnight,Hybrid_ID %in% unique(Outputs$Geno)),Hybrid_ID,resp.tnight.seq,grain.nb15.seq)
colnames(Hybrids_TnResp) <- c('Geno','resp.tnight.seq','grain.nb15.seq')

```

```{r Sensitivity of GN to Tnight - B73, echo=FALSE, fig.height=5, fig.width=5, message=FALSE, warning=FALSE}

# Show results B73 :
ggplot(subset(Outputs,Geno=='B73_H'),aes(x=Tnight,y=KNO,label=site))+
  geom_point(size=3)+
  geom_text_repel(size=3)+
  geom_abline(slope=B73_TnResp$resp.tnight.seq,
              intercept=B73_TnResp$grain.nb15.seq-(B73_TnResp$resp.tnight.seq*15),
              color='black')+
  geom_smooth(method='lm',color='black')+
  #scale_x_continuous(limits=c(-10,10))+
  scale_y_continuous(limits=c(2000,6000))+
  theme_light()+
  ggtitle('B73 - DROPS network')

```

```{r Sensitivity of GN to Tnight - 16 Hybrids, echo=FALSE, fig.height=6.5, fig.width=6.5, message=FALSE, warning=FALSE}

# Show results 16 Hybrids :
ggplot(Outputs,aes(x=Tnight,y=KNO,label=site))+
  geom_point(size=2)+
  #geom_text_repel(size=3)+
  geom_abline(data=Hybrids_TnResp,
              aes(slope=resp.tnight.seq,
              intercept=grain.nb15.seq-(resp.tnight.seq*15)),
              color='blue')+
  geom_smooth(method='lm',color='black')+
  scale_x_continuous(limits=c(7,20))+
  scale_y_continuous(limits=c(2000,6000))+
  theme_light()+
  facet_wrap(~Geno)+
  labs(x='Night temperature (°C)',y='Grain number (grain/m²)')+
  ggtitle('16 Hybrids - DROPS network')

```

```{r Simulation multi-genotype / With Tnight Response, eval=FALSE, message=FALSE, warning=FALSE, include=FALSE}

GenoToSimulate$respTnight <- -GenoToSimulate$Res.Tnight
MyParams <- subset(GenoToSimulate,select=colnames(GenoToSimulate)[c(2:11,16,17,18)])


# Loading XML file : ----
ApsimFile<-paste(PATHS[1],'DROPS_Network_CalibrationvTnight',sep='/')
ApsimFile<-paste(ApsimFile,'apsim',sep='.')
SimFileTree<-xmlTreeParse(ApsimFile,getDTD = F , useInternalNodes = T)

# Prepare data : ----
SimToUse <- SimFileTree

# Test variables: ----
#geno <- 'B104_H'
#var <- colnames(MyParams)[2]
Stopped_At <- 1
WorkOnErrors <- F

# 1/ Loop over genotypes : ----
Start <- Sys.time()
for (i in Stopped_At:length(unique(MyParams$Hybrid_ID)))
#for (i in 2)
{
  Genotypes <- unique(MyParams$Hybrid_ID)
  geno <- as.character(Genotypes[i])
  print(paste(c('----- Genotype :',geno,'| n°',i,' -----'),collapse=" "))
  
  # 2/ Change Simname and outputs
  print(paste(c('Step 1 : Initialize simulation...'),collapse=" "))
  ParamsToUse <- subset(MyParams,Hybrid_ID==as.character(geno))
  SimulationName<-paste(ParamsToUse[,'Hybrid_ID'])
  NewApsimName<-paste(c(SimulationName,"apsim"),collapse=".")
  
  # CHeck if we have values for the genotype :
  if (nrow(ParamsToUse)>0)
  {
    # Loop over genotypic parameters and change them :
    for (var in colnames(ParamsToUse[2:ncol(ParamsToUse)]))
    {
       VarNode<-getNodeSet(SimToUse, paste(c("//",var),collapse="") )
       NewValue<- ParamsToUse[,var]
       
      for (node in 1:length(VarNode))
      {
         ChangeXmlValue(VarNode[node],NewValue) 
      }
       
    }
    print(paste('==> Done !'),collapse=" ")
  
    # Save the new apsimfile in Model Directory :
    print(paste(c('Step 2 : Saving simulation file...'),collapse=" "))
    saveXML(xmlRoot(SimFileTree), 
            file= paste0(PATH.APSIM,NewApsimName), 
            prefix= '<?xml version="1.0" encoding="utf-8"?>',indent= T)
    print(paste('==> Done !'),collapse=" ")
  
    # Simulate :
    print(paste(c('Step 3 : Simulation...'),collapse=" "))
    setwd(PATH.APSIM)
    
    command<-paste("apsim",NewApsimName)

    if (WorkOnErrors==T)
    {
      result <- tryCatch(evalWithTimeout(system(command,show.output.on.console=T),
                                 timeout = 30,
                                 onTimeout = "error"),
                 error=
                   {
                        #Get all tasks running :
                        Tasks <- system('tasklist',show.output.on.console=T,intern=T)
                        # Kill jobrunner :
                        while (sum(grepl("JobRunner.exe", Tasks))>0)
                        {  
                          print('Killing JobRunner...')
                          system('taskkill /IM JobRunner.exe /F',
                                 show.output.on.console=F,intern=T)
                          Tasks <- system('tasklist',show.output.on.console=T,intern=T)
                        }
                        # Kill all Apsim.exe :
                        while (sum(grepl("Apsim.exe", Tasks))>0)
                        {
                          print('Killing APSIM...')
                          system('taskkill /IM Apsim.exe /F',
                                 show.output.on.console=F,intern=T)       
                          Tasks <- system('tasklist',show.output.on.console=T,intern=T)
                        }
                    }
                 )
      
      # If it runs smoothly :
      if (result==0) {print(paste('==> Done !'),collapse=" ")
      # If there was an error : 
      } else {print(paste('==> Error Handled !'),collapse=" ")}
    }
    else {system(command)}
      
    
    # Get/Save outputs  : 
    print(paste(c('Step 4 : Loading outputs...'),collapse=" "))
    
    Output_Sim <- loadApsim(PATH.APSIM, 
                            loadAll = TRUE, 
                            ext = ".out", 
                            returnFrame = TRUE, 
                            n = 0, fill = TRUE, 
                            addConstants = TRUE)
    
    Output_Sim$Genotype<-geno
  
    
    if (i ==1)
    {Global_Output_Tnight <- Output_Sim
    }else{
      Global_Output_Tnight <- rbind(Global_Output_Tnight,Output_Sim)
    }
    rm(Output_Sim)
  }
  
  print(paste('==> Done !'),collapse=" ")  
  
}

Stop <- Sys.time()
TimeDiff <- difftime(Stop, Start, units="secs")
MyTimeDiff <- format(.POSIXct(TimeDiff,tz="GMT"), "%H:%M:%S")
print(paste(c(''),collapse=" "))
print(paste(c('----- Simulation loops finished -----'),collapse=" "))
print(paste(c(''),collapse=" "))
print(paste(c('Time of execution is '),MyTimeDiff,collapse=" "))


# 2/ Save and treat outputs : ----
save(file=paste0(PATHS[4],'/Global_Output_Tnight.RData'),Global_Output_Tnight)
  
Mau16_ToTreat <- subset(Global_Output_Tnight,Title==" Mauguio_2016")

which(!is.na(Global_Output_Tnight$RADi))

```

```{r Treat Data -  GN / GY / With Tnight Response, echo=FALSE, fig.height=5, fig.width=5, message=FALSE, warning=FALSE}


# Obs : 
# ------

## GN :
Measured_GN15 <- read.table(paste0(PATHS[3],'/BLUES_KNO.csv'), 
                       sep=";",                                           
                       header=TRUE,dec=".",
                       row.names=NULL,
                       as.is=FALSE, 
                       fill=TRUE)

# Separate data to fit the one we already have (in lines not columns...)
Measured_GN <- gather(Measured_GN15,one_of(colnames(Measured_GN15)[5:33]),key = 'Site',value = 'KNO')
Measured_GN_v2 <- separate(Measured_GN,Site,sep="_",into=c('Site','Year_Scenario'),extra='merge')
Measured_GN_v3 <- separate(Measured_GN_v2,Year_Scenario,sep="_",into=c('Year','Scenario'),extra='merge')
Measured_GN_v4 <- separate(Measured_GN_v3,Site,sep="[.]",into=c('Var','Site'),extra='merge')
Measured_GN_v5 <- select(subset(Measured_GN_v4,Scenario=="watered"),Hybrid_ID,Site,Year,KNO)
Measured_Data_GN <- unite(Measured_GN_v5,Site,Year,col="Title",sep='_')
Measured_Data_GN$Title <- factor(Measured_Data_GN$Title)
levels(Measured_Data_GN$Title) <- c('Martonvasar_2013','Gaillac_2012','Gaillac_2013','Graneros_2013',
                                 'Nerac_2012','Nerac_2013','Caracal_2012','Karlsruhe_2012',
                                 'Karlsruhe_2013','Murony_2013','Karacal_2012',
                                 'Campagnola_2013','Cadriano_2012')
colnames(Measured_Data_GN) <- c("Genotype","Title","M.GrainNo")  

## GN :
Measured_GY15 <- read.table(paste0(PATHS[3],'/BLUES_GY15.csv'), 
                       sep=",",                                           
                       header=TRUE,dec=".",
                       row.names=NULL,  
                       as.is=FALSE, 
                       fill=TRUE)

# Separate data to fit the one we already have (in lines not columns...)
Measured_GY <- unite(
                 select(
                   subset(
                     separate(
                        separate(
                          separate(
                            gather(Measured_GY15,
                                   one_of(colnames(Measured_GY15)[5:33]),key = 'Site',value ='Yield'),
                            Site,sep="_",into=c('Site','Year_Scenario'),extra='merge'),
                        Year_Scenario,sep="_",into=c('Year','Scenario'),extra='merge'),
                     Site,sep="[.]",into=c('Var','Site'),extra='merge'),
                   Scenario=="watered"),
                 Hybrid_ID,Site,Year,Yield),
               Site,Year,col="Title",sep='_')

Measured_GY$Title <- factor(Measured_GY$Title)


levels(Measured_GY$Title) <- c('Martonvasar_2013','Gaillac_2012','Gaillac_2013','Graneros_2013',
                                 'Nerac_2012',' Nerac_2013','Caracal_2012','Karlsruhe_2012',
                                 'Karlsruhe_2013','Murony_2013','Karacal_2012',
                                 'Campagnola_2012','Campagnola_2013','Cadriano_2012')
colnames(Measured_GY) <- c("Genotype","Title","M.Yield") 



MeasuredYield <- merge(Measured_Data_GN,Measured_GY,by=c('Genotype','Title'))
MeasuredYield$M.Yield.kgha <- MeasuredYield$M.Yield *1000

# -----------------------------------

# Sim :
# ------
load(paste0(PATHS[4],'/Global_Output_Tnight.RData'))
HarvestDataOutputs <- Get_Harvest_Data(Global_Output_Tnight,list(Global_Output_Tnight$Title,Global_Output_Tnight$Genotype))
HarvestData <- HarvestDataOutputs[!duplicated(HarvestDataOutputs[,c(64,70)]),]
Network_GN_GY <- select(HarvestData,Title,Genotype,GrainNo,yield)
Network_GN_GY$Title  <- gsub(' ', '', Network_GN_GY$Title)

MeasuredYield$Title <- as.character(MeasuredYield$Title)
MeasuredYield$Titlev2 <- gsub(' ', '', MeasuredYield$Title)

MeasuredYield$Genotype <- as.character(MeasuredYield$Genotype)

Data_GY_GN <- merge(MeasuredYield,Network_GN_GY,by=c('Genotype','Title'))



  

```

```{r Show Sim/Obs GN/GY - B73 / With Tnight Response, echo=FALSE, fig.height=5, fig.width=5, message=FALSE, warning=FALSE}

ToPlot <- subset(Data_GY_GN,Genotype=='B73_H')


R2 <- round(summary(lm(ToPlot,formula = 'M.GrainNo~GrainNo'))$r.squared,digits=3)
RMSE <- sqrt(sum((ToPlot$M.GrainNo-ToPlot$GrainNo)^2,na.rm=T)*(1/nrow(ToPlot)))

ggplot(data=ToPlot,aes(x=M.GrainNo,y=GrainNo,label=Title))+
  geom_point(size=2)+
  geom_text_repel(size=3.5)+
  geom_smooth(method='lm',se=F)+
  geom_abline(slope=1,intercept=0,color='black')+
  scale_x_continuous(limits=c(1000,6000))+
  scale_y_continuous(limits=c(1000,6000))+
  #geom_text(aes(x=2500,y=2000,label=paste0('R² = ',round(R2,digits=2))),color='black')+
  #geom_text(aes(x=2500,y=1500,label=paste0('RMSE = ',round(RMSE,digits=2),' gr/m²')),color='black')+
  theme_light()+
  labs(x='Measured grain number (gr/m²)',y='Simulated Grain number (gr/m²)')+
  ggtitle('Genotype B73 - DROPS network')


R2 <- round(summary(lm(ToPlot,formula = 'M.Yield~yield'))$r.squared,digits=3)
RMSE <- sqrt(sum((ToPlot$M.Yield-(ToPlot$yield/1000))^2,na.rm=T)*(1/nrow(ToPlot)))

ggplot(data=ToPlot,aes(x=M.Yield,y=yield/1000,label=Title))+
  geom_point(size=2)+
  geom_text_repel(size=3.5)+
  geom_smooth(method='lm',se=F)+
  geom_abline(slope=1,intercept=0,color='black')+
  scale_x_continuous(limits=c(0,20))+
  scale_y_continuous(limits=c(0,20))+
  #geom_text(aes(x=6,y=6,label=paste0('R² = ',round(R2,digits=2))),color='black')+
 # geom_text(aes(x=6,y=5,label=paste0('RMSE = ',round(RMSE,digits=2),' t/ha')),color='black')+
  theme_light()+
  labs(x='Measured grain yield (t/ha)',y='Simulated grain yield (t/ha)')+
  ggtitle('Genotype B73 - DROPS network')

R2 <- round(summary(lm(ToPlot,formula = 'Sim~Obs'))$r.squared,digits=3)
RMSE <- sqrt(sum((ToPlot$Sim-ToPlot$Obs)^2,na.rm=T)*(1/nrow(ToPlot)))

colnames(ToPlot)
```

```{r Show Sim/Obs GN/GY - 16 Hybrids / With Tnight Response, echo=FALSE, fig.height=5, fig.width=5, message=FALSE, warning=FALSE}


ggplot(data=Data_GY_GN,aes(x=M.GrainNo,y=GrainNo,label=Title))+
  geom_point(size=2)+
  geom_smooth(method='lm',se=F)+
  geom_abline(slope=1,intercept=0,color='black')+
  scale_x_continuous(limits=c(1000,6000))+
  scale_y_continuous(limits=c(1000,6000))+
  theme_light()+
  facet_wrap(~Genotype)+
  labs(x='Measured grain number (gr/m²)',y='Simulated grain number (gr/m²)')+
  ggtitle('16 hybrids - DROPS network')



```

## Calibrate kernel weight :

```{r Simulation multi-genotype - Fixed PotKW , eval=FALSE, message=FALSE, warning=FALSE, include=FALSE}

GenoToSimulate$respTnight <- -GenoToSimulate$Res.Tnight
MyParams <- subset(GenoToSimulate,select=colnames(GenoToSimulate)[c(2:9,11,16,17,18)])

# Loading XML file : ----
ApsimFile<-paste(PATHS[1],'DROPS_Network_CalibrationvPotKW',sep='/')
ApsimFile<-paste(ApsimFile,'apsim',sep='.')
SimFileTree<-xmlTreeParse(ApsimFile,getDTD = F , useInternalNodes = T)

# Prepare data : ----
SimToUse <- SimFileTree

# Test variables: ----
#geno <- 'B104_H'
#var <- colnames(MyParams)[2]
Stopped_At <- 1
WorkOnErrors <- F

# 1/ Loop over genotypes : ----
Start <- Sys.time()
for (i in Stopped_At:length(unique(MyParams$Hybrid_ID)))
#for (i in 2)
{
  Genotypes <- unique(MyParams$Hybrid_ID)
  geno <- as.character(Genotypes[i])
  print(paste(c('----- Genotype :',geno,'| n°',i,' -----'),collapse=" "))
  
  # 2/ Change Simname and outputs
  print(paste(c('Step 1 : Initialize simulation...'),collapse=" "))
  ParamsToUse <- subset(MyParams,Hybrid_ID==as.character(geno))
  SimulationName<-paste(ParamsToUse[,'Hybrid_ID'])
  NewApsimName<-paste(c(SimulationName,"apsim"),collapse=".")
  
  # CHeck if we have values for the genotype :
  if (nrow(ParamsToUse)>0)
  {
    # Loop over genotypic parameters and change them :
    for (var in colnames(ParamsToUse[2:ncol(ParamsToUse)]))
    {
       VarNode<-getNodeSet(SimToUse, paste(c("//",var),collapse="") )
       NewValue<- ParamsToUse[,var]
       
      for (node in 1:length(VarNode))
      {
         ChangeXmlValue(VarNode[node],NewValue) 
      }
       
    }
    print(paste('==> Done !'),collapse=" ")
  
    # Save the new apsimfile in Model Directory :
    print(paste(c('Step 2 : Saving simulation file...'),collapse=" "))
    saveXML(xmlRoot(SimFileTree), 
            file= paste0(PATH.APSIM,NewApsimName), 
            prefix= '<?xml version="1.0" encoding="utf-8"?>',indent= T)
    print(paste('==> Done !'),collapse=" ")
  
    # Simulate :
    print(paste(c('Step 3 : Simulation...'),collapse=" "))
    setwd(PATH.APSIM)
    
    command<-paste("apsim",NewApsimName)

    if (WorkOnErrors==T)
    {
      result <- tryCatch(evalWithTimeout(system(command,show.output.on.console=T),
                                 timeout = 30,
                                 onTimeout = "error"),
                 error=
                   {
                        #Get all tasks running :
                        Tasks <- system('tasklist',show.output.on.console=T,intern=T)
                        # Kill jobrunner :
                        while (sum(grepl("JobRunner.exe", Tasks))>0)
                        {  
                          print('Killing JobRunner...')
                          system('taskkill /IM JobRunner.exe /F',
                                 show.output.on.console=F,intern=T)
                          Tasks <- system('tasklist',show.output.on.console=T,intern=T)
                        }
                        # Kill all Apsim.exe :
                        while (sum(grepl("Apsim.exe", Tasks))>0)
                        {
                          print('Killing APSIM...')
                          system('taskkill /IM Apsim.exe /F',
                                 show.output.on.console=F,intern=T)       
                          Tasks <- system('tasklist',show.output.on.console=T,intern=T)
                        }
                    }
                 )
      
      # If it runs smoothly :
      if (result==0) {print(paste('==> Done !'),collapse=" ")
      # If there was an error : 
      } else {print(paste('==> Error Handled !'),collapse=" ")}
    }
    else {system(command)}
      
    
    # Get/Save outputs  : 
    print(paste(c('Step 4 : Loading outputs...'),collapse=" "))
    
    Output_Sim <- loadApsim(PATH.APSIM, 
                            loadAll = TRUE, 
                            ext = ".out", 
                            returnFrame = TRUE, 
                            n = 0, fill = TRUE, 
                            addConstants = TRUE)
    
    Output_Sim$Genotype<-geno
  
    if (i ==1)
    {Global_Output_FixedPotKW <- Output_Sim
    }else{
     Global_Output_FixedPotKW <- rbind(Global_Output_FixedPotKW,Output_Sim)
    }
    rm(Output_Sim)
  }
  
  print(paste('==> Done !'),collapse=" ")  
  
}

Stop <- Sys.time()
TimeDiff <- difftime(Stop, Start, units="secs")
MyTimeDiff <- format(.POSIXct(TimeDiff,tz="GMT"), "%H:%M:%S")
print(paste(c(''),collapse=" "))
print(paste(c('----- Simulation loops finished -----'),collapse=" "))
print(paste(c(''),collapse=" "))
print(paste(c('Time of execution is '),MyTimeDiff,collapse=" "))


# 2/ Save and treat outputs : ----
save(file=paste0(PATHS[4],'/Global_Output_FixedPotKW.RData'),Global_Output_FixedPotKW)

```

```{r Simulation multi-genotype - Genotypic PotKW - Cali (-10%) , eval=FALSE, message=FALSE, warning=FALSE, include=FALSE}

GenoToSimulate$respTnight <- -GenoToSimulate$Res.Tnight
MyParams <- subset(GenoToSimulate,select=colnames(GenoToSimulate)[c(2:11,16,17,18)])
MyParams$potKernelWt <- MyParams$potKernelWt - (20/100*MyParams$potKernelWt)

# Loading XML file : ----
ApsimFile<-paste(PATHS[1],'DROPS_Network_CalibrationvPotKW',sep='/')
ApsimFile<-paste(ApsimFile,'apsim',sep='.')
SimFileTree<-xmlTreeParse(ApsimFile,getDTD = F , useInternalNodes = T)

# Prepare data : ----
SimToUse <- SimFileTree

# Test variables: ----
#geno <- 'B104_H'
#var <- colnames(MyParams)[2]
Stopped_At <- 1
WorkOnErrors <- F

# 1/ Loop over genotypes : ----
Start <- Sys.time()
for (i in Stopped_At:length(unique(MyParams$Hybrid_ID)))
#for (i in 2)
{
  Genotypes <- unique(MyParams$Hybrid_ID)
  geno <- as.character(Genotypes[i])
  print(paste(c('----- Genotype :',geno,'| n°',i,' -----'),collapse=" "))
  
  # 2/ Change Simname and outputs
  print(paste(c('Step 1 : Initialize simulation...'),collapse=" "))
  ParamsToUse <- subset(MyParams,Hybrid_ID==as.character(geno))
  SimulationName<-paste(ParamsToUse[,'Hybrid_ID'])
  NewApsimName<-paste(c(SimulationName,"apsim"),collapse=".")
  
  # CHeck if we have values for the genotype :
  if (nrow(ParamsToUse)>0)
  {
    # Loop over genotypic parameters and change them :
    for (var in colnames(ParamsToUse[2:ncol(ParamsToUse)]))
    {
       VarNode<-getNodeSet(SimToUse, paste(c("//",var),collapse="") )
       NewValue<- ParamsToUse[,var]
       
      for (node in 1:length(VarNode))
      {
         ChangeXmlValue(VarNode[node],NewValue) 
      }
       
    }
    print(paste('==> Done !'),collapse=" ")
  
    # Save the new apsimfile in Model Directory :
    print(paste(c('Step 2 : Saving simulation file...'),collapse=" "))
    saveXML(xmlRoot(SimFileTree), 
            file= paste0(PATH.APSIM,NewApsimName), 
            prefix= '<?xml version="1.0" encoding="utf-8"?>',indent= T)
    print(paste('==> Done !'),collapse=" ")
  
    # Simulate :
    print(paste(c('Step 3 : Simulation...'),collapse=" "))
    setwd(PATH.APSIM)
    
    command<-paste("apsim",NewApsimName)

    if (WorkOnErrors==T)
    {
      result <- tryCatch(evalWithTimeout(system(command,show.output.on.console=T),
                                 timeout = 30,
                                 onTimeout = "error"),
                 error=
                   {
                        #Get all tasks running :
                        Tasks <- system('tasklist',show.output.on.console=T,intern=T)
                        # Kill jobrunner :
                        while (sum(grepl("JobRunner.exe", Tasks))>0)
                        {  
                          print('Killing JobRunner...')
                          system('taskkill /IM JobRunner.exe /F',
                                 show.output.on.console=F,intern=T)
                          Tasks <- system('tasklist',show.output.on.console=T,intern=T)
                        }
                        # Kill all Apsim.exe :
                        while (sum(grepl("Apsim.exe", Tasks))>0)
                        {
                          print('Killing APSIM...')
                          system('taskkill /IM Apsim.exe /F',
                                 show.output.on.console=F,intern=T)       
                          Tasks <- system('tasklist',show.output.on.console=T,intern=T)
                        }
                    }
                 )
      
      # If it runs smoothly :
      if (result==0) {print(paste('==> Done !'),collapse=" ")
      # If there was an error : 
      } else {print(paste('==> Error Handled !'),collapse=" ")}
    }
    else {system(command)}
      
    
    # Get/Save outputs  : 
    print(paste(c('Step 4 : Loading outputs...'),collapse=" "))
    
    Output_Sim <- loadApsim(PATH.APSIM, 
                            loadAll = TRUE, 
                            ext = ".out", 
                            returnFrame = TRUE, 
                            n = 0, fill = TRUE, 
                            addConstants = TRUE)
    
    Output_Sim$Genotype<-geno
  
    if (i ==1)
    {Global_Output_GenoPotKW_Cali <- Output_Sim
    }else{
     Global_Output_GenoPotKW_Cali <- rbind(Global_Output_GenoPotKW_Cali,Output_Sim)
    }
    rm(Output_Sim)
  }
  
  print(paste('==> Done !'),collapse=" ")  
  
}

Stop <- Sys.time()
TimeDiff <- difftime(Stop, Start, units="secs")
MyTimeDiff <- format(.POSIXct(TimeDiff,tz="GMT"), "%H:%M:%S")
print(paste(c(''),collapse=" "))
print(paste(c('----- Simulation loops finished -----'),collapse=" "))
print(paste(c(''),collapse=" "))
print(paste(c('Time of execution is '),MyTimeDiff,collapse=" "))


# 2/ Save and treat outputs : ----
save(file=paste0(PATHS[4],'/Global_Output_GenoPotKW_Cali.RData'),Global_Output_GenoPotKW_Cali)

```

```{r Simulation multi-genotype - Genotypic PotKW , eval=FALSE, message=FALSE, warning=FALSE, include=FALSE}

GenoToSimulate$respTnight <- -GenoToSimulate$Res.Tnight
MyParams <- subset(GenoToSimulate,select=colnames(GenoToSimulate)[c(2:9,11,16,17,18)])

# Loading XML file : ----
ApsimFile<-paste(PATHS[1],'DROPS_Network_CalibrationvPotKW',sep='/')
ApsimFile<-paste(ApsimFile,'apsim',sep='.')
SimFileTree<-xmlTreeParse(ApsimFile,getDTD = F , useInternalNodes = T)

# Prepare data : ----
SimToUse <- SimFileTree

# Test variables: ----
#geno <- 'B104_H'
#var <- colnames(MyParams)[2]
Stopped_At <- 1
WorkOnErrors <- F

# 1/ Loop over genotypes : ----
Start <- Sys.time()
for (i in Stopped_At:length(unique(MyParams$Hybrid_ID)))
#for (i in 2)
{
  Genotypes <- unique(MyParams$Hybrid_ID)
  geno <- as.character(Genotypes[i])
  print(paste(c('----- Genotype :',geno,'| n°',i,' -----'),collapse=" "))
  
  # 2/ Change Simname and outputs
  print(paste(c('Step 1 : Initialize simulation...'),collapse=" "))
  ParamsToUse <- subset(MyParams,Hybrid_ID==as.character(geno))
  SimulationName<-paste(ParamsToUse[,'Hybrid_ID'])
  NewApsimName<-paste(c(SimulationName,"apsim"),collapse=".")
  
  # CHeck if we have values for the genotype :
  if (nrow(ParamsToUse)>0)
  {
    # Loop over genotypic parameters and change them :
    for (var in colnames(ParamsToUse[2:ncol(ParamsToUse)]))
    {
       VarNode<-getNodeSet(SimToUse, paste(c("//",var),collapse="") )
       NewValue<- ParamsToUse[,var]
       
      for (node in 1:length(VarNode))
      {
         ChangeXmlValue(VarNode[node],NewValue) 
      }
       
    }
    print(paste('==> Done !'),collapse=" ")
  
    # Save the new apsimfile in Model Directory :
    print(paste(c('Step 2 : Saving simulation file...'),collapse=" "))
    saveXML(xmlRoot(SimFileTree), 
            file= paste0(PATH.APSIM,NewApsimName), 
            prefix= '<?xml version="1.0" encoding="utf-8"?>',indent= T)
    print(paste('==> Done !'),collapse=" ")
  
    # Simulate :
    print(paste(c('Step 3 : Simulation...'),collapse=" "))
    setwd(PATH.APSIM)
    
    command<-paste("apsim",NewApsimName)

    if (WorkOnErrors==T)
    {
      result <- tryCatch(evalWithTimeout(system(command,show.output.on.console=T),
                                 timeout = 30,
                                 onTimeout = "error"),
                 error=
                   {
                        #Get all tasks running :
                        Tasks <- system('tasklist',show.output.on.console=T,intern=T)
                        # Kill jobrunner :
                        while (sum(grepl("JobRunner.exe", Tasks))>0)
                        {  
                          print('Killing JobRunner...')
                          system('taskkill /IM JobRunner.exe /F',
                                 show.output.on.console=F,intern=T)
                          Tasks <- system('tasklist',show.output.on.console=T,intern=T)
                        }
                        # Kill all Apsim.exe :
                        while (sum(grepl("Apsim.exe", Tasks))>0)
                        {
                          print('Killing APSIM...')
                          system('taskkill /IM Apsim.exe /F',
                                 show.output.on.console=F,intern=T)       
                          Tasks <- system('tasklist',show.output.on.console=T,intern=T)
                        }
                    }
                 )
      
      # If it runs smoothly :
      if (result==0) {print(paste('==> Done !'),collapse=" ")
      # If there was an error : 
      } else {print(paste('==> Error Handled !'),collapse=" ")}
    }
    else {system(command)}
      
    
    # Get/Save outputs  : 
    print(paste(c('Step 4 : Loading outputs...'),collapse=" "))
    
    Output_Sim <- loadApsim(PATH.APSIM, 
                            loadAll = TRUE, 
                            ext = ".out", 
                            returnFrame = TRUE, 
                            n = 0, fill = TRUE, 
                            addConstants = TRUE)
    
    Output_Sim$Genotype<-geno
  
    if (i ==1)
    {Global_Output_GenoPotKW <- Output_Sim
    }else{
     Global_Output_GenoPotKW <- rbind(Global_Output_GenoPotKW,Output_Sim)
    }
    rm(Output_Sim)
  }
  
  print(paste('==> Done !'),collapse=" ")  
  
}

Stop <- Sys.time()
TimeDiff <- difftime(Stop, Start, units="secs")
MyTimeDiff <- format(.POSIXct(TimeDiff,tz="GMT"), "%H:%M:%S")
print(paste(c(''),collapse=" "))
print(paste(c('----- Simulation loops finished -----'),collapse=" "))
print(paste(c(''),collapse=" "))
print(paste(c('Time of execution is '),MyTimeDiff,collapse=" "))


# 2/ Save and treat outputs : ----
save(file=paste0(PATHS[4],'/Global_Output_GenoPotKW.RData'),Global_Output_GenoPotKW)

```

```{r Show differences with and without potKW, eval=FALSE, message=FALSE, warning=FALSE, include=FALSE}

# Sim PotKW:
# --------------
load(paste0(PATHS[4],'/Global_Output_GenoPotKW.RData'))
HarvestDataOutputs <- Get_Harvest_Data(Global_Output_GenoPotKW,list(Global_Output_GenoPotKW$Title,Global_Output_GenoPotKW$Genotype))
HarvestData <- HarvestDataOutputs[!duplicated(HarvestDataOutputs[,c(64,70)]),]
Network_GN_GY <- select(HarvestData,Title,Genotype,GrainNo,yield)
Network_GN_GY$Title  <- gsub(' ', '', Network_GN_GY$Title)

MeasuredYield$Title <- as.character(MeasuredYield$Title)
MeasuredYield$Titlev2 <- gsub(' ', '', MeasuredYield$Title)

MeasuredYield$Genotype <- as.character(MeasuredYield$Genotype)

Data_GY_GN <- merge(MeasuredYield,Network_GN_GY,by=c('Genotype','Title'))

ToPlotv1 <- Data_GY_GN

#R2 <- round(summary(lm(ToPlot,formula = 'M.Yield~yield'))$r.squared,digits=3)
#RMSE <- sqrt(sum((ToPlot$M.Yield-(ToPlot$yield/1000))^2,na.rm=T)*(1/nrow(ToPlot)))

ggplot(data=ToPlotv2,aes(x=M.Yield,y=yield/1000,label=Title))+
  geom_point(size=2)+
  geom_smooth(method='lm',se=F)+
  geom_abline(slope=1,intercept=0,color='black')+
  scale_x_continuous(limits=c(0,20))+
  scale_y_continuous(limits=c(0,20))+
  facet_wrap(~Genotype)+
  #geom_text(aes(x=6,y=6,label=paste0('R² = ',round(R2,digits=2))),color='black')+
  #geom_text(aes(x=6,y=5,label=paste0('RMSE = ',round(RMSE,digits=2),' t/ha')),color='black')+
  theme_light()+
  labs(x='Measured grain yield (t/ha)',y='Simulated grain yield (t/ha)')+
  ggtitle('Genotype B73 - DROPS network')

# Sim No PotKW :
# --------------
load(paste0(PATHS[4],'/Global_Output_FixedPotKW.RData'))
HarvestDataOutputs <- Get_Harvest_Data(Global_Output_FixedPotKW,
                                       list(Global_Output_FixedPotKW$Title,Global_Output_FixedPotKW$Genotype))
HarvestData <- HarvestDataOutputs[!duplicated(HarvestDataOutputs[,c(64,70)]),]
Network_GN_GY <- select(HarvestData,Title,Genotype,GrainNo,yield)
Network_GN_GY$Title  <- gsub(' ', '', Network_GN_GY$Title)

MeasuredYield$Title <- as.character(MeasuredYield$Title)
MeasuredYield$Titlev2 <- gsub(' ', '', MeasuredYield$Title)

MeasuredYield$Genotype <- as.character(MeasuredYield$Genotype)

Data_GY_GN <- merge(MeasuredYield,Network_GN_GY,by=c('Genotype','Title'))


ToPlotv2 <- Data_GY_GN


# Sim PotKW Cali :
# --------------
load(paste0(PATHS[4],'/Global_Output_GenoPotKW_Cali.RData'))
HarvestDataOutputs <- Get_Harvest_Data(Global_Output_GenoPotKW_Cali,
                                       list(Global_Output_GenoPotKW_Cali$Title,Global_Output_GenoPotKW_Cali$Genotype))
HarvestData <- HarvestDataOutputs[!duplicated(HarvestDataOutputs[,c(64,70)]),]
Network_GN_GY <- subset(HarvestData,select=c('Title','Genotype','GrainNo','yield'))
Network_GN_GY$Title  <- gsub(' ', '', Network_GN_GY$Title)

MeasuredYield$Title <- as.character(MeasuredYield$Title)
MeasuredYield$Titlev2 <- gsub(' ', '', MeasuredYield$Title)

MeasuredYield$Genotype <- as.character(MeasuredYield$Genotype)

Data_GY_GN <- merge(MeasuredYield,Network_GN_GY,by=c('Genotype','Title'))

ToPlotv3 <- Data_GY_GN

#R2 <- round(summary(lm(ToPlot,formula = 'M.Yield~yield'))$r.squared,digits=3)
#RMSE <- sqrt(sum((ToPlot$M.Yield-(ToPlot$yield/1000))^2,na.rm=T)*(1/nrow(ToPlot)))

ggplot(data=subset(ToPlotv3,Title!='Graneros_2013'),aes(x=M.Yield,y=yield/1000,label=Title,color=Title))+
  geom_point(size=2)+
  geom_smooth(method='lm',se=F,color='black')+
  geom_abline(slope=1,intercept=0,color='black')+
  scale_x_continuous(limits=c(0,20))+
  scale_y_continuous(limits=c(0,20))+
  #geom_text(aes(x=6,y=6,label=paste0('R² = ',round(R2,digits=2))),color='black')+
 # geom_text(aes(x=6,y=5,label=paste0('RMSE = ',round(RMSE,digits=2),' t/ha')),color='black')+
  facet_wrap(~Genotype)+
  theme_light()+
  labs(x='Measured grain yield (t/ha)',y='Simulated grain yield (t/ha)')+
  ggtitle('16 Hybrids - DROPS network')


### All Together : 
YieldFor1 <- select(ToPlotv1,Genotype,Title,yield,M.Yield.kgha)
YieldFor2 <- select(ToPlotv2,yield)
YieldFor3 <- select(ToPlotv3,yield)

ToPlot <- cbind(YieldFor1,YieldFor2)
colnames(ToPlot) <- c('Genotype','Title','Genotypic_PotKW','Measured','Fixed_PotKW')

ggplot(data=subset(ToPlot,Title!='Graneros_2013'))+
  geom_point(aes(x=Measured,y=Genotypic_PotKW),color='blue')+
  geom_smooth(aes(x=Measured,y=Genotypic_PotKW),color='blue',method='lm',se=F)+
  geom_point(aes(x=Measured,y=Fixed_PotKW),color='red')+
  geom_smooth(aes(x=Measured,y=Fixed_PotKW),color='red',method='lm',se=F)+
  facet_wrap(~Genotype)+
  scale_x_continuous(limits=c(0,15000))+
  scale_y_continuous(limits=c(0,15000))+
  geom_abline(slope=1,intercept=0,color='black')
  

ggplot(data=subset(ToPlotv3,Title!='Graneros_2013'),aes(x=M.Yield,y=yield/1000,label=Title,color=Genotype))+
  geom_point(size=2)+
  geom_smooth(method='lm',se=F,color='black')+
  geom_abline(slope=1,intercept=0,color='black')+
  scale_x_continuous(limits=c(0,20))+
  scale_y_continuous(limits=c(0,20))+
  #geom_text(aes(x=6,y=6,label=paste0('R² = ',round(R2,digits=2))),color='black')+
 # geom_text(aes(x=6,y=5,label=paste0('RMSE = ',round(RMSE,digits=2),' t/ha')),color='black')+
  facet_wrap(~Title)+
  theme_light()+
  labs(x='Measured grain yield (t/ha)',y='Simulated grain yield (t/ha)')+
  ggtitle('16 Hybrids - DROPS network')

ggplot(data=subset(ToPlotv3,Title!='Graneros_2013'),aes(x=M.Yield,y=yield/1000,label=Title,color=Genotype))+
  geom_point(size=2)+
  geom_smooth(method='lm',se=F,color='black')+
  geom_abline(slope=1,intercept=0,color='black')+
  scale_x_continuous(limits=c(0,20))+
  scale_y_continuous(limits=c(0,20))+
  #geom_text(aes(x=6,y=6,label=paste0('R² = ',round(R2,digits=2))),color='black')+
 # geom_text(aes(x=6,y=5,label=paste0('RMSE = ',round(RMSE,digits=2),' t/ha')),color='black')+
  #facet_wrap(~Genotype)+
  theme_light()+
  labs(x='Measured grain yield (t/ha)',y='Simulated grain yield (t/ha)')+
  ggtitle('16 Hybrids - DROPS network')
```

```{r Prepare PARAMS - Simulation des 250 genos sur panel DROPS, eval=FALSE, message=FALSE, warning=FALSE, include=FALSE}


Parameters <- read.table('F:/Sebastien/Sims DROPS GxE/3.Data/Parameters_DROPS_Panel_ForAPSIM_WithNA.csv', 
                          sep=",", header=TRUE,dec=".",row.names=NULL,as.is=FALSE, fill=TRUE)
Parameters$respTnight <- -Parameters$Res.Tnight

# replace missing params with NAs
for (col in colnames(Parameters))
{
  ToTreat <- subset(Parameters,select=col)
  ToTreat[is.na(ToTreat[,col]),] <- mean(ToTreat[,col],na.rm=T)
  Parameters[col] <- ToTreat
}


```

## Estimate parameters effect on  :

```{r Simulation des 250 genos sur panel DROPS, eval=FALSE, message=FALSE, warning=FALSE, include=FALSE}

MyParams <- subset(Parameters,select=colnames(Parameters)[c(2:11,16,17)])
MyParams$potKernelWt <- Parameters$potKernelWt - (20/100*Parameters$potKernelWt)
MyParams$b <- (-0.0758*MyParams$c)-0.3672

# Loading XML file : ----
ApsimFile<-paste(PATHS[1],'DROPS_Network_CalibrationvPotKW',sep='/')
ApsimFile<-paste(ApsimFile,'apsim',sep='.')
SimFileTree<-xmlTreeParse(ApsimFile,getDTD = F , useInternalNodes = T)

# Prepare data : ----
SimToUse <- SimFileTree

# Test variables: ----
#geno <- 'B104_H'
#var <- colnames(MyParams)[2]
Stopped_At <- 1
WorkOnErrors <- F

# 1/ Loop over genotypes : ----
Start <- Sys.time()
for (i in Stopped_At:length(unique(MyParams$Hybrid_ID)))
#for (i in 2)
{
  Genotypes <- unique(MyParams$Hybrid_ID)
  geno <- as.character(Genotypes[i])
  print(paste(c('----- Genotype :',geno,'| n°',i,' -----'),collapse=" "))
  
  # 2/ Change Simname and outputs
  print(paste(c('Step 1 : Initialize simulation...'),collapse=" "))
  ParamsToUse <- subset(MyParams,Hybrid_ID==as.character(geno))
  SimulationName<-paste(ParamsToUse[,'Hybrid_ID'])
  NewApsimName<-paste(c(SimulationName,"apsim"),collapse=".")
  
  # CHeck if we have values for the genotype :
  if (nrow(ParamsToUse)>0)
  {
    # Loop over genotypic parameters and change them :
    for (var in colnames(ParamsToUse[2:ncol(ParamsToUse)]))
    {
       VarNode<-getNodeSet(SimToUse, paste(c("//",var),collapse="") )
       NewValue<- ParamsToUse[,var]
       
      for (node in 1:length(VarNode))
      {
         ChangeXmlValue(VarNode[node],NewValue) 
      }
       
    }
    print(paste('==> Done !'),collapse=" ")
  
    # Save the new apsimfile in Model Directory :
    print(paste(c('Step 2 : Saving simulation file...'),collapse=" "))
    saveXML(xmlRoot(SimFileTree), 
            file= paste0(PATH.APSIM,NewApsimName), 
            prefix= '<?xml version="1.0" encoding="utf-8"?>',indent= T)
    print(paste('==> Done !'),collapse=" ")
  
    # Simulate :
    print(paste(c('Step 3 : Simulation...'),collapse=" "))
    setwd(PATH.APSIM)
    
    command<-paste("apsim",NewApsimName)

    if (WorkOnErrors==T)
    {
      result <- tryCatch(evalWithTimeout(system(command,show.output.on.console=T),
                                 timeout = 30,
                                 onTimeout = "error"),
                 error=
                   {
                        #Get all tasks running :
                        Tasks <- system('tasklist',show.output.on.console=T,intern=T)
                        # Kill jobrunner :
                        while (sum(grepl("JobRunner.exe", Tasks))>0)
                        {  
                          print('Killing JobRunner...')
                          system('taskkill /IM JobRunner.exe /F',
                                 show.output.on.console=F,intern=T)
                          Tasks <- system('tasklist',show.output.on.console=T,intern=T)
                        }
                        # Kill all Apsim.exe :
                        while (sum(grepl("Apsim.exe", Tasks))>0)
                        {
                          print('Killing APSIM...')
                          system('taskkill /IM Apsim.exe /F',
                                 show.output.on.console=F,intern=T)       
                          Tasks <- system('tasklist',show.output.on.console=T,intern=T)
                        }
                    }
                 )
      
      # If it runs smoothly :
      if (result==0) {print(paste('==> Done !'),collapse=" ")
      # If there was an error : 
      } else {print(paste('==> Error Handled !'),collapse=" ")}
    }
    else {system(command)}
      
    
    # Get/Save outputs  : 
    print(paste(c('Step 4 : Loading outputs...'),collapse=" "))
    
    Output_Sim <- loadApsim(PATH.APSIM, 
                            loadAll = TRUE, 
                            ext = ".out", 
                            returnFrame = TRUE, 
                            n = 0, fill = TRUE, 
                            addConstants = TRUE)
    
    Output_Sim$Genotype<-geno
    
  
    if (i ==1)
    {Global_Output_AllGenotypes <- Output_Sim
    }else{
     Global_Output_AllGenotypes <- bind_rows(Global_Output_AllGenotypes,Output_Sim)
    }
  }
  
  print(paste('==> Done !'),collapse=" ")  
  
}

Stop <- Sys.time()
TimeDiff <- difftime(Stop, Start, units="secs")
MyTimeDiff <- format(.POSIXct(TimeDiff,tz="GMT"), "%H:%M:%S")
print(paste(c(''),collapse=" "))
print(paste(c('----- Simulation loops finished -----'),collapse=" "))
print(paste(c(''),collapse=" "))
print(paste(c('Time of execution is '),MyTimeDiff,collapse=" "))


# 2/ Save and treat outputs : ----
save(file=paste0(PATHS[4],'/Global_Output_AllGenotypes.RData'),Global_Output_AllGenotypes)

```

```{r Data Analysis - Genotypic variability, eval=FALSE, message=FALSE, warning=FALSE, include=FALSE}

## Getting all data for harvest dates only (max DAS):
Get_Harvest_Data_Das <- function(Output_Sim,By)
{
  HarvestDateSim <-   
    do.call("rbind",lapply(split(Output_Sim,By,drop=T),
                         function(x)
                         {
                           Temp<-subset(x, DaysAfterSowing==max(x$DaysAfterSowing))
                           return(Temp)
                         }
                       ) ## End lapply
                      ) ## End do.call rbind  
  return(HarvestDateSim)
}
library(dplyr)

load('F:/Sebastien/Sims DROPS GxE/4.Out/Global_Output_AllGenotypes.RData')

head(Global_Output_AllGenotypes)

Harvest_PanelNetwork <- Get_Harvest_Data_Das(Global_Output_AllGenotypes,
                                         list(Global_Output_AllGenotypes$Title,
                                              Global_Output_AllGenotypes$Genotype))


Global_Genotypic_Analysis <- merge(Harvest_PanelNetwork,MyParams,by.x='Genotype',by.y='Hybrid_ID')
  
Anova_GE <- anova(lm(data=Global_Genotypic_Analysis,yield~Title+Genotype+Genotype*Title))

DataToTreat <- select(Global_Genotypic_Analysis,
                      yield,Biomass,Title,Genotype,a,b,c,SensiRAD,respTnight,First_LiguloChrone,Phyllochron,GNmax,potKernelWt)


ggplot(data=DataToTreat,aes(x=yield,y=a))+
  geom_point()

```
